<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Sarah Williams" />


<title>Dataset: IBD cosMX GarridoTrigo2023</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">spatialsnippets</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tests</a>
</li>
<li>
  <a href="index_data.html">Data</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/swbioinf/spatialsnippets">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Dataset: IBD cosMX GarridoTrigo2023</h1>
<h4 class="author">Sarah Williams</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-03-25
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>spatialsnippets/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20231017code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20231017)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20231017code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20231017)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree0a3a1ae402c4ecb2bb0bf5bfe879bed5e5981dd1targetblank0a3a1aea">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/swbioinf/spatialsnippets/tree/0a3a1ae402c4ecb2bb0bf5bfe879bed5e5981dd1" target="_blank">0a3a1ae</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree0a3a1ae402c4ecb2bb0bf5bfe879bed5e5981dd1targetblank0a3a1aea"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/swbioinf/spatialsnippets/tree/0a3a1ae402c4ecb2bb0bf5bfe879bed5e5981dd1" target="_blank">0a3a1ae</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/d_DigitsDish.nb.html
    Ignored:    analysis/d_cosmxIBD.nb.html
    Ignored:    analysis/e_CompositionChange.nb.html
    Ignored:    analysis/e_DEPseudobulk.nb.html
    Ignored:    analysis/e_DEWithoutReps.nb.html
    Ignored:    notes/
    Ignored:    renv/library/
    Ignored:    renv/staging/

Unstaged changes:
    Modified:   .gitignore
    Deleted:    code/d_cosmxLiver_subsetting.R
    Modified:   renv.lock

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/d_cosmxIBD.Rmd</code>) and HTML
(<code>docs/d_cosmxIBD.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/0a3a1ae402c4ecb2bb0bf5bfe879bed5e5981dd1/analysis/d_cosmxIBD.Rmd" target="_blank">0a3a1ae</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-25
</td>
<td>
wflow_publish("analysis/")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/d_cosmxIBD.html" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/6295b7d314030df2d6cbc32d1c08ab215a1c8389/analysis/d_cosmxIBD.Rmd" target="_blank">6295b7d</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
<td>
wflow_publish(c("analysis/index.Rmd", "analysis/d_cosmxIBD.Rmd"))
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>(FIX SPELLING IN OBJECT) of CHRONES</p>
<p>Data is from paper <a
href="https://www.ncbi.nlm.nih.gov/pubmed/37495570"><em>Macrophage and
neutrophil heterogeneity at single-cell spatial resolution in human
inflammatory bowel disease</em></a> from Garrido-Trigo et al 2023.</p>
<p>The study included 9 cosmx slides of colonic biopsies</p>
<ul>
<li>3x HC - Healthy controls</li>
<li>3x UC - Ulcerative colitis</li>
<li>3x CD - Chrones’s disease</li>
</ul>
<p>Fastantically - not only have they made their raw and annotated data
available, but have also shared their analysis code; <a
href="https://github.com/HelenaLC/CosMx-SMI-IBD"
class="uri">https://github.com/HelenaLC/CosMx-SMI-IBD</a></p>
<p>They have also shared browseable interface here: <a
href="https://servidor2-ciberehd.upc.es/external/garrido/app/"
class="uri">https://servidor2-ciberehd.upc.es/external/garrido/app/</a></p>
<div id="libraries" class="section level1">
<h1>Libraries</h1>
<pre class="r"><code>library(Seurat)</code></pre>
<pre><code>Loading required package: SeuratObject</code></pre>
<pre><code>Loading required package: sp</code></pre>
<pre><code>
Attaching package: &#39;SeuratObject&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    intersect</code></pre>
<pre class="r"><code>library(SeuratObject)
library(tidyverse)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.4     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     </code></pre>
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div id="data-download" class="section level1">
<h1>Data download</h1>
<p>From the project description:</p>
<p>Raw data on GEO here; <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE234713"
class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE234713</a></p>
<p><em>Inflammatory bowel diseases (IBDs) including ulcerative colitis
(UC) and Crohn’s disease (CD) are chronic inflammatory diseases with
increasing worldwide prevalence that show a perplexing heterogeneity in
manifestations and response to treatment. We applied spatial
transcriptomics at single-cell resolution (CosMx Spatial Molecular
Imaging) to human inflamed and uninflamed intestine.</em></p>
<p>The following files were download from GEO</p>
<ul>
<li>GSE234713_CosMx_annotation.csv.gz</li>
<li>GSE234713_CosMx_normalized_matrix.txt.gz</li>
<li>GSE234713_RAW.tar: RAW data was downloaded via custom downlod in 3
batches, one per group
<ul>
<li>GSE234713_RAW_CD.tar</li>
<li>GSE234713_RAW_HC.tar</li>
<li>GSE234713_RAW_UC.tar</li>
</ul></li>
<li>GSE234713_ReadMe_SMI_Data_File.html</li>
</ul>
<pre class="sh"><code>cd raw_data
tar -xf GSE234713_RAW_CD.tar
tar -xf GSE234713_RAW_HC.tar
tar -xf GSE234713_RAW_UC.tar</code></pre>
</div>
<div id="data-load" class="section level1">
<h1>Data load</h1>
<pre class="r"><code>dataset_dir      &lt;- &#39;~/data_local/datasets/&#39;
project_data_dir &lt;- file.path(dataset_dir,&#39;GSE234713_IBDcosmx_GarridoTrigo2023&#39;)
sample_dir            &lt;- file.path(project_data_dir, &quot;raw_data/&quot;) 
annotation_file       &lt;- file.path(project_data_dir,&quot;GSE234713_CosMx_annotation.csv.gz&quot;)
data_dir              &lt;- file.path(project_data_dir, &quot;processed_data/&quot;) 


seurat_file_01_loaded &lt;- file.path(data_dir, &quot;GSE234713_CosMx_IBD_seurat_01_loaded.RDS&quot;)

# config
min_count_per_cell &lt;- 100
max_pc_negs        &lt;- 1.5
max_avg_neg        &lt;- 0.5



sample_codes &lt;- c(HC=&quot;Healthy controls&quot;,UC=&quot;Ulcerative colitis&quot;,CD=&quot;Chrones&#39;s disease&quot;)</code></pre>
<div id="raw-data" class="section level2">
<h2>Raw data</h2>
</div>
<div id="load-data" class="section level2">
<h2>Load data</h2>
<pre class="r"><code>load_sample_into_seurat &lt;- function(the_sample){
  
  metadata_cols &lt;- c(&quot;fov&quot;,&quot;Area&quot;) 
  
  sample_metadata_file &lt;- file.path(sample_dir, paste0(the_sample,&#39;_metadata_file.csv.gz&#39;))
  sample_mtx_file      &lt;- file.path(sample_dir, paste0(the_sample,&#39;_exprMat_file.csv.gz&#39;))
  sample_molecules_file &lt;- file.path(sample_dir, paste0(the_sample,&#39;_tx_file.csv.gz&#39;))
  
  ns &lt;- ReadNanostring(
                 data.dir           = sample_dir,
                 mtx.file           = sample_mtx_file,
                 metadata.file      = sample_metadata_file,
                 molecules.file     = sample_molecules_file,
                 segmentations.file = NULL,
                 metadata           = metadata_cols , # Can only draw from selected feilds. So add later. Using ensures table-ness.
                 type=&quot;centroids&quot;
        )
  
  # Add the rest of the metadata
  metadata_table &lt;- read_csv(sample_metadata_file)
  
  # Check order matches.
  stopifnot(all(paste0(metadata_table$cell_ID, &quot;_&quot;, metadata_table$fov)  == ns$metadata$cell)) #im not paranoid, who&#39;s paranoid? not me.
  # pull the whole of annodation in now, and shuffle column order.
  ns$metadata &lt;- cbind(ns$metadata, metadata_table[,!colnames(metadata_table) %in% metadata_cols])
  ns$metadata &lt;- ns$metadata[, c(2,1,3:ncol(ns$metadata))]
  rownames(ns$metadata) &lt;- ns$metadata$cell
  
  cents  &lt;- CreateCentroids(ns$centroids)
  coords &lt;- CreateFOV(coords = list(&quot;centroids&quot; = cents), type = &quot;centroids&quot;)
  so     &lt;- CreateSeuratObject(counts    = ns$matrix,
                               meta.data = ns$metadata)
  so$orig.ident &lt;- the_sample
  
  # sample info
  so$group     &lt;- substr(the_sample, 12, 13)
  so$group     &lt;- factor(so$group, levels=names(sample_codes))
  so$condition &lt;- factor(as.character(sample_codes[so$group]), levels=sample_codes)
  

  # Put neg probes into their own assay.
  neg_probes &lt;- rownames(so)[grepl(x=rownames(so), pattern=&quot;NegPrb&quot;)]
  neg_matrix         &lt;- GetAssayData(so,assay = &#39;RNA&#39;, layer = &#39;counts&#39;)[neg_probes,]
  so[[&quot;negprobes&quot;]] &lt;- CreateAssayObject(counts = neg_matrix)
  
  ## and remove from the main one
  #rna_probes &lt;- rownames(so)[(! rownames(so) %in% neg_probes)]
  #so         &lt;- so[rna_probes,]
  

  return(so)
  
}</code></pre>
<pre class="r"><code>samples &lt;- c(&#39;GSM7473682_HC_a&#39;,&#39;GSM7473683_HC_b&#39;,&#39;GSM7473684_HC_c&#39;,
             &#39;GSM7473685_UC_a&#39;,&#39;GSM7473686_UC_b&#39;,&#39;GSM7473687_UC_c&#39;,
             &#39;GSM7473688_CD_a&#39;,&#39;GSM7473689_CD_b&#39;,&#39;GSM7473690_CD_c&#39;)
sample_prefix &lt;- paste0(substr(samples, 12,15))

so.list &lt;- lapply(FUN=load_sample_into_seurat, X=samples)</code></pre>
<pre><code>Rows: 42636 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 54818 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 31340 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 50757 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 78115 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 57436 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 38779 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 72837 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 55559 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
dbl (20): fov, cell_ID, Area, AspectRatio, CenterX_local_px, CenterY_local_p...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>#NB: merge is in SeuratObject packages, but must be called without :: 
so.raw &lt;- merge(so.list[[1]], y=so.list[2:length(so.list)], add.cell.ids=sample_prefix)
rm(so.list)</code></pre>
</div>
</div>
<div id="negative-probe-handling" class="section level1">
<h1>Negative probe handling</h1>
<pre class="r"><code>so.raw$pc_neg &lt;-  ( so.raw$nCount_negprobes / so.raw$nCount_RNA ) * 100
so.raw$avg_neg &lt;-  colMeans(so.raw[[&quot;negprobes&quot;]])</code></pre>
<!-- ```{r} -->
<!-- # Put neg probes into their own assay. -->
<!-- neg_probes <- rownames(so.raw)[grepl(x=rownames(so.raw), pattern="NegPrb")] -->
<!-- neg_matrix         <- GetAssayData(so.raw,assay = 'RNA', layer = 'counts')[neg_probes,] -->
<!-- so.raw[["negprobes"]] <- CreateAssayObject(counts = neg_matrix) -->
<!-- ## and remove from the main one -->
<!-- rna_probes <- rownames(so)[(! rownames(so.raw) %in% neg_probes)] -->
<!-- so.raw         <- so[rna_probes,] -->
<!-- ``` -->
</div>
<div id="pull-in-annotation" class="section level1">
<h1>Pull in annotation</h1>
<p>Most cell annotations.</p>
<pre class="r"><code>anno_table &lt;- read_csv(annotation_file)</code></pre>
<pre><code>Rows: 463967 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
chr (3): id, subset, SingleR2

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>anno_table &lt;- as.data.frame(anno_table)
rownames(anno_table) &lt;- anno_table$id

head(so.raw@meta.data)</code></pre>
<pre><code>              orig.ident nCount_RNA nFeature_RNA  Area fov cell cell_ID
HC_a_1_1 GSM7473682_HC_a        368          189  6153   1  1_1       1
HC_a_2_1 GSM7473682_HC_a        810          286 16178   1  2_1       2
HC_a_3_1 GSM7473682_HC_a        119           74  3119   1  3_1       3
HC_a_4_1 GSM7473682_HC_a        106           59  3988   1  4_1       4
HC_a_5_1 GSM7473682_HC_a        465          209  4773   1  5_1       5
HC_a_6_1 GSM7473682_HC_a       1322          441 11588   1  6_1       6
         AspectRatio CenterX_local_px CenterY_local_px CenterX_global_px
HC_a_1_1        0.67             2119             3443          19274.56
HC_a_2_1        0.85             1847             3404          19002.56
HC_a_3_1        1.52             1986             3458          19141.56
HC_a_4_1        1.52             2362             3454          19517.56
HC_a_5_1        0.53             2159             3415          19314.56
HC_a_6_1        0.87             1664             3402          18819.56
         CenterY_global_px Width Height Mean.MembraneStain Max.MembraneStain
HC_a_1_1          173198.6    79    118                680              2218
HC_a_2_1          173159.6   137    161                615              4453
HC_a_3_1          173213.6    79     52                989              2827
HC_a_4_1          173209.6    91     60                449              2099
HC_a_5_1          173170.6    63    118               1002              2425
HC_a_6_1          173157.6   116    134               1024              2320
         Mean.PanCK Max.PanCK Mean.CD45 Max.CD45 Mean.CD3 Max.CD3 Mean.DAPI
HC_a_1_1      17170     29735       198     9639      234    1237        20
HC_a_2_1      16775     29702       222    21371      234    1520        12
HC_a_3_1      24033     29662       171     6858      298    1721        13
HC_a_4_1      13575     29651       146      935      206    1402         6
HC_a_5_1      24946     29670       299     6387      330    1443        22
HC_a_6_1      25700     29712       354    23243      328    1610        23
         Max.DAPI group        condition nCount_negprobes nFeature_negprobes
HC_a_1_1      131    HC Healthy controls                2                  2
HC_a_2_1      122    HC Healthy controls                5                  3
HC_a_3_1       58    HC Healthy controls                1                  1
HC_a_4_1       56    HC Healthy controls                0                  0
HC_a_5_1      116    HC Healthy controls                3                  2
HC_a_6_1      113    HC Healthy controls                8                  5
            pc_neg    avg_neg
HC_a_1_1 0.5434783 0.10526316
HC_a_2_1 0.6172840 0.26315789
HC_a_3_1 0.8403361 0.05263158
HC_a_4_1 0.0000000 0.00000000
HC_a_5_1 0.6451613 0.15789474
HC_a_6_1 0.6051437 0.42105263</code></pre>
<pre class="r"><code>head(anno_table)</code></pre>
<pre><code>               id   subset             SingleR2
HC_c_2_1 HC_c_2_1   stroma            Pericytes
HC_c_3_1 HC_c_3_1   stroma          Endothelium
HC_c_4_1 HC_c_4_1   stroma            Pericytes
HC_c_5_1 HC_c_5_1      epi Secretory progenitor
HC_c_7_1 HC_c_7_1 myeloids                   M2
HC_c_8_1 HC_c_8_1      epi           Tuft cells</code></pre>
<pre class="r"><code>so.raw$full_cell_id &lt;- as.character(rownames(so.raw@meta.data))

so.raw$celltype_subset   &lt;- anno_table[so.raw$full_cell_id,]$subset
so.raw$celltype_SingleR2 &lt;- anno_table[so.raw$full_cell_id,]$SingleR2

table(is.na(so.raw$celltype_subset))</code></pre>
<pre><code>
 FALSE   TRUE 
459145  21240 </code></pre>
</div>
<div id="basic-qc-filter" class="section level1">
<h1>Basic QC filter</h1>
<div id="min-count-per-cell" class="section level2">
<h2>Min count per cell</h2>
<pre class="r"><code>ggplot(so.raw@meta.data, aes(x=nCount_RNA, col=orig.ident)) +
  geom_density() + 
  geom_vline(xintercept = min_count_per_cell, lty=3) +
  scale_x_log10() +
  theme_bw() +
  ggtitle(&quot;Counts per cell&quot;)</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-8-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="percent-negative-probes" class="section level2">
<h2>Percent Negative probes</h2>
<pre class="r"><code>ggplot(so.raw@meta.data, aes(x=pc_neg, col=orig.ident)) +
  geom_density() + 
  geom_vline(xintercept = max_pc_negs, lty=3) +
  scale_x_log10() +
  theme_bw() +
  ggtitle(&quot;Negative probe composition&quot;)</code></pre>
<pre><code>Warning: Transformation introduced infinite values in continuous x-axis</code></pre>
<pre><code>Warning: Removed 227705 rows containing non-finite values (`stat_density()`).</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-9-1">
Past versions of unnamed-chunk-9-1.png
</button>
</p>
<div id="fig-unnamed-chunk-9-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-9-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>ggplot(so.raw@meta.data, aes(x=avg_neg, col=orig.ident)) +
  geom_density() + 
  geom_vline(xintercept = max_avg_neg, lty=3) +
  theme_bw() +
  ggtitle(&quot;Negative probe average&quot;)</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-10-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Use bottom left corner;</p>
<pre class="r"><code>ggplot(so.raw@meta.data, aes(y=avg_neg, x=nCount_RNA)) +
  geom_point(pch=3, alpha=0.1) + 
  geom_hline(yintercept = max_avg_neg, lty=3) +
  geom_vline(xintercept = min_count_per_cell, lty=3) +
  scale_x_log10() + 
  theme_bw() +
  ggtitle(&quot;Negative probes vs counts&quot;)</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-1">
Past versions of unnamed-chunk-11-1.png
</button>
</p>
<div id="fig-unnamed-chunk-11-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-11-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="apply-filteres" class="section level2">
<h2>Apply filteres</h2>
<p>From paper: <em>“Cells with an average negative control count greater
than 0.5 and less than 20 detected features were filtered out.”</em></p>
<p>Also remove cell with no cell type annotations.</p>
<pre class="r"><code>so &lt;- so.raw[ ,so.raw$nCount_RNA &gt;= min_count_per_cell &amp;
               so.raw$avg_neg &lt;= max_avg_neg &amp;
               !(is.na(so.raw$celltype_subset) )]
ncol(so.raw)</code></pre>
<pre><code>[1] 480385</code></pre>
<pre class="r"><code>ncol(so)</code></pre>
<pre><code>[1] 354191</code></pre>
<pre class="r"><code>so &lt;- readRDS(seurat_file_01_loaded)</code></pre>
</div>
<div id="how-many-cells-per-sample" class="section level2">
<h2>How many cells per sample?</h2>
<pre class="r"><code>table(so@meta.data$orig.ident)</code></pre>
<pre><code>
GSM7473682_HC_a GSM7473683_HC_b GSM7473684_HC_c GSM7473685_UC_a GSM7473686_UC_b 
          30426           46203           15482           44070           60127 
GSM7473687_UC_c GSM7473688_CD_a GSM7473689_CD_b GSM7473690_CD_c 
          38439           14579           69147           35718 </code></pre>
</div>
</div>
<div id="basic-preprocessing" class="section level1">
<h1>Basic preprocessing</h1>
<pre class="r"><code>num_dims &lt;- 15
# Run through preprocessing
so &lt;- NormalizeData(so)

## Do per sample to mimic paper approach somewhat.
so &lt;- FindVariableFeatures(so, nfeatures = 200) 


so &lt;- ScaleData(so) # Just 2k variable features
so &lt;- RunPCA(so, features = VariableFeatures(so))
so &lt;- RunUMAP(so, dims=1:num_dims)
so &lt;- FindNeighbors(so, dims = 1:num_dims)
so &lt;- FindClusters(so)</code></pre>
</div>
<div id="basic-plots" class="section level1">
<h1>Basic plots</h1>
<div id="umap" class="section level2 tabset tabset-pills">
<h2 class="tabset tabset-pills">UMAP</h2>
<div id="sample" class="section level3">
<h3>Sample</h3>
<pre class="r"><code>DimPlot(so, group.by = &#39;orig.ident&#39;)</code></pre>
<pre><code>Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-16-1">
Past versions of unnamed-chunk-16-1.png
</button>
</p>
<div id="fig-unnamed-chunk-16-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-16-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="condition" class="section level3">
<h3>Condition</h3>
<pre class="r"><code>DimPlot(so, group.by = &#39;condition&#39;)</code></pre>
<pre><code>Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-17-1">
Past versions of unnamed-chunk-17-1.png
</button>
</p>
<div id="fig-unnamed-chunk-17-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-17-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="condition-x-sample" class="section level3">
<h3>Condition X Sample</h3>
<pre class="r"><code>DimPlot(so, group.by = &#39;orig.ident&#39;,split.by = &quot;condition&quot;)</code></pre>
<pre><code>Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-18-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-18-1">
Past versions of unnamed-chunk-18-1.png
</button>
</p>
<div id="fig-unnamed-chunk-18-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-18-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="total-counts" class="section level3">
<h3>Total counts</h3>
<pre class="r"><code>FeaturePlot(so, &#39;nCount_RNA&#39;) + scale_colour_viridis_c(option=&quot;magma&quot;, direction = -1)</code></pre>
<pre><code>Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`</code></pre>
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-19-1">
Past versions of unnamed-chunk-19-1.png
</button>
</p>
<div id="fig-unnamed-chunk-19-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-19-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="clustered" class="section level3">
<h3>Clustered</h3>
<pre class="r"><code>DimPlot(so, group.by = &#39;seurat_clusters&#39;)</code></pre>
<pre><code>Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-20-1">
Past versions of unnamed-chunk-20-1.png
</button>
</p>
<div id="fig-unnamed-chunk-20-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-20-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="celltype_subset" class="section level3">
<h3>celltype_subset</h3>
<p>Classifications from Garrido-Trigo et al 2023.</p>
<pre class="r"><code>DimPlot(so, group.by = &#39;celltype_subset&#39;)</code></pre>
<pre><code>Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-21-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="celltype_singler2" class="section level3">
<h3>celltype_SingleR2</h3>
<p>Classifications from Garrido-Trigo et al 2023.</p>
<pre class="r"><code>DimPlot(so, group.by = &#39;celltype_SingleR2&#39;)</code></pre>
<pre><code>Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-1">
Past versions of unnamed-chunk-22-1.png
</button>
</p>
<div id="fig-unnamed-chunk-22-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-22-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<p>Double checking the clusters line up with annotaiaon</p>
<pre class="r"><code>table(so$celltype_subset)</code></pre>
<pre><code>
     epi myeloids  plasmas   stroma   tcells 
   82743    44006   130781    84915    11746 </code></pre>
<pre class="r"><code>table(so$celltype_SingleR2)</code></pre>
<pre><code>
                      B cell                  BEST4 OTOP2 
                        5314                        10155 
                         CD4                          CD8 
                        2733                         2881 
                 Colonocytes                Cycling cells 
                       17673                         2112 
             Cycling myeloid              Cycling T cells 
                        1287                          323 
                  Cycling TA                          DCs 
                        1846                         1469 
                          DN                  Endothelium 
                         608                        13808 
             Enteroendocrine                  Eosinophils 
                         852                          194 
            Epithelium Ribhi                  Fibroblasts 
                        4379                         1666 
                        FRCs                    GC B cell 
                       10027                         3566 
                      gd IEL                         Glia 
                         367                         4353 
                      Goblet                         ILC4 
                        8864                          115 
    Inflammatory fibroblasts       Inflammatory monocytes 
                       12593                         1735 
                          M0                           M1 
                        3001                          937 
                          M2              Macrophage NRG1 
                       21296                         9581 
                        MAIT                         Mast 
                        1233                         2939 
               Memory B cell                   MT T cells 
                        7040                          175 
              Myofibroblasts                           N1 
                       16697                          556 
                          N2                           N3 
                         433                          578 
                Naïve B cell                           NK 
                        8906                          240 
                 Paneth-like PC  immediate early response 
                        1820                         9826 
                      PC IgA            PC IgA heat shock 
                        1299                         5090 
                  PC IgA IgM                       PC IgG 
                       12245                        75381 
                   Pericytes                Ribhi T cells 
                        9303                           87 
                          S1                          S2a 
                        5214                         3567 
                         S2b                           S3 
                        4606                         3081 
        Secretory progenitor                T cells CCL20 
                       34665                          817 
                       Tregs                   Tuft cells 
                        2167                         2489 </code></pre>
<p>Composition</p>
<pre class="r"><code>heatmap(table(so$celltype_subset,so$celltype_SingleR2))</code></pre>
<p><img src="figure/d_cosmxIBD.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/cf8a2ee50b7775199ac0db3b77a063375425ccd9/docs/figure/d_cosmxIBD.Rmd/unnamed-chunk-24-1.png" target="_blank">cf8a2ee</a>
</td>
<td>
Sarah Williams
</td>
<td>
2024-03-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="save-data" class="section level1">
<h1>Save data</h1>
<pre class="r"><code>saveRDS(so, seurat_file_01_loaded)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Brisbane
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] lubridate_1.9.3    forcats_1.0.0      stringr_1.5.0      dplyr_1.1.3       
 [5] purrr_1.0.2        readr_2.1.4        tidyr_1.3.0        tibble_3.2.1      
 [9] ggplot2_3.4.4      tidyverse_2.0.0    Seurat_5.0.1       SeuratObject_5.0.0
[13] sp_2.1-1           workflowr_1.7.0   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.15.0      jsonlite_1.8.7        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   farver_2.1.1          
  [7] rmarkdown_2.23         fs_1.6.3               vctrs_0.6.3           
 [10] ROCR_1.0-11            spatstat.explore_3.2-5 htmltools_0.5.5       
 [13] sass_0.4.7             sctransform_0.4.1      parallelly_1.36.0     
 [16] KernSmooth_2.23-22     bslib_0.5.0            htmlwidgets_1.6.2     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.10.3         
 [22] zoo_1.8-12             cachem_1.0.8           whisker_0.4.1         
 [25] igraph_1.5.1           mime_0.12              lifecycle_1.0.3       
 [28] pkgconfig_2.0.3        Matrix_1.6-1.1         R6_2.5.1              
 [31] fastmap_1.1.1          fitdistrplus_1.1-11    future_1.33.0         
 [34] shiny_1.7.5.1          digest_0.6.33          colorspace_2.1-0      
 [37] patchwork_1.1.3        ps_1.7.5               rprojroot_2.0.3       
 [40] tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1         
 [43] labeling_0.4.3         progressr_0.14.0       timechange_0.2.0      
 [46] fansi_1.0.5            spatstat.sparse_3.0-3  httr_1.4.6            
 [49] polyclip_1.10-6        abind_1.4-5            compiler_4.3.1        
 [52] bit64_4.0.5            withr_2.5.1            fastDummies_1.7.3     
 [55] highr_0.10             R.utils_2.12.3         MASS_7.3-60           
 [58] tools_4.3.1            lmtest_0.9-40          httpuv_1.6.11         
 [61] future.apply_1.11.0    goftest_1.2-3          R.oo_1.25.0           
 [64] glue_1.6.2             callr_3.7.3            nlme_3.1-162          
 [67] promises_1.2.0.1       grid_4.3.1             Rtsne_0.16            
 [70] getPass_0.2-2          cluster_2.1.4          reshape2_1.4.4        
 [73] generics_0.1.3         gtable_0.3.4           spatstat.data_3.0-3   
 [76] tzdb_0.4.0             R.methodsS3_1.8.2      hms_1.1.3             
 [79] data.table_1.14.8      utf8_1.2.4             spatstat.geom_3.2-7   
 [82] RcppAnnoy_0.0.21       ggrepel_0.9.4          RANN_2.6.1            
 [85] pillar_1.9.0           vroom_1.6.4            spam_2.10-0           
 [88] RcppHNSW_0.5.0         later_1.3.1            splines_4.3.1         
 [91] lattice_0.21-8         bit_4.0.5              renv_1.0.0            
 [94] survival_3.5-7         deldir_1.0-9           tidyselect_1.2.0      
 [97] miniUI_0.1.1.1         pbapply_1.7-2          knitr_1.43            
[100] git2r_0.32.0           gridExtra_2.3          scattermore_1.2       
[103] xfun_0.39              matrixStats_1.0.0      stringi_1.7.12        
[106] lazyeval_0.2.2         yaml_2.3.7             evaluate_0.21         
[109] codetools_0.2-19       BiocManager_1.30.22    cli_3.6.1             
[112] uwot_0.1.16            xtable_1.8-4           reticulate_1.34.0     
[115] munsell_0.5.0          processx_3.8.2         jquerylib_0.1.4       
[118] Rcpp_1.0.11            globals_0.16.2         spatstat.random_3.2-1 
[121] png_0.1-8              parallel_4.3.1         ellipsis_0.3.2        
[124] dotCall64_1.1-0        listenv_0.9.0          viridisLite_0.4.2     
[127] scales_1.2.1           ggridges_0.5.4         crayon_1.5.2          
[130] leiden_0.4.3           rlang_1.1.1            cowplot_1.1.1         </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
