<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Sarah Williams" />


<title>Dataset: Digits in a dish.</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">spatialsnippets</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tests</a>
</li>
<li>
  <a href="index_data.html">Data</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/swbioinf/spatialsnippets">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Dataset: Digits in a dish.</h1>
<h4 class="author">Sarah Williams</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-11-01
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>spatialsnippets/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20231017code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20231017)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20231017code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20231017)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree8b1c005f8c609ac8c7c66b28c945c9d6dccd1b1etargetblank8b1c005a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/swbioinf/spatialsnippets/tree/8b1c005f8c609ac8c7c66b28c945c9d6dccd1b1e" target="_blank">8b1c005</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree8b1c005f8c609ac8c7c66b28c945c9d6dccd1b1etargetblank8b1c005a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/swbioinf/spatialsnippets/tree/8b1c005f8c609ac8c7c66b28c945c9d6dccd1b1e" target="_blank">8b1c005</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/d_DigitsDish.nb.html
    Ignored:    analysis/e_DEPseudobulk.nb.html
    Ignored:    analysis/e_DEWithoutReps.nb.html
    Ignored:    renv/library/
    Ignored:    renv/staging/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/d_DigitsDish.Rmd</code>) and HTML
(<code>docs/d_DigitsDish.html</code>) files. If you’ve configured a
remote Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/8b1c005f8c609ac8c7c66b28c945c9d6dccd1b1e/analysis/d_DigitsDish.Rmd" target="_blank">8b1c005</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-11-01
</td>
<td>
wflow_publish("analysis")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Preparation of data from: <em>Digits in a dish: An in vitro system to
assess the molecular genetics of hand/foot development at single-cell
resolution</em> Allison M. Fuiten, Yuki Yoshimoto, Chisa Shukunami, H.
Scott Stadler. Fronteirs in Cell and Developmental Biology 2023.</p>
<p><a
href="https://www.frontiersin.org/articles/10.3389/fcell.2023.1135025/full"
class="uri">https://www.frontiersin.org/articles/10.3389/fcell.2023.1135025/full</a></p>
<p>Data from GEO, GSE221883. <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE221883"
class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE221883</a></p>
<p>Data processing here is simplified for demonstrative purposes - and
differs from that used in the paper!</p>
<!-- Package install
Seurat5
#renv::install('satijalab/seurat-wrappers@seurat5') # Yes.
#renv::install('satijalab/seurat@seurat5')
-->
<div id="libraries" class="section level1">
<h1>Libraries</h1>
<pre class="r"><code>library(Seurat)</code></pre>
<pre><code>Loading required package: SeuratObject</code></pre>
<pre><code>Loading required package: sp</code></pre>
<pre><code>
Attaching package: &#39;SeuratObject&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    intersect</code></pre>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.4     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     </code></pre>
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div id="data-download" class="section level1">
<h1>Data Download</h1>
<p>Download counts matricies from GEO. Note the read10X function used
later expects a folder per sample with files exactly named
barcodes.tsv.gz/features.tsv.gz and matrix.mtx.gz</p>
<pre class="sh"><code>wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE221883&amp;format=file
tar -xzf GSE221883_RAW.tar

mkdir data_for_seurat



mkdir data_for_seurat
mkdir seurat_objects
for sample in GSM6908653_Day2_A GSM6908655_Day7_A GSM6908657_Day10_A GSM6908654_Day2_B GSM6908656_Day7_B GSM6908658_Day10_B
do
echo ${sample}
mkdir data_for_seurat/${sample}
cp ${sample}_barcodes.tsv.gz data_for_seurat/${sample}/barcodes.tsv.gz
cp ${sample}_features.tsv.gz data_for_seurat/${sample}/features.tsv.gz
cp ${sample}_matrix.mtx.gz data_for_seurat/${sample}/matrix.mtx.gz
done</code></pre>
<p>Contains the following files:</p>
<pre><code>GSM6908653_Day2_A_barcodes.tsv.gz       
GSM6908653_Day2_A_features.tsv.gz       
GSM6908653_Day2_A_matrix.mtx.gz     
GSM6908654_Day2_B_barcodes.tsv.gz       
GSM6908654_Day2_B_features.tsv.gz       
GSM6908654_Day2_B_matrix.mtx.gz     
GSM6908655_Day7_A_barcodes.tsv.gz       
GSM6908655_Day7_A_features.tsv.gz       
GSM6908655_Day7_A_matrix.mtx.gz     
GSM6908656_Day7_B_barcodes.tsv.gz       
GSM6908656_Day7_B_features.tsv.gz       
GSM6908656_Day7_B_matrix.mtx.gz     
GSM6908657_Day10_A_barcodes.tsv.gz      
GSM6908657_Day10_A_features.tsv.gz      
GSM6908657_Day10_A_matrix.mtx.gz        
GSM6908658_Day10_B_barcodes.tsv.gz      
GSM6908658_Day10_B_features.tsv.gz      
GSM6908658_Day10_B_matrix.mtx.g     </code></pre>
</div>
<div id="data-load" class="section level1">
<h1>Data Load</h1>
<pre class="r"><code>data_dir &lt;- &#39;/Users/s2992547/data_local/datasets/GSE221883_DigitsDish_ScRNAseq/data_for_seurat/&#39;
seurat_objects_dir &lt;- &#39;/Users/s2992547/data_local/datasets/GSE221883_DigitsDish_ScRNAseq/seurat_objects/&#39;</code></pre>
<pre class="r"><code>samples &lt;- list.files(data_dir)

sample_dirs &lt;- file.path(data_dir, samples)
names(sample_dirs) &lt;- samples

data &lt;- Read10X(data.dir = sample_dirs)
so   &lt;- CreateSeuratObject(counts = data, project = &quot;Fuiten2023&quot;)

so[[&quot;percent.mt&quot;]] &lt;- PercentageFeatureSet(so, pattern = &quot;^mt-&quot;)</code></pre>
<pre class="r"><code>VlnPlot(so, features = c(&quot;nFeature_RNA&quot;))</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;RNA&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<p><img src="figure/d_DigitsDish.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>VlnPlot(so, features = c(&quot;nCount_RNA&quot;)) + scale_y_log10()</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;RNA&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<pre><code>Scale for y is already present.
Adding another scale for y, which will replace the existing scale.</code></pre>
<p><img src="figure/d_DigitsDish.Rmd/unnamed-chunk-5-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>VlnPlot(so, features = c(&quot;percent.mt&quot;))</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;RNA&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<p><img src="figure/d_DigitsDish.Rmd/unnamed-chunk-5-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>Add basic sample information.</p>
<pre class="r"><code>anno_table &lt;- as_tibble(str_split_fixed(rownames(so@meta.data), &quot;_&quot;, n = 4 ))</code></pre>
<pre><code>Warning: The `x` argument of `as_tibble.matrix()` must have unique column names if
`.name_repair` is omitted as of tibble 2.0.0.
ℹ Using compatibility `.name_repair`.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
<pre class="r"><code>colnames(anno_table) &lt;- c(&quot;Accession&quot;, &quot;Day&quot;,&quot;Rep&quot;,&quot;Cell&quot;)
so[[&quot;Sample&quot;]] &lt;- paste(anno_table$Day, anno_table$Rep, anno_table$Accession, sep=&quot;_&quot;)
so[[&quot;Accession&quot;]] &lt;- anno_table$Accession
so[[&quot;Day&quot;]] &lt;- anno_table$Day
so[[&quot;Rep&quot;]] &lt;- anno_table$Rep
so[[&quot;Cell&quot;]] &lt;- anno_table$Cell</code></pre>
<p>Do routine processing (absolutely not optimised for this study, just
need something reasonable.)</p>
<pre class="r"><code>so &lt;- subset(so, subset = nFeature_RNA &gt; 2000 &amp; percent.mt &lt; 25)
so &lt;- NormalizeData(so)</code></pre>
<pre><code>Normalizing layer: counts</code></pre>
<pre class="r"><code>so &lt;- FindVariableFeatures(so, selection.method = &quot;vst&quot;, nfeatures = 2000)</code></pre>
<pre><code>Finding variable features for layer counts</code></pre>
<pre class="r"><code>so &lt;- ScaleData(so) # No cc regression.</code></pre>
<pre><code>Centering and scaling data matrix</code></pre>
<pre class="r"><code>so &lt;- RunPCA(so, features = VariableFeatures(object = so))</code></pre>
<pre><code>PC_ 1 
Positive:  Tmsb4x, Actb, Basp1, Lgals1, Tmsb10, Tagln, Actg1, Tpm4, Abracl, Tuba1a 
       Tpm1, Tagln2, Myl9, Acta2, Pdlim7, Jpt1, Hmga2, Dstn, Filip1l, Tpm2 
       Myl12a, Cks2, Cald1, Cnn1, Flna, Pclaf, Csrp1, Msn, Actn1, Rtn4 
Negative:  Col2a1, Col9a2, Col9a3, Col11a1, Hapln1, Col11a2, Acan, Col9a1, Matn1, Mia 
       Col27a1, Cnmd, Snorc, Comp, Susd5, S100b, Fgfr3, Csgalnact1, Papss2, Matn4 
       Matn3, Chadl, Bnip3, Ncmap, Scrg1, Cpe, Scin, Fbln7, Higd1a, Cmtm5 
PC_ 2 
Positive:  Hist1h1e, Hmgb2, Stmn1, Lmnb1, Hist1h2ap, Hist1h4d, Hist1h1b, Hist1h2ae, Hmgb3, Top2a 
       Hist1h3c, Hist1h4h, Mki67, Pclaf, Smc2, Hist1h3e, Kif15, H1fx, Cdca8, Kif11 
       Nusap1, Hist1h1d, Nnat, Hist1h1a, H2afx, Cenpe, Sox11, Knl1, Birc5, Spc25 
Negative:  Sparc, S100a6, Bgn, Timp1, Gsn, Col1a2, Tspo, Timp2, Lmna, Anxa2 
       Igfbp7, Col1a1, Thbs1, Vim, Nupr1, Lox, Ctsl, Mmp23, Anxa1, Serpinb6a 
       Ifitm3, Ctsd, Ccn4, S100a4, Cst3, Ass1, Cyba, Cdkn2a, Cryab, Fbln5 
PC_ 3 
Positive:  Cald1, Tpm1, Myh10, Fermt2, Sparc, Fn1, Tagln, Myl9, Col12a1, Cnn2 
       Thbs1, Lpp, Col1a1, Igfbp7, Mfap4, Sox4, Ltbp1, Acta2, Nxn, Prss23 
       Cnn1, Fbln2, Palld, Tpm2, Vcl, Ccn4, Col4a1, Phldb2, Ptn, Fbln5 
Negative:  Fcer1g, Tyrobp, Trem2, Laptm5, Spi1, Csf1r, Ctss, Lcp1, C3ar1, Rac2 
       Pld4, Cd68, Clec4d, Coro1a, Cd14, Ncf4, Fcgr3, Adgre1, Mpeg1, Ncf2 
       Gmfg, C5ar1, Ms4a7, Msr1, Ms4a6d, Cd53, Fyb, Cxcl2, Lilrb4a, C1qa 
PC_ 4 
Positive:  Col11a1, Prc1, Hmmr, Nusap1, Anln, Ckap2l, Cdk1, Top2a, Mki67, Smc4 
       Kif23, Tpx2, Depdc1a, Tubb4b, Tuba1c, Pbk, Cenpf, Cks2, Kif11, Aurkb 
       Ube2c, Racgap1, Plk1, Ccna2, Spc25, Cdca3, Cdca8, Birc5, Knl1, Sgo2a 
Negative:  Tnnt1, Actc1, Ttn, Myog, Neb, Arpp21, Myod1, Myl1, Chrna1, Rbm24 
       Mylpf, Tnnc1, Atp2a1, Mymk, Rapsn, Cdh15, Klhl41, Mylk4, Tnnt2, Smyd1 
       Fitm1, Myh3, Tnni1, Kremen2, Mymx, Lrrn1, Mrln, Ank1, Fndc5, Apobec2 
PC_ 5 
Positive:  Sox4, Vcan, Mfap4, Gsta4, Bcl11a, Fos, Chd3, Sfrp2, Hmcn1, Dnm3os 
       Gas2, Marcksl1, Gas1, Foxp2, Robo2, Csrp2, Pdgfra, Crabp2, Nnat, Ebf1 
       Mex3b, Bex4, Creb5, Amot, Epha7, Sox11, Gm26771, Scx, Tmsb10, Mab21l2 
Negative:  Tubb4b, Snorc, Tubb2a, Hapln1, Matn1, Acan, Prc1, Col11a2, Hmmr, Matn3 
       Comp, Tuba1c, Ube2c, Nusap1, Tubb6, Cks2, Anln, Cnmd, Tpx2, Plk1 
       Cdc20, Lgals3, Racgap1, Ccna2, Cdk1, Ckap2l, Depdc1a, Col9a3, Pbk, Cdca8 </code></pre>
<pre class="r"><code>so &lt;- RunUMAP(so, dims = 1:20)</code></pre>
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;
This message will be shown once per session</code></pre>
<pre><code>16:30:24 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
<pre><code>16:30:24 Read 49947 rows and found 20 numeric columns</code></pre>
<pre><code>16:30:24 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>16:30:24 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>**************************************************|
16:30:27 Writing NN index file to temp file /var/folders/tp/b078yqdd4ydff9fx87lfttpj_sc0x3/T//RtmpE5kviT/file59c83e162b39
16:30:27 Searching Annoy index using 1 thread, search_k = 3000
16:30:38 Annoy recall = 100%
16:30:39 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
16:30:39 Initializing from normalized Laplacian + noise (using RSpectra)
16:30:40 Commencing optimization for 200 epochs, with 2105356 positive edges
16:30:58 Optimization finished</code></pre>
<pre class="r"><code>so &lt;- FindNeighbors(so, dims = 1:20)</code></pre>
<pre><code>Computing nearest neighbor graph
Computing SNN</code></pre>
<pre class="r"><code>so &lt;- FindClusters(so)</code></pre>
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 49947
Number of edges: 1587201

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8881
Number of communities: 20
Elapsed time: 9 seconds</code></pre>
<pre class="r"><code>ElbowPlot(so,ndims = 30)</code></pre>
<p><img src="figure/d_DigitsDish.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>UMAP views</p>
<pre class="r"><code>DimPlot(so)</code></pre>
<p><img src="figure/d_DigitsDish.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>DimPlot(so, group.by=&#39;Sample&#39;)</code></pre>
<p><img src="figure/d_DigitsDish.Rmd/unnamed-chunk-9-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>DimPlot(so, group.by=&#39;Day&#39;)</code></pre>
<p><img src="figure/d_DigitsDish.Rmd/unnamed-chunk-9-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>DimPlot(so, split.by=&#39;Sample&#39;, ncol=3)</code></pre>
<p><img src="figure/d_DigitsDish.Rmd/unnamed-chunk-9-4.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>saveRDS(so, file.path(seurat_objects_dir,&quot;Fuiten2023_DigitsInDish_00_load.RDS&quot;))</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.5

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Brisbane
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] lubridate_1.9.3    forcats_1.0.0      stringr_1.5.0      dplyr_1.1.3       
 [5] purrr_1.0.2        readr_2.1.4        tidyr_1.3.0        tibble_3.2.1      
 [9] ggplot2_3.4.4      tidyverse_2.0.0    Seurat_5.0.0       SeuratObject_5.0.0
[13] sp_2.1-1           workflowr_1.7.0   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.15.0      jsonlite_1.8.7        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   farver_2.1.1          
  [7] rmarkdown_2.23         fs_1.6.3               vctrs_0.6.3           
 [10] ROCR_1.0-11            spatstat.explore_3.2-5 htmltools_0.5.5       
 [13] sass_0.4.7             sctransform_0.4.1      parallelly_1.36.0     
 [16] KernSmooth_2.23-22     bslib_0.5.0            htmlwidgets_1.6.2     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.10.3         
 [22] zoo_1.8-12             cachem_1.0.8           whisker_0.4.1         
 [25] igraph_1.5.1           mime_0.12              lifecycle_1.0.3       
 [28] pkgconfig_2.0.3        Matrix_1.6-1.1         R6_2.5.1              
 [31] fastmap_1.1.1          fitdistrplus_1.1-11    future_1.33.0         
 [34] shiny_1.7.5.1          digest_0.6.33          colorspace_2.1-0      
 [37] patchwork_1.1.3        ps_1.7.5               rprojroot_2.0.3       
 [40] tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1         
 [43] labeling_0.4.3         progressr_0.14.0       timechange_0.2.0      
 [46] fansi_1.0.5            spatstat.sparse_3.0-3  httr_1.4.6            
 [49] polyclip_1.10-6        abind_1.4-5            compiler_4.3.1        
 [52] withr_2.5.1            fastDummies_1.7.3      highr_0.10            
 [55] MASS_7.3-60            tools_4.3.1            lmtest_0.9-40         
 [58] httpuv_1.6.11          future.apply_1.11.0    goftest_1.2-3         
 [61] glue_1.6.2             callr_3.7.3            nlme_3.1-162          
 [64] promises_1.2.0.1       grid_4.3.1             Rtsne_0.16            
 [67] getPass_0.2-2          cluster_2.1.4          reshape2_1.4.4        
 [70] generics_0.1.3         gtable_0.3.4           spatstat.data_3.0-3   
 [73] tzdb_0.4.0             hms_1.1.3              data.table_1.14.8     
 [76] utf8_1.2.4             spatstat.geom_3.2-7    RcppAnnoy_0.0.21      
 [79] ggrepel_0.9.4          RANN_2.6.1             pillar_1.9.0          
 [82] spam_2.10-0            RcppHNSW_0.5.0         later_1.3.1           
 [85] splines_4.3.1          lattice_0.21-8         renv_1.0.0            
 [88] survival_3.5-5         deldir_1.0-9           tidyselect_1.2.0      
 [91] miniUI_0.1.1.1         pbapply_1.7-2          knitr_1.43            
 [94] git2r_0.32.0           gridExtra_2.3          scattermore_1.2       
 [97] xfun_0.39              matrixStats_1.0.0      stringi_1.7.12        
[100] lazyeval_0.2.2         yaml_2.3.7             evaluate_0.21         
[103] codetools_0.2-19       cli_3.6.1              uwot_0.1.16           
[106] xtable_1.8-4           reticulate_1.34.0      munsell_0.5.0         
[109] processx_3.8.2         jquerylib_0.1.4        Rcpp_1.0.11           
[112] globals_0.16.2         spatstat.random_3.2-1  png_0.1-8             
[115] parallel_4.3.1         ellipsis_0.3.2         dotCall64_1.1-0       
[118] listenv_0.9.0          viridisLite_0.4.2      scales_1.2.1          
[121] ggridges_0.5.4         crayon_1.5.2           leiden_0.4.3          
[124] rlang_1.1.1            cowplot_1.1.1         </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
