<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Sarah Williams" />


<title>scRNAseq differential expression without replicates</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.33/datatables.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Spatial Sampler</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tests</a>
</li>
<li>
  <a href="index_data.html">Data</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/swbioinf/spatialsnippets">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">scRNAseq differential expression without
replicates</h1>
<h4 class="author">Sarah Williams</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-10-04
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>spatialsnippets/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20231017code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20231017)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20231017code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20231017)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstreec178f627cc9175364ba4c3f7787d32d93b3d3c00targetblankc178f62a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/swbioinf/spatialsnippets/tree/c178f627cc9175364ba4c3f7787d32d93b3d3c00" target="_blank">c178f62</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstreec178f627cc9175364ba4c3f7787d32d93b3d3c00targetblankc178f62a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/swbioinf/spatialsnippets/tree/c178f627cc9175364ba4c3f7787d32d93b3d3c00" target="_blank">c178f62</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/e_neighbourcellchanges.nb.html
    Ignored:    analysis/glossary.nb.html
    Ignored:    renv/library/
    Ignored:    renv/staging/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/e_DEWithoutReps.Rmd</code>) and
HTML (<code>docs/e_DEWithoutReps.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/c178f627cc9175364ba4c3f7787d32d93b3d3c00/analysis/e_DEWithoutReps.Rmd" target="_blank">c178f62</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-04
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/2a0a17b632d124fba5ea4b1fef3c771afb501bf9/docs/e_DEWithoutReps.html" target="_blank">2a0a17b</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/4edfd350647b5f6026a3dde61385bc403c446072/analysis/e_DEWithoutReps.Rmd" target="_blank">4edfd35</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-26
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/cd1b7adaf0bf112b7420314f5c4d44ea72689013/docs/e_DEWithoutReps.html" target="_blank">cd1b7ad</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/a42f6d30a349a3f156a9347e8996bd75b1e7f128/analysis/e_DEWithoutReps.Rmd" target="_blank">a42f6d3</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-26
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/2b1c8e2f1b31eb5a5b7a7c9ac3ea4328dbbec41d/docs/e_DEWithoutReps.html" target="_blank">2b1c8e2</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/511594fa32c4ab59aa41a582bc691aca9fde7e67/analysis/e_DEWithoutReps.Rmd" target="_blank">511594f</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-19
</td>
<td>
wflow_publish("analysis/")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/e7a4c1232709d2520c307c758f907eb9005812a0/docs/e_DEWithoutReps.html" target="_blank">e7a4c12</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-10-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/507ead4d3149a8703afa56cbf10bf74e5cefafa7/analysis/e_DEWithoutReps.Rmd" target="_blank">507ead4</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-10-18
</td>
<td>
wflow_publish("analysis/")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/584cf7388ac5d6a551b4245bd6116c4eec38f733/analysis/e_DEWithoutReps.Rmd" target="_blank">584cf73</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-10-17
</td>
<td>
adding
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/584cf7388ac5d6a551b4245bd6116c4eec38f733/docs/e_DEWithoutReps.html" target="_blank">584cf73</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-10-17
</td>
<td>
adding
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<!-- keywords
# Not yet used, but add keywords as appriopriate. 

insituspatial
seurat
differntialexpression
test

-->
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>Sometimes there are no biological replicates, yet you still want to
make a comparison. While not ideal, its possible. Individual cells may
be treated as ‘replicates’ to explore the difference between two
samples.</p>
<p><img src="assets/de_1v1.png" /></p>
<p>This requires:</p>
<ul>
<li>Cell clusters</li>
<li>2 or more samples (or grouping to compare)</li>
</ul>
<p>For example:</p>
<ul>
<li>What genes are differentially expressed between the knockout (n=1)
and control(n=1), for every cell type.</li>
<li>In my n=1 pilot study - what is differentially expressed between two
of my stromal clusters?</li>
</ul>
<p>Steps:</p>
<ol style="list-style-type: decimal">
<li>Filter to testable genes (enough expression to see changes)</li>
<li>Test for changes in gene expression</li>
<li>Plot DE results and individual genes.</li>
</ol>
<div id="gotchas" class="section level2">
<h2>Gotchas</h2>
<p><strong>Caveat on 1 vs 1 comparisons!</strong></p>
<p>Essentially the results we get will be testing the difference between
<em>this</em> specific sample and <em>that</em> specific sample. The
p-values for such a test can be very significant!</p>
<p>But those p-values can’t be compared to those using multiple samples,
and we cannot tell if our results will generalise to other samples. They
are likely good candidates for further work though!</p>
<p><strong>But what about pooled samples? </strong></p>
<p>Note that ‘pooling’ multiple unlabelled biological samples before
library prep will still count as ‘one’ replicate no many how many
samples are in the pool - because we have no way to tell what sample a
cell comes from. A change in gene expression could be from a single
outlier sample.</p>
<p>NB: Using cell hashing approaches to tag samples before pooling
avoids this issue.</p>
</div>
</div>
<div id="worked-example" class="section level1">
<h1>Worked example</h1>
<p>The data for this example is from paper <a
href="https://www.biorxiv.org/content/10.1101/2023.11.19.567482v1.full">Forming
nephrons promote nephron progenitor maintenance and branching
morphogenesis via paracrine BMP4 signalling under the control of
Wnt4</a> <span class="citation">(Moreau et al. 2023)</span></p>
<p>This study included 10X chromium single cell RNAseq data from 4
conditions, with 3-4 E14.5 mice pooled per group.</p>
<ul>
<li>Sample1 (Wnt4FloxKO): Wnt4Flox/Flox Six2-Conditional Wnt4
Knockout</li>
<li>Sample2 (Wnt4FloxHet): Wnt4Flox/+ Six2-Conditional Wnt4 Het</li>
<li>Sample3 (Wnt4Het): Wnt4 GCE/+ Control Wnt4 Het</li>
<li>Sample4 (Wnt4KO): Wnt4 GCE/GCE Knockout Wnt4</li>
</ul>
<p>In that paper they explain that complete or conditional homozygous
knockout of Wnt4 gene results in abnormal kidney development, and they
use scRNAseq data to explore effects at cellular level. <span
class="citation">(Moreau et al. 2023)</span></p>
<p>In this vignette, we will test for differential expression in each
cell type for two 1 vs 1 comparisons:</p>
<ul>
<li>Wnt4KO vs Wnt4Het : Complete knockout vs heterozgote</li>
<li>Wnt4FloxKO vs Wnt4FloxHet : Conditional (Flox) knockout vs
heterozygote</li>
</ul>
<div id="load-libraries-and-data" class="section level2">
<h2>Load Libraries and Data</h2>
<pre class="r"><code>library(Seurat)
library(edgeR)
library(limma)
library(DT)
library(tidyverse)


dataset_dir      &lt;- &#39;~/projects/spatialsnippets/datasets&#39;
seurat_file      &lt;- file.path(dataset_dir, &#39;Wnt4KO_Moreau2023&#39;,  &quot;Wnt4KOE14.5_11_ss.rds&quot;)

so &lt;- readRDS(seurat_file)</code></pre>
</div>
<div id="experimental-design" class="section level2">
<h2>Experimental Design</h2>
<p>In this case the sample (sample 1-4), ‘genotype’ GT effect and GT
short columns are just different labels for this same groups. We will
use GTshort as our ‘sample’ names throughout.</p>
<p>There are between 4000 and 10000 cells per sample.</p>
<pre class="r"><code>select(so@meta.data, sample, Genotype, GTeffect, GTshort) %&gt;% 
  as_tibble() %&gt;% 
  group_by( sample, Genotype, GTeffect, GTshort) %&gt;%
  summarise(num_cells=n(), .groups = &#39;drop&#39;) %&gt;%
  DT::datatable()</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-8839f08f82529a5247be" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8839f08f82529a5247be">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4"],["Sample1","Sample2","Sample3","Sample4"],["Six2TGC/+, Wnt4 FLOX/FLOX","Six2TGC/+, Wnt4 FLOX/+","Wnt4GCE HET","Wnt4GCE KO"],["Six2-Conditional Wnt4 Knockout","Six2-Conditional Wnt4 Het","Control Wnt4 Het","Knockout Wnt4"],["Wnt4FloxKO","Wnt4FloxHet","Wnt4Het","Wnt4KO"],[10511,6836,8340,4103]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>Genotype<\/th>\n      <th>GTeffect<\/th>\n      <th>GTshort<\/th>\n      <th>num_cells<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":5},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"Genotype","targets":2},{"name":"GTeffect","targets":3},{"name":"GTshort","targets":4},{"name":"num_cells","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="filter" class="section level2">
<h2>Filter</h2>
<div id="counts-per-cell" class="section level3">
<h3>Counts per cell</h3>
<p>A minimum-counts-per-cell threshold was already applied to this
dataset (during preprocessing) so nothing to do here.</p>
<pre class="r"><code># This is plenty
min(so$nCount_RNA)</code></pre>
<pre><code>[1] 1456</code></pre>
</div>
<div id="cells-per-group" class="section level3">
<h3>Cells per group</h3>
<p>How many cells are there for each celltype for each sample? If there
are too few on either side of the contrast, we won’t be able to
test.</p>
<p>How many is too few? 50 might be a good threshold. In this case
however, we might consider being very permissive and allowing just 20
cells to look at some celltypes that are clearly reduced or increased in
number between conditions (e.g. c9,c10,c11).</p>
<p>They’re interesting in this experiment, but have to keep in mind
during interpretation that results may be less reliable.</p>
<p>We will apply that filter later when running the differential
expression.</p>
<pre class="r"><code>min_cells_per_group &lt;- 20 # used later
table(so$CelltypeCode, so$GTshort)</code></pre>
<pre><code>                                    
                                     Wnt4FloxHet Wnt4FloxKO Wnt4Het Wnt4KO
  c0: Blood                                 1531        567    2597    189
  c1: Stroma                                 553       2014     750    130
  c2: High Mitochondria - Non Wnt4KO         868        678    1353     69
  c3: NP                                     587       1225     684    376
  c4: Stroma                                 403       1088     435    283
  c5: Stroma                                 283        587     339    511
  c6: Stroma                                 343        841     391     32
  c7: Stroma                                 326        672     242    330
  c8: Immune                                 408        701     228    144
  c9: Stroma - Wnt4KO                         33         41      23   1383
  c10: UE - Cortical                         325        408     345     35
  c11: UE - Tip                              251        466     307     47
  c12: Stroma                                189        658     177     28
  c13: Vasculature                           295        245     227     22
  c14: High Mitochondria                     125         16     140    381
  c15: Podocyte                               61         65      49    110
  c16: Blood                                 171         28       8      9
  c17: Stroma                                 21         94      13     17
  c18: Immune                                 29         59      24      6
  c19: UE - Medullary                         34         58       8      1</code></pre>
</div>
</div>
<div id="run-just-one-differential-expression" class="section level2">
<h2>Run just one differential expression</h2>
<p>Talk about options. …</p>
<ul>
<li>layer vs assay</li>
<li>SCT vs RNA</li>
<li>count vs data vs data scale</li>
<li>subsampling</li>
<li>min.pct</li>
</ul>
<pre class="r"><code>de_result &lt;- FindMarkers(so.this, 
                               ident.1  = &#39;Wnt4KO&#39;, 
                               ident.2  = &#39;Wnt4Het&#39;,
                               slot     = &#39;data&#39;,
                               test.use = &quot;wilcox&quot;, # the default
                               min.pct  = 0.01,   # Note 
                               max.cells.per.ident = 100 # 1000 # if you have really big clusters, set this to subsample!
                               )</code></pre>
</div>
<div id="run-differential-expression-for-every-cluster"
class="section level2">
<h2>Run differential expression for every cluster</h2>
<p>We typically want to know about changes in every cluster. We can
simply loop through and run them one-by-one. This can make for a lot of
contrasts!</p>
<pre><code>2x comparisons * 19x clusters = 38 sets of differential results </code></pre>
<p>The below code makes a big table of all the tests - recording which
contrast was being done in which cluster.</p>
<p>NB: Note that the <em>FindMarkers()</em> function returns a table
with feature names as a rowname. Rownames are expected to be unique. So
while that is fine for one comparison, but when we want to put more
together, we need to give it its own column. If we don’t, there’ll be
gene names with ‘.1’ appended to the end!</p>
<pre class="r"><code># Set threhoehsolds
min_cells_per_group &lt;- 20

# Calculate DE across every celltype
# Empty list to collect results
de_result_list     &lt;- list()
de_result_sig_list &lt;- list()

# the contrasts
# If we store them in a list we can give the nice names
contrast_list &lt;-  list( # &#39;Test&#39;      vs  &#39;control&#39;  
                       FloxWnt4KOvsFloxHet  = c(&#39;Wnt4FloxKO&#39;,      &#39;Wnt4FloxHet&#39; ) ,
                       TotalWnt4KOvsHet     = c(&#39;Wnt4KO&#39;,          &#39;Wnt4Het&#39;     ) )

## Or you could autogenerated them
#make_contrast_name &lt;- function(contrast_parts){ paste0(contrast_parts[1],&quot;vs&quot;,contrast_parts[1])}
#contrast_names &lt;- sapply(FUN=make_contrast_name, X=contrast_list) 
#names(contrast_list) &lt;- contrast_names


# note keeping all 4 samples before two contrasts. (actually does that matter for wilcox ranK sum - CHECK)

#the_celltype = &#39;c15: Podocyte&#39;
#the_celltype = &quot;c16: Blood&quot;
#contrast_name &lt;- &#39;TotalWnt4KOvsHet&#39;
  
for (the_celltype in levels(so$CelltypeCode)[5:7]) {
  
  # Subset to one cell type. 
  print(the_celltype)
  
  so.this &lt;- subset(so, CelltypeCode == the_celltype)
  
  # count how many cells within each sample (GTshort)
  # And list which samples have more than the minimum
  cells_per_sample &lt;- table(so.this$GTshort)
  print(cells_per_sample)
  samples_with_enough_cells &lt;- names(cells_per_sample)[cells_per_sample &gt; min_cells_per_group]
  
  # SUbset to only those samples (might be all cells, might be no cells)
  so.this &lt;- subset(so.this, GTshort %in% samples_with_enough_cells)
  

  # For each listed contrast, do both sides of the contrast have enough cells?
  # This is of course much simpler if you only have one contrast!
  for (contrast_name in names(contrast_list)) {
    
    # from mycontrastname pull out list of the two samples involved;
    # c(&#39;test&#39;, &#39;control&#39;)
    contrast &lt;- contrast_list[[contrast_name]]
    
    # Only run this contrast if both sides pass!
    if (all(contrast %in% samples_with_enough_cells)) {
      print(contrast_name) 

      # We need to tell Seurat to group by _sample_ not cluster.
      Idents(so.this) &lt;- so.this$GTshort
      
      
      de_result &lt;- FindMarkers(so.this, 
                               ident.1  = contrast[1], 
                               ident.2 = contrast[2],
                               slot     = &#39;data&#39;,
                               test.use = &quot;wilcox&quot;, # the default
                               min.pct  = 0.01,   # Note 
                               max.cells.per.ident = 100 # 1000 # if you have really big clusters, set this to subsample!
                               )
  
      
      # It can be helpful to know the average expression of a gene
      # This will give us the average (per cell) within this celltype.
      avg_expression &lt;- rowMeans(GetAssayData(so.this, assay = &#39;SCT&#39;, layer=&quot;data&quot;))
      
      
      
      de_result.formatted &lt;- de_result %&gt;%
      rownames_to_column(&quot;target&quot;) %&gt;%
      mutate(contrast=contrast_name,
             celltype=the_celltype,
             avg_expression=avg_expression[target]) %&gt;%
      select(celltype,contrast,target,avg_expression, everything()) %&gt;%
      arrange(p_val)
  
      # Filter to just significant results, optionally by log2FC.
      de_result.sig &lt;- filter(de_result.formatted, 
                         p_val_adj &lt; 0.01,
                         abs(avg_log2FC) &gt; log2(1.5) )

      # Record these results in a list to combine
      full_name &lt;- paste(contrast_name, the_celltype)
      de_result_list[[full_name]] &lt;- de_result.formatted
      de_result_sig_list[[full_name]] &lt;- de_result.sig
      
    }
  }
}

# Join together results for all celltypes, and pull out those with a singificant adjusted p-value
de_results_all &lt;- bind_rows(de_result_list)
de_results_sig &lt;- bind_rows(de_result_sig_list)</code></pre>
<p>Check out the full set of significant DE genes</p>
<pre class="r"><code>DT::datatable(de_results_sig)</code></pre>
<p>Save the results.</p>
<pre class="r"><code># Save the full set of results as a tab-deliminated text file.
# This is useful for parsing later e.g. for functional enrichment
write_tsv(x = de_results_all, &quot;~/myproject/de_results_all.tsv&quot;)

# If you don&#39;t have too many contrasts and celltypes, 
# you can save it as an excel file, with each contrast in a seprate tab.
library(writexl)
write_xlsx(de_result_list,      &quot;~/myproject/de_results_all.xlsx&quot;)
write_xlsx(de_result_sig_list,  &quot;~/myproject/de_results_sig.xlsx&quot;)</code></pre>
</div>
<div id="check-individual-results" class="section level2">
<h2>Check individual results</h2>
<p>Plot some example genes.</p>
<p>Discuss (how to pick stuff thats real)</p>
<ul>
<li>further filtering - where small significant changes dominate</li>
<li>violin plots are hard to read</li>
<li>MA style plots.</li>
</ul>
</div>
</div>
<div id="code-snippet" class="section level1">
<h1>Code Snippet</h1>
<p>Copy and paste this bit into your own script to adapt to your own
data. Make as generic as possible, just a framework.</p>
<pre class="r"><code># This code does not run</code></pre>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<pre class="r"><code>DT::datatable(de_result.sig)</code></pre>
<ul>
<li><strong>celltype</strong>: The cluster or celltype the contrat is
within. Added by running code.</li>
<li><strong>contrast</strong>: The name of what is being compared. Added
by running code.</li>
<li><strong>target</strong> : Gene name. (NB: Added by our running code,
in the direct output of FindMarkers(), these are row names instead)</li>
<li><strong>avg_expression</strong> : Average expression of target
within celltype group. (Added by running code)</li>
<li><strong>p_val</strong> : P value without multiple hypothesis
correction - see <em>p_val_adj</em></li>
<li><strong>avg_log2FC</strong> : Average log 2 fold change of the
comparison. Is calculated as: <em>log2(test expression) - log2(control
expression)</em> A value of 0 represents no change, +1 is a doubling,
and -1 is a halving of expression.</li>
<li><strong>pct.1</strong> : Percent of cells in group 1 (generally
test) that express the gene at all.</li>
<li><strong>pct.2</strong> : Percent of cells in group 2 (generally
control/WT/reference) that express the gene at all.</li>
<li><strong>p_val_adj</strong> : Pvalue with multiple hypothesis
correction. This may be used for filtering.</li>
</ul>
</div>
<div id="more-information" class="section level1">
<h1>More information</h1>
<p>List of useful resources. Papers, vignettes, pertinent forum
posts</p>
<ul>
<li><p><a href="d_Wnt4KO.html">Wnt4 KO in developing mouse kidney - 10X
Chromium scRNAseq</a>: <a
href="https://www.biorxiv.org/content/10.1101/2023.11.19.567482v1.full">Forming
nephrons promote nephron progenitor maintenance and branching
morphogenesis via paracrine BMP4 signalling under the control of
Wnt4</a> <span class="citation">(Moreau et al. 2023)</span> : The paper
with this data.</p></li>
<li><p><a
href="https://satijalab.org/seurat/articles/de_vignette">Seurat
Differential Expression Vignette</a> : How to do differential expression
with the seurat package. Tells you how to use the multiple statistical
tests that seurat offers.</p></li>
<li><p><a
href="https://bioconductor.org/books/3.19/OSCA.multisample/multi-sample-comparisons.html">OSCA
Differntial Expression</a>: The excellent book <a
href="https://bioconductor.org/books/release/OSCA/book-contents.html#multi-sample">Orchestrating
single cell analysis</a> includes a section on differential expression,
it focusses on pseuboulk approaches with replicates (which can’t be used
for a 1 vs 1) using the bioconductor toolkit, but provides useful
background.</p></li>
</ul>
</div>
<div id="refereneces" class="section level1">
<h1>Refereneces</h1>
<!-- nothing here, autopopulated from bibliography: in json -->
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-moreauFormingNephronsPromote2023a" class="csl-entry">
Moreau, Julie L. M., Sarah Williams, Jihane Homman-Ludiye, Andrew J.
Mallett, and Alexander N. Combes. 2023. <span>“Forming Nephrons Promote
Nephron Progenitor Maintenance and Branching Morphogenesis via Paracrine
<span>BMP4</span> Signalling Under the Control of
<span><em>Wnt4</em></span>.”</span> November 19, 2023. <a
href="https://doi.org/10.1101/2023.11.19.567482">https://doi.org/10.1101/2023.11.19.567482</a>.
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] lubridate_1.9.3    forcats_1.0.0      stringr_1.5.1      dplyr_1.1.4       
 [5] purrr_1.0.2        readr_2.1.5        tidyr_1.3.1        tibble_3.2.1      
 [9] ggplot2_3.5.0      tidyverse_2.0.0    DT_0.33            edgeR_4.0.16      
[13] limma_3.58.1       Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-3          
[17] workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.16.0      jsonlite_1.8.8        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   rmarkdown_2.26        
  [7] fs_1.6.3               vctrs_0.6.5            ROCR_1.0-11           
 [10] spatstat.explore_3.2-7 htmltools_0.5.8.1      sass_0.4.9            
 [13] sctransform_0.4.1      parallelly_1.37.1      KernSmooth_2.23-22    
 [16] bslib_0.7.0            htmlwidgets_1.6.4      ica_1.0-3             
 [19] plyr_1.8.9             plotly_4.10.4          zoo_1.8-12            
 [22] cachem_1.0.8           whisker_0.4.1          igraph_2.0.3          
 [25] mime_0.12              lifecycle_1.0.4        pkgconfig_2.0.3       
 [28] Matrix_1.6-5           R6_2.5.1               fastmap_1.1.1         
 [31] fitdistrplus_1.1-11    future_1.33.2          shiny_1.8.1.1         
 [34] digest_0.6.35          colorspace_2.1-0       patchwork_1.2.0       
 [37] ps_1.7.6               rprojroot_2.0.4        tensor_1.5            
 [40] RSpectra_0.16-1        irlba_2.3.5.1          crosstalk_1.2.1       
 [43] progressr_0.14.0       timechange_0.3.0       fansi_1.0.6           
 [46] spatstat.sparse_3.0-3  httr_1.4.7             polyclip_1.10-6       
 [49] abind_1.4-5            compiler_4.3.2         withr_3.0.0           
 [52] fastDummies_1.7.3      MASS_7.3-60.0.1        tools_4.3.2           
 [55] lmtest_0.9-40          httpuv_1.6.15          future.apply_1.11.2   
 [58] goftest_1.2-3          glue_1.7.0             callr_3.7.6           
 [61] nlme_3.1-164           promises_1.2.1         grid_4.3.2            
 [64] Rtsne_0.17             getPass_0.2-4          cluster_2.1.6         
 [67] reshape2_1.4.4         generics_0.1.3         gtable_0.3.4          
 [70] spatstat.data_3.0-4    tzdb_0.4.0             hms_1.1.3             
 [73] data.table_1.15.4      utf8_1.2.4             spatstat.geom_3.2-9   
 [76] RcppAnnoy_0.0.22       ggrepel_0.9.5          RANN_2.6.1            
 [79] pillar_1.9.0           spam_2.10-0            RcppHNSW_0.6.0        
 [82] later_1.3.2            splines_4.3.2          lattice_0.22-6        
 [85] renv_1.0.5             survival_3.5-8         deldir_2.0-4          
 [88] tidyselect_1.2.1       locfit_1.5-9.9         miniUI_0.1.1.1        
 [91] pbapply_1.7-2          knitr_1.45             git2r_0.33.0          
 [94] gridExtra_2.3          scattermore_1.2        xfun_0.43             
 [97] statmod_1.5.0          matrixStats_1.2.0      stringi_1.8.3         
[100] lazyeval_0.2.2         yaml_2.3.8             evaluate_0.23         
[103] codetools_0.2-20       BiocManager_1.30.22    cli_3.6.2             
[106] uwot_0.1.16            xtable_1.8-4           reticulate_1.35.0     
[109] munsell_0.5.1          processx_3.8.4         jquerylib_0.1.4       
[112] Rcpp_1.0.12            globals_0.16.3         spatstat.random_3.2-3 
[115] png_0.1-8              parallel_4.3.2         dotCall64_1.1-1       
[118] listenv_0.9.1          viridisLite_0.4.2      scales_1.3.0          
[121] ggridges_0.5.6         leiden_0.4.3.1         rlang_1.1.3           
[124] cowplot_1.1.3         </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
