<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Sarah Williams" />


<title>scRNAseq differential expression without replicates</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.33/datatables.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Spatial Sampler</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tests</a>
</li>
<li>
  <a href="index_data.html">Data</a>
</li>
<li>
  <a href="contributing.html">Contributing</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/swbioinf/spatialsnippets">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">scRNAseq differential expression without
replicates</h1>
<h4 class="author">Sarah Williams</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-10-09
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>spatialsnippets/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20231017code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20231017)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20231017code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20231017)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree31da75c06ee0f907841393e188e9294117443eb7targetblank31da75ca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/swbioinf/spatialsnippets/tree/31da75c06ee0f907841393e188e9294117443eb7" target="_blank">31da75c</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree31da75c06ee0f907841393e188e9294117443eb7targetblank31da75ca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/swbioinf/spatialsnippets/tree/31da75c06ee0f907841393e188e9294117443eb7" target="_blank">31da75c</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/contributions.nb.html
    Ignored:    analysis/e_neighbourcellchanges.nb.html
    Ignored:    analysis/glossary.nb.html
    Ignored:    renv/library/
    Ignored:    renv/staging/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/e_DEWithoutReps.Rmd</code>) and
HTML (<code>docs/e_DEWithoutReps.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/31da75c06ee0f907841393e188e9294117443eb7/analysis/e_DEWithoutReps.Rmd" target="_blank">31da75c</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
<td>
wflow_publish("analysis/")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/b681e0dc966ab9d7609526177446dc63de7ab354/docs/e_DEWithoutReps.html" target="_blank">b681e0d</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/7c61fe814d26b9ba1212bf979e85e13744c46090/analysis/e_DEWithoutReps.Rmd" target="_blank">7c61fe8</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
<td>
wflow_publish("analysis/")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/e_DEWithoutReps.html" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/ca62223f695fa56ddcf8f1a5af89d767c8a678a3/analysis/e_DEWithoutReps.Rmd" target="_blank">ca62223</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/87886e17e98fe7f1db4bfadbcb54ab241f3e60cb/docs/e_DEWithoutReps.html" target="_blank">87886e1</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/501fbfd4e11cc1a6ee7cb33e28cf8b09e615328a/analysis/e_DEWithoutReps.Rmd" target="_blank">501fbfd</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-08
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/41a750ed56732d7386e797fbb9556d78f29767db/docs/e_DEWithoutReps.html" target="_blank">41a750e</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/eda7dacf83becc6272102be1d8550d02b2c2d2e1/analysis/e_DEWithoutReps.Rmd" target="_blank">eda7dac</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-04
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/98d7c091334e65919c1e14678cd42780fcd3e236/docs/e_DEWithoutReps.html" target="_blank">98d7c09</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/c178f627cc9175364ba4c3f7787d32d93b3d3c00/analysis/e_DEWithoutReps.Rmd" target="_blank">c178f62</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-04
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/2a0a17b632d124fba5ea4b1fef3c771afb501bf9/docs/e_DEWithoutReps.html" target="_blank">2a0a17b</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/4edfd350647b5f6026a3dde61385bc403c446072/analysis/e_DEWithoutReps.Rmd" target="_blank">4edfd35</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-26
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/cd1b7adaf0bf112b7420314f5c4d44ea72689013/docs/e_DEWithoutReps.html" target="_blank">cd1b7ad</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/a42f6d30a349a3f156a9347e8996bd75b1e7f128/analysis/e_DEWithoutReps.Rmd" target="_blank">a42f6d3</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-26
</td>
<td>
wflow_publish("analysis/e_DEWithoutReps.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/2b1c8e2f1b31eb5a5b7a7c9ac3ea4328dbbec41d/docs/e_DEWithoutReps.html" target="_blank">2b1c8e2</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/511594fa32c4ab59aa41a582bc691aca9fde7e67/analysis/e_DEWithoutReps.Rmd" target="_blank">511594f</a>
</td>
<td>
swbioinf
</td>
<td>
2024-09-19
</td>
<td>
wflow_publish("analysis/")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/e7a4c1232709d2520c307c758f907eb9005812a0/docs/e_DEWithoutReps.html" target="_blank">e7a4c12</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-10-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/507ead4d3149a8703afa56cbf10bf74e5cefafa7/analysis/e_DEWithoutReps.Rmd" target="_blank">507ead4</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-10-18
</td>
<td>
wflow_publish("analysis/")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/584cf7388ac5d6a551b4245bd6116c4eec38f733/analysis/e_DEWithoutReps.Rmd" target="_blank">584cf73</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-10-17
</td>
<td>
adding
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/584cf7388ac5d6a551b4245bd6116c4eec38f733/docs/e_DEWithoutReps.html" target="_blank">584cf73</a>
</td>
<td>
Sarah Williams
</td>
<td>
2023-10-17
</td>
<td>
adding
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<!-- keywords
# Not yet used, but add keywords as appriopriate. 

insituspatial
seurat
differntialexpression
test

-->
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>Sometimes there are no biological replicates, yet you still want to
compare gene expression between groups. While not ideal, its possible.
Individual cells may be treated as ‘replicates’ to explore the
difference between two samples.</p>
<p>Commonly, this can be done for each cluster or celltype in your
experiment. But in a single-sample pilot you could compare clusters to
each other</p>
<p><img src="assets/de_1v1.png" /></p>
<p>This requires:</p>
<ul>
<li>Cell clusters / cell types</li>
<li>2 or more samples (or grouping to compare)</li>
</ul>
<p>For example:</p>
<ul>
<li>What genes are differentially expressed between the knockout (n=1)
and control(n=1), for every cell type.</li>
<li>In my n=1 pilot study - what is differentially expressed between two
of my stromal clusters?</li>
</ul>
<!--  Not really helpful
Steps:

1. Test for changes in gene expression
2. Plot DE results 
3. Check individual genes of interest
-->
<hr />
<p>This vignette describes a default approach using Seurat built-in
methods. It is by no means the only way.</p>
<p>It’s highly recommended to check out the paper <a
href="https://www.nature.com/articles/s41467-021-25960-2">Confronting
false discoveries in single-cell differential expression</a><span
class="citation">(Squair et al. 2021)</span>, which outlines the
importance of replicates, benchmarking of different scRNAseq
differenetial expression tools and common biases in results.</p>
<p><strong>Not suitable for replicates</strong></p>
<p>The approach described in this document is really only useful when
there are no replicates - using your replicates will proved much more
biologically relevant results! If you were to use this approach you will
loose the individual information!</p>
<p><strong>Caveat on 1 vs 1 comparisons!</strong></p>
<p>Essentially the results we get will be testing the difference between
<em>this</em> specific sample and <em>that</em> specific sample. The
p-values for such a test can be very significant!</p>
<p>But those p-values can’t be compared to those using multiple samples,
and we cannot tell if our results will generalise to other samples. They
are likely good candidates for further work though!</p>
<p><strong>But what about pooled samples? </strong></p>
<p>Note that ‘pooling’ multiple unlabelled biological samples before
library prep will still count as ‘one’ replicate no many how many
samples are in the pool - because we have no way to tell what sample a
cell comes from. A change in gene expression could be from a single
outlier sample.</p>
<p>NB: Using cell hashing approaches to tag samples before pooling
avoids this issue.</p>
</div>
<div id="worked-example" class="section level1">
<h1>Worked example</h1>
<p>The data for this example is from paper <a
href="https://www.biorxiv.org/content/10.1101/2023.11.19.567482v1.full">Forming
nephrons promote nephron progenitor maintenance and branching
morphogenesis via paracrine BMP4 signalling under the control of
Wnt4</a> <span class="citation">(Moreau et al. 2023)</span></p>
<p>This study included 10X chromium single cell RNAseq data from 4
conditions, with 3-4 E14.5 mice pooled per group.</p>
<ul>
<li>Sample1 (Wnt4FloxKO): Wnt4Flox/Flox Six2-Conditional Wnt4
Knockout</li>
<li>Sample2 (Wnt4FloxHet): Wnt4Flox/+ Six2-Conditional Wnt4 Het</li>
<li>Sample3 (Wnt4Het): Wnt4 GCE/+ Control Wnt4 Het</li>
<li>Sample4 (Wnt4KO): Wnt4 GCE/GCE Knockout Wnt4</li>
</ul>
<p>In that paper they explain that complete or conditional homozygous
knockout of Wnt4 gene results in abnormal kidney development, and they
use scRNAseq data to explore effects at cellular level. <span
class="citation">(<strong>moreauFormingNephronsPromote2023a?</strong>)</span></p>
<p>In this vignette, we will test for differential expression in each
cell type for two 1 vs 1 comparisons:</p>
<ul>
<li>Wnt4KO vs Wnt4Het : Complete knockout vs heterozgote</li>
<li>Wnt4FloxKO vs Wnt4FloxHet : Conditional (Flox) knockout vs
heterozygote</li>
</ul>
<div id="load-libraries-and-data" class="section level2">
<h2>Load Libraries and Data</h2>
<pre class="r"><code>library(Seurat)
library(DT)
library(tidyverse)


dataset_dir      &lt;- &#39;~/projects/spatialsnippets/datasets&#39;
seurat_file      &lt;- file.path(dataset_dir, &#39;Wnt4KO_Moreau2023&#39;,  &quot;Wnt4KOE14.5_11_ss.rds&quot;)

so &lt;- readRDS(seurat_file)</code></pre>
</div>
<div id="experimental-design" class="section level2">
<h2>Experimental Design</h2>
<p>In this case the sample (sample 1-4), ‘genotype’ GT effect and GT
short columns are just different labels for this same groups. We will
use GTshort as our ‘sample’ names throughout.</p>
<p>There are between 4000 and 10000 cells per sample.</p>
<pre class="r"><code>select(so@meta.data, sample, Genotype, GTeffect, GTshort) %&gt;% 
  as_tibble() %&gt;% 
  group_by( sample, Genotype, GTeffect, GTshort) %&gt;%
  summarise(num_cells=n(), .groups = &#39;drop&#39;) %&gt;%
  DT::datatable()</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-8839f08f82529a5247be" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8839f08f82529a5247be">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4"],["Sample1","Sample2","Sample3","Sample4"],["Six2TGC/+, Wnt4 FLOX/FLOX","Six2TGC/+, Wnt4 FLOX/+","Wnt4GCE HET","Wnt4GCE KO"],["Six2-Conditional Wnt4 Knockout","Six2-Conditional Wnt4 Het","Control Wnt4 Het","Knockout Wnt4"],["Wnt4FloxKO","Wnt4FloxHet","Wnt4Het","Wnt4KO"],[10511,6836,8340,4103]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>Genotype<\/th>\n      <th>GTeffect<\/th>\n      <th>GTshort<\/th>\n      <th>num_cells<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":5},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"Genotype","targets":2},{"name":"GTeffect","targets":3},{"name":"GTshort","targets":4},{"name":"num_cells","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="filter" class="section level2">
<h2>Filter</h2>
<div id="counts-per-cell" class="section level3">
<h3>Counts per cell</h3>
<p>A minimum-counts-per-cell threshold was already applied to this
dataset (during preprocessing) so nothing to do here.</p>
<pre class="r"><code># This is plenty
min(so$nCount_RNA)</code></pre>
<pre><code>[1] 1456</code></pre>
</div>
<div id="cells-per-group" class="section level3">
<h3>Cells per group</h3>
<p>How many cells are there for each celltype for each sample? If there
are too few on either side of the contrast, we won’t be able to
test.</p>
<p>How many is too few? 50 might be a good threshold. In this case
however, we might consider being very permissive and allowing just 20
cells to look at some celltypes that are clearly reduced or increased in
number between conditions (e.g. c9,c10,c11).</p>
<p>They’re interesting in this experiment, but we have to keep in mind
during interpretation that results may be less reliable.</p>
<p>We will apply that filter later (‘min_cells_per_group’) when running
the differential expression.</p>
<pre class="r"><code>table(so$CelltypeCode, so$GTshort)</code></pre>
<pre><code>                                    
                                     Wnt4FloxHet Wnt4FloxKO Wnt4Het Wnt4KO
  c0: Blood                                 1531        567    2597    189
  c1: Stroma                                 553       2014     750    130
  c2: High Mitochondria - Non Wnt4KO         868        678    1353     69
  c3: NP                                     587       1225     684    376
  c4: Stroma                                 403       1088     435    283
  c5: Stroma                                 283        587     339    511
  c6: Stroma                                 343        841     391     32
  c7: Stroma                                 326        672     242    330
  c8: Immune                                 408        701     228    144
  c9: Stroma - Wnt4KO                         33         41      23   1383
  c10: UE - Cortical                         325        408     345     35
  c11: UE - Tip                              251        466     307     47
  c12: Stroma                                189        658     177     28
  c13: Vasculature                           295        245     227     22
  c14: High Mitochondria                     125         16     140    381
  c15: Podocyte                               61         65      49    110
  c16: Blood                                 171         28       8      9
  c17: Stroma                                 21         94      13     17
  c18: Immune                                 29         59      24      6
  c19: UE - Medullary                         34         58       8      1</code></pre>
</div>
</div>
<div id="what-data-is-used-to-calculate-differential-expression"
class="section level2">
<h2>What data is used to calculate differential expression?</h2>
<p>In short - we want to run this differential expression on the
normalised data (which is on a log scale), and typically that’s in the
<em>data</em> layer of the <em>RNA</em> assay.</p>
<hr />
<p>The longer explanation:</p>
<p>From Seurat v5, a Seruat object has mutiple <em>assays</em> (but
usually just one, called RNA), each of which has multiple
<em>layers</em>.</p>
<p>Each <em>assay</em> generally has some or all of the following
layers:</p>
<ul>
<li><strong>counts</strong> : Just the plain old counts matrix.</li>
<li><strong>data</strong> : A normalised counts matrix</li>
<li><strong>scale.data</strong> : A normalised and scaled counts matrix,
which often only contains a subset of the highly variable genes. (Used
for PCA e.t.c)</li>
</ul>
<p>Depending on how you setup your object, you might instead have one
set of counts/data/scale.data layers per sample. (e.g. counts.sample1,
counts.sample2 …). In that case you’ll want to proceed with differential
expression on a merged layer - which you can produce with the
JoinLayers() function. See how that’s used in <a
href="https://satijalab.org/seurat/articles/seurat5_integration#layers-in-the-seurat-v5-object">seurats
v5 integration tutorial</a></p>
<p>Multiple assays come into play more often with multimodal data - a
different assay per modality (RNA, ADT, e.t.c). See <a
href="https://satijalab.org/seurat/articles/multimodal_vignette">multimodal
vigette</a> for details.</p>
<p>Beware: The notation around Layers and Assays has changed between
Seurat v4 and v5! Layers were called ‘slots’ before v5, and there was no
concept of joining and splitting layers with different samples.</p>
<hr />
<p>This dataset however is a little different.It was analysed with an
earlier version of Seurat, and has ended up with two assays; an RNA and
an SCT assay. NB: The newer <a
href="https://satijalab.org/seurat/articles/seurat5_integration#layers-in-the-seurat-v5-object">seurat
5 integration workflow</a> behaves differently now, and doesn’t make
that assay like that (NB: to be confirmed).</p>
<pre class="r"><code>Assays(so)</code></pre>
<pre><code>[1] &quot;RNA&quot; &quot;SCT&quot;</code></pre>
<pre class="r"><code>DefaultAssay(so)</code></pre>
<pre><code>[1] &quot;SCT&quot;</code></pre>
<p>Alternatively the seurat summary tells us the assasy, and layers in
the default assay</p>
<pre class="r"><code>so</code></pre>
<pre><code>An object of class Seurat 
50377 features across 29790 samples within 2 assays 
Active assay: SCT (19360 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: umap, pca</code></pre>
<p>In any case, we want the unintegrated ‘RNA’ assay for calculating
differential expression, not our integrated SCT assay (This is the
reccomendation of the Seurat developers, see <a
href="https://github.com/satijalab/seurat/discussions/4032">discussion</a>
). The between sample integration was useful for visualising the
cell-cell relationships together (e.g. superimposing samples on a UMAP
layout), but for differential expression, we need to look at the
variation between samples.</p>
<pre class="r"><code>DefaultAssay(so) &lt;- &#39;RNA&#39;
so</code></pre>
<pre><code>An object of class Seurat 
50377 features across 29790 samples within 2 assays 
Active assay: RNA (31017 features, 0 variable features)
 2 layers present: counts, data
 1 other assay present: SCT
 2 dimensional reductions calculated: umap, pca</code></pre>
<p>We don’t have the normalised RNA data assay, so create it now.</p>
<pre class="r"><code>so &lt;- NormalizeData(so)</code></pre>
<!-- ```{r} -->
<!-- GetAssayData(so, assay = "RNA", layer = "counts")[1:10,1:3] -->
<!-- GetAssayData(so, assay = "RNA", layer = "data")[1:10,1:3] -->
<!-- GetAssayData(so, assay = "SCT", layer = "data")[1:10,1:3] -->
<!-- GetAssayData(so, assay = "SCT", layer = "scale.data")[1:10,1:3] -->
<!-- ``` -->
</div>
<div id="run-a-single-differential-expression-test"
class="section level2">
<h2>Run a single differential expression test</h2>
<p>Lets say we want to test the difference between samples Wnt4KO and
Wnt4Het within Nephron progenitors (NP),</p>
<p>We can use Seurat’s <em>FindMarkers</em> function to find
differential expression between two groups. Its the same method used by
the convenience function <em>FindAllMarkers</em> used for finding
celltype markers.</p>
<p>Bt we first need to configure our Seuruat object for use with this
function. Though we can specify our layer with the slot parameter,
there’s no parameter for assay in FindMarkers, you need to make sure
your ‘Default’ assay is the one you want to use.</p>
<pre class="r"><code>DefaultAssay(so)</code></pre>
<pre><code>[1] &quot;RNA&quot;</code></pre>
<pre class="r"><code>DefaultAssay(so) &lt;- &quot;RNA&quot; # Set it to RNA</code></pre>
<p>Likewise, there’s no parameter to specify which column of the
metadata out groups may be found in. We instead set Idents()</p>
<pre class="r"><code>Idents(so) &lt;- so$GTshort
levels(Idents(so))</code></pre>
<pre><code>[1] &quot;Wnt4FloxHet&quot; &quot;Wnt4FloxKO&quot;  &quot;Wnt4Het&quot;     &quot;Wnt4KO&quot;     </code></pre>
<p>Now make a subset of our Seurat object that’s just the NP cells.</p>
<pre class="r"><code>so.np &lt;- subset(so, CelltypeCode == &quot;c3: NP&quot;)
table(so$GTshort)</code></pre>
<pre><code>
Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
       6836       10511        8340        4103 </code></pre>
<p>Finally - run FindMarkers. We will stick with the default method that
uses a wilcoxon rank-sum test (but to explore other tests see the <a
href="https://satijalab.org/seurat/articles/de_vignette">seurat DE
vignette</a>.</p>
<pre class="r"><code>de_result &lt;- FindMarkers(so.np, 
                         ident.1  = &#39;Wnt4KO&#39;, 
                         ident.2  = &#39;Wnt4Het&#39;,
                         slot     = &#39;data&#39;,   # This is the default
                         test.use = &quot;wilcox&quot;  # This is the default

)</code></pre>
<p>Check out those p-values! They’re so significant because each cell is
treated as its own replicate.</p>
<pre class="r"><code>DT::datatable(de_result[1:100,])</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-a4d51c6334d8450be088" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-a4d51c6334d8450be088">{"x":{"filter":"none","vertical":false,"data":[["Rps29","Rpl37a","Rps27","Rpl39","mt-Atp6","Rps28","Rpl35","mt-Nd3","mt-Co3","Rps21","mt-Co2","mt-Co1","mt-Nd2","Rps19","Rpl38","Rpl36","Rps18","Rpl37","Rpl41","mt-Nd4","mt-Nd1","Rpl27a","Tpm2","Rpl36a","Prdx1","Rplp1","Rpl17","Rps23","mt-Cytb","Gm42418","Eef1a1","Sumo2","Rpl26","Oaz1","Rps26","Rpl28","Rps14","Rpl34","Pfn1","Rpl32","Rps9","Pdia6","Rack1","Rpl3","mt-Nd5","Ldha","Hnrnpa2b1","Tpt1","Pdia3","Actb","Rpl35a","Id3","Hspd1","H2afz","Mdk","Id2","Marcksl1","Gapdh","Rpl12","Ybx1","Ubc","Bsg","Eef2","Ran","Pgam1","Atp5k","Rpl13a","Rps12","Rplp2","Ptges3","U2af1","Psma3","Sox4","Eif5a","mt-Atp8","Aldoa","Rpl4","Ptma","Bzw1","Pgk1","Pgf","Foxp1","Ttc28","Spcs2","Pcbp1","Rhoa","Rps15a","Cox5a","Idh2","Igfbp4","Calm1","Atp5g3","Ostc","Tmed10","Pkm","Uqcr11","Ube2n","Heyl","Btg2","Snrpb"],[5.672010085064606e-145,1.211207884095274e-135,1.582890838048106e-123,3.019671532859554e-122,7.999740850588445e-117,9.360018853174469e-115,1.067709089815159e-111,6.277808959822926e-111,8.069471342041734e-110,4.946837463295477e-104,9.248632347876383e-99,3.795627870505453e-98,3.209834968188937e-97,6.245879735970555e-95,7.561698100973594e-92,1.924796126116113e-88,2.032642097881945e-78,1.049381530773901e-76,2.078630479344822e-76,4.810084713479711e-68,2.647642733505963e-67,2.114302228864748e-66,1.047168249116902e-63,3.188133718141319e-61,1.052564207395053e-60,8.3633334665012e-58,1.132084770525729e-50,1.751359126399225e-50,1.375259083715029e-48,7.38020771554834e-48,1.483838440196425e-47,3.130107838629643e-45,1.637012296705122e-44,1.952019452171758e-43,1.210872122968884e-42,3.134881916036424e-42,1.009197686797193e-41,1.119929700823194e-41,3.07781068585982e-41,7.764097935771811e-38,1.013600719921148e-36,1.186054836430618e-35,2.488971366435717e-35,3.32737836068836e-35,4.582039882518706e-35,4.662338588839319e-35,2.32545901831462e-33,2.653828794325416e-33,2.390336866323259e-32,3.096965549278416e-32,6.288958672462525e-32,1.426108283794575e-31,4.891465751383152e-31,6.721493881615951e-31,1.508253050559916e-30,1.805463045255496e-30,3.425541081219259e-30,6.520890371111313e-30,7.325732056669258e-30,1.168477905568295e-29,5.645544348902752e-29,1.179412318307755e-28,4.034677354965059e-28,5.636266463740808e-28,7.104707131787309e-28,9.013583685861374e-28,1.02444284457668e-27,2.247275509838223e-27,2.390700899804107e-27,4.419074082188423e-27,5.998720460517117e-27,6.725313363836557e-27,2.645548293270255e-26,2.496981210411345e-25,2.673331020891897e-25,7.207972753152911e-25,2.283502435800657e-24,3.289392142115892e-24,3.905270578129058e-24,8.241989133962067e-24,1.06253593683238e-23,2.076519839628049e-23,4.247929770401124e-23,6.018392197386493e-23,7.88997473247782e-23,8.054084610879452e-23,2.455432170571237e-22,4.417435626254982e-22,4.665995783954582e-22,7.591924482570746e-22,9.992331064856833e-22,1.360602275028694e-21,3.064242044359963e-21,3.543432417251729e-21,4.025363431618035e-21,4.933124799701297e-21,6.0159957835612e-21,6.932595246382151e-21,1.021117219011012e-20,1.597295108725575e-20],[1.141345377095659,0.869948158976003,1.131732531636984,0.7100687085146395,1.203380090199911,0.9988497205912195,0.77925762102445,1.533841209912128,1.070890138904474,0.6967146987523241,0.9624946432396726,0.9848340201209647,1.181692380444596,0.6016724510897635,0.5699001208678403,0.6956616727799427,0.5227137667650403,0.5479668015859875,0.4373721178260714,0.6925103206238745,0.8356694625503467,0.4248615910802265,1.11369090369812,0.4320665140807747,-0.7793263864593905,0.4478194443932111,0.3664971297296873,0.380378592624516,0.5211904784011967,0.7770391352942347,0.3676695842072863,-0.4720961191840711,0.4332936105784997,-0.5095029995112768,0.3467681032182419,0.2728818225764682,0.3015851129728429,0.3148486076651933,-0.4790808620970823,0.2897749863250993,0.271590353703683,-0.7543951523049797,0.3196610874230217,0.2460276395307899,0.7415625382230455,-0.7431771522005204,-0.5184237129745313,0.3078896039516534,-0.7074014971975613,-0.3481638740619317,0.2488499814347902,-0.828012276334489,-0.5305865942244994,-0.5858291156143283,-0.4961901636662605,-1.437231234614104,-0.6182550253919761,-0.5243573535587656,0.3972780133953515,-0.3719384443453544,0.6895217256711013,-0.4768892480020828,0.4130697614883356,-0.4712729918516172,-0.4704772390940821,0.6394783652773837,0.4948681327060904,0.323159753931888,0.2725190206364458,-0.5389488496164732,-0.5403437344041042,-0.4497098300253994,0.6638750449537907,-0.5024201001590591,1.663987312087541,-0.6314161895810686,0.3352779866392255,-0.2463867194479672,-0.5890607463397404,-0.5968997150102153,1.160809346665692,0.804955370001186,0.6970066529544816,-0.5172880913337086,-0.4415894586459497,-0.4762548374378963,0.1750448887864628,-0.4309431938780124,-0.4958970900219848,-0.8852987602286426,-0.5105000541285056,-0.461305428986172,-0.515931877020406,-0.5736847756263361,-0.4237704378203908,0.5195907840539356,-0.4785835485912162,1.471266146066189,1.250909108211322,-0.3826988394183308],[1,1,1,1,1,1,1,0.989,1,1,0.997,0.997,0.995,1,1,1,1,1,1,0.997,0.995,1,0.989,1,0.997,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0.973,1,1,0.965,0.979,0.995,1,0.949,1,1,0.907,0.992,1,1,0.46,1,1,1,1,0.984,0.995,1,0.997,0.997,0.976,0.987,1,1,0.973,0.979,0.997,0.971,0.997,0.476,0.976,0.997,1,0.944,0.971,0.694,0.904,0.923,0.984,0.989,0.984,1,0.995,0.984,0.803,0.981,0.981,0.955,0.947,0.997,0.973,0.96,0.455,0.612,0.992],[0.99,0.997,0.972,1,0.994,0.985,0.996,0.889,0.999,0.999,0.993,0.993,0.965,1,1,0.99,0.999,0.997,1,0.987,0.978,1,0.886,0.999,1,1,0.997,1,0.994,0.953,1,1,1,1,0.999,1,1,0.999,0.997,1,1,0.965,0.997,1,0.841,0.98,0.994,1,0.946,1,1,0.971,0.993,1,0.999,0.718,1,1,1,1,0.858,0.991,0.99,1,0.985,0.87,0.9399999999999999,1,0.996,0.969,0.974,0.981,0.886,0.993,0.192,0.972,0.981,1,0.9360000000000001,0.959,0.412,0.677,0.712,0.972,0.99,0.962,1,0.99,0.974,0.849,0.985,0.975,0.953,0.9370000000000001,0.994,0.873,0.9419999999999999,0.19,0.352,0.993],[1.759287368084489e-140,3.756803494098311e-131,4.909652512373809e-119,9.366115193470476e-118,2.481279619627018e-112,2.903197047689125e-110,3.311713283879679e-107,1.947188005068277e-106,2.502907926161085e-105,1.534360575990358e-99,2.868648295340818e-94,1.177289896594676e-93,9.955945120831625e-93,1.937284517705987e-90,2.34541189997898e-87,5.970140144374348e-84,6.304645995000429e-74,3.254866694001407e-72,6.447288157783836e-72,1.491943975580002e-63,8.212193466515444e-63,6.557931223269789e-62,3.248001758285895e-59,9.888634353558928e-57,3.264738402077234e-56,2.594055141304677e-53,3.511387332739653e-46,5.432190602352477e-46,4.265641099958906e-44,2.289119027131628e-43,4.602421689957251e-43,9.708655483077561e-41,5.077521040690277e-40,6.05457873480114e-39,3.755762063812587e-38,9.723463238970176e-38,3.130228465138854e-37,3.473685953043301e-37,9.546445404331404e-37,2.408190256738342e-33,3.143885352979424e-32,3.678786286156848e-31,7.720042487273664e-31,1.032052946134709e-30,1.421211310360827e-30,1.446117560100291e-30,7.212876237106457e-29,8.231380771359142e-29,7.414107858274853e-28,9.605858044196863e-28,1.950646311437701e-27,4.423360063845632e-27,1.517185932106512e-26,2.084805757260819e-26,4.678148486921693e-26,5.600004727468971e-26,1.062500077161777e-25,2.022584566407596e-25,2.272222312017104e-25,3.624267919701182e-25,1.751078490699167e-24,3.658183187695164e-24,1.251435875189512e-23,1.748200769058486e-23,2.20366701106647e-23,2.795743251843622e-23,3.177514371023489e-23,6.970374448865216e-23,7.415236980922399e-23,1.370664208072383e-22,1.860623125238594e-22,2.085990446061185e-22,8.205697141236351e-22,7.74488662033287e-21,8.291870827500398e-21,2.235696908845439e-20,7.082739505122897e-20,1.020270760720086e-19,1.21129777521829e-19,2.556417769681015e-19,3.295667715272994e-19,6.440741586574318e-19,1.317580376885317e-18,1.866724707863368e-18,2.447233462772645e-18,2.49813542375648e-18,7.616013963460807e-18,1.370156008195508e-17,1.447251912309193e-17,2.354787216758968e-17,3.099321326386644e-17,4.2201800764565e-17,9.504359548991299e-17,1.099066432858969e-16,1.248546975584966e-16,1.530107319123351e-16,1.865981412187177e-16,2.150283067570352e-16,3.167199278206456e-16,4.954330238734116e-16]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>p_val<\/th>\n      <th>avg_log2FC<\/th>\n      <th>pct.1<\/th>\n      <th>pct.2<\/th>\n      <th>p_val_adj<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"p_val","targets":1},{"name":"avg_log2FC","targets":2},{"name":"pct.1","targets":3},{"name":"pct.2","targets":4},{"name":"p_val_adj","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
<p>Some improvements/changes can be made:</p>
<ul>
<li><strong>Speed</strong>:
<ul>
<li>Add the <em>max.cells.per.ident</em> parameter to subsample down to
a specified maximum number of cells per group. It can be useful to set a
very low number when testing (e.g 20), then a much higher one (e.g. 500,
or no limit at all) for your real run.</li>
<li>If Seurat suggests you install a library called <em>presto</em> to
speed this up, do so.</li>
</ul></li>
<li><strong>Adjust filtering</strong>:
<ul>
<li>By default the results are filtered by a <em>logfc.threshold</em> of
0.1. Consider removing this threshold (set it to 0) to get stats for
genes that are present but unchanging.</li>
<li><em>min.cells.pct</em> by default 0.01, genes with very low
expression in both groups will be ignored. You can’t find differences
when there is that little expression, and keeping them in makes your
multiple hypothesis correction more aggressive. You might consider
disabling this test if you’ve already done your one filtering. But this
default is good enough.</li>
</ul></li>
</ul>
<pre class="r"><code>de_result &lt;- FindMarkers(so.np, 
                         ident.1  = &#39;Wnt4KO&#39;, 
                         ident.2  = &#39;Wnt4Het&#39;,
                         slot     = &#39;data&#39;,   # This is the default
                         test.use = &quot;wilcox&quot;,  # This is the default
                         logfc.threshold     = 0, 
                         min.pct             = 0.1, # this is the default.
                         max.cells.per.ident = 50  # subset for speed e.g 500, 200, or even less for testing.
)</code></pre>
<p>Below is the same contrast with fewer cells - notice the change in
p-values.</p>
<pre class="r"><code>DT::datatable(de_result[1:100,])</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-0b50cf44ed0c70fb0261" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-0b50cf44ed0c70fb0261">{"x":{"filter":"none","vertical":false,"data":[["Rps29","Rpl37a","Rps27","Rpl39","Rps28","Rps19","Rpl38","Rpl35","Rps21","mt-Nd3","mt-Atp6","Rpl41","mt-Co1","mt-Nd2","Rpl36","Rps18","mt-Co3","Rpl26","Rpl36a","mt-Co2","Rplp1","Rpl37","Rpl17","Prdx1","Rpl27a","Rps23","mt-Nd4","mt-Nd1","Actb","Rpl3","Tpm2","Rpl32","Oaz1","Spcs1","Rpl13a","Atp5k","Eef1a1","Sparc","Uba52","Rpl34","Sumo2","Idh2","Rps5","Rpl12","Atp5g3","Rpl4","mt-Cytb","Spcs2","Fjx1","Rps26","Rpl35a","Rps9","Sec13","Pgk1","Gapdh","Emd","Rack1","Marcksl1","Sf3b5","Fam136a","Tmem258","Rpl28","Tpt1","Myl12b","Hbb-y","Hnrnpa2b1","Gm42418","Rpl23","Pkm","Rps14","Ptges3","4732440D04Rik","Mdk","Id2","Calm1","Psmb2","Ndufc2","Ywhae","Ldhb","Cd81","Smim27","Rap1gds1","Bex2","Ddx39b","Spry1","Dcun1d5","Rps12","Ung","Wnt4","Tpi1","Cdc42","Cox7c","AU015791","Serpinh1","Eid2b","Siva1","Sh3glb1","Snrpg","2610001J05Rik","Pigx"],[3.739190024415634e-17,2.020149843731889e-16,1.637897277112716e-14,3.462082325291892e-14,6.193572581462564e-14,8.941486415112145e-14,3.796094168474121e-13,6.434089009989823e-12,1.091727223486018e-11,5.397099818415829e-11,1.288802440763671e-10,1.768218763984862e-10,5.990946365117422e-10,9.86651796225596e-10,1.725394872271081e-09,1.800446295837737e-09,2.526901541289027e-09,4.540477412852899e-09,7.760743214060704e-09,1.370288848121278e-08,3.419000052170309e-08,2.13022164198444e-07,2.210480895526745e-07,3.969710363453131e-07,5.489574436157695e-07,1.075055369275294e-06,1.657590302359313e-06,3.668349099684782e-06,4.192079216634898e-06,4.192079216634898e-06,5.725798230216098e-06,5.823819865105902e-06,6.633394451685836e-06,6.891984796222938e-06,8.868007446328295e-06,9.759312228714684e-06,9.761710633723086e-06,1.161985646779706e-05,1.494029313519279e-05,1.948416740862417e-05,2.416452817816957e-05,2.814290034574668e-05,2.814290034574668e-05,2.900999962280169e-05,3.476826119640425e-05,6.470282283706046e-05,7.059035528951115e-05,7.266301438527293e-05,7.57381262962068e-05,7.923180736547099e-05,8.391869292415667e-05,9.144291358863725e-05,9.770991689920417e-05,0.0001147384498494361,0.0001180221821304931,0.0001545603932689136,0.0001560364787527338,0.0001791111232737129,0.0001964010099829764,0.0002078049866046248,0.0002080906236574616,0.0002228122183097693,0.0002352055289019803,0.0002744526176934795,0.0002763783842187437,0.0002838649668247406,0.0003241795552481462,0.0003509707647241048,0.0003603321819866389,0.0003699266881303995,0.0003998406617858139,0.0004259811177813767,0.0004326952121699585,0.0004401283401219188,0.000467610199442518,0.0004924572634527742,0.0005319207204376999,0.0006043037374767204,0.0006275379218478792,0.0006357205511889689,0.0006523004553352842,0.0006545668225963414,0.0006871781829588234,0.0007394324525202081,0.0007404749301442385,0.0007520745483122146,0.0008801722846266725,0.0009531895565818192,0.0009533628147512753,0.0009591637420943635,0.0009713614591665939,0.0009954873607059295,0.0009966000512385384,0.00104541191761698,0.00108484953404232,0.001124656446562812,0.001128012112819663,0.001180371694780655,0.00118828048990355,0.00119478034486768],[1.141345377095659,0.869948158976003,1.131732531636984,0.7100687085146395,0.9988497205912195,0.6016724510897635,0.5699001208678403,0.77925762102445,0.6967146987523241,1.533841209912128,1.203380090199911,0.4373721178260714,0.9848340201209647,1.181692380444596,0.6956616727799427,0.5227137667650403,1.070890138904474,0.4332936105784997,0.4320665140807747,0.9624946432396726,0.4478194443932111,0.5479668015859875,0.3664971297296873,-0.7793263864593905,0.4248615910802265,0.380378592624516,0.6925103206238745,0.8356694625503467,-0.3481638740619317,0.2460276395307899,1.11369090369812,0.2897749863250993,-0.5095029995112768,-0.4200619289467297,0.4948681327060904,0.6394783652773837,0.3676695842072863,-0.4693970207649021,0.6041957912990945,0.3148486076651933,-0.4720961191840711,-0.4958970900219848,0.1888325715332524,0.3972780133953515,-0.461305428986172,0.3352779866392255,0.5211904784011967,-0.5172880913337086,-1.436226039989695,0.3467681032182419,0.2488499814347902,0.271590353703683,-0.6729829939964139,-0.5968997150102153,-0.5243573535587656,-0.5062902341052649,0.3196610874230217,-0.6182550253919761,-0.3323146214630175,-0.4029594794988882,0.3334148699119655,0.2728818225764682,0.3078896039516534,-0.7430604907039461,0.0731469673524705,-0.5184237129745313,0.7770391352942347,0.1644507313064478,-0.4237704378203908,0.3015851129728429,-0.5389488496164732,0.1152251368362309,-0.4961901636662605,-1.437231234614104,-0.5105000541285056,-0.4385236105156078,-0.2164611783674979,-0.2240720455697556,-0.5868033568586851,-0.5091763246137351,0.7599209907913267,-0.05725138992545764,-1.049689170633957,-0.3996942545272111,-1.631815147625863,-0.4770146029406968,0.323159753931888,-0.6677027923144085,-6.328811135600175,-0.4717712482180383,-0.3927475680422439,0.2520666894420671,0.5118615269701821,-0.4053480460179597,-0.2887563459296212,-0.4371097118525649,-0.3818469652153275,0.3587409219959197,0.3384799566958521,-0.5801263437847349],[1,1,1,1,1,1,1,1,1,0.989,1,1,0.997,0.995,1,1,1,1,1,0.997,1,1,1,0.997,1,1,0.997,0.995,1,1,0.989,1,1,0.912,0.987,0.976,1,0.957,0.888,1,1,0.984,1,1,0.981,0.997,1,0.984,0.157,1,1,1,0.739,0.971,1,0.766,1,1,0.947,0.801,0.984,1,1,0.899,0.995,0.995,1,1,0.997,1,0.973,0.237,1,0.46,0.981,0.987,0.995,0.997,0.862,0.963,0.37,0.21,0.364,0.995,0.191,0.904,1,0.511,0.005,0.944,0.976,0.997,0.282,0.997,0.242,0.965,0.753,0.987,0.439,0.79],[0.99,0.997,0.972,1,0.985,1,1,0.996,0.999,0.889,0.994,1,0.993,0.965,0.99,0.999,0.999,1,0.999,0.993,1,0.997,0.997,1,1,1,0.987,0.978,1,1,0.886,1,1,0.885,0.9399999999999999,0.87,1,0.977,0.649,0.999,1,0.974,1,1,0.975,0.981,0.994,0.972,0.32,0.999,1,1,0.762,0.959,1,0.776,0.997,1,0.911,0.819,0.915,1,1,0.874,0.968,0.994,0.953,1,0.994,1,0.969,0.174,0.999,0.718,0.985,0.98,0.975,0.994,0.868,0.9340000000000001,0.205,0.184,0.591,0.993,0.349,0.915,1,0.61,0.205,0.9389999999999999,0.95,0.987,0.164,0.991,0.237,0.953,0.716,0.949,0.344,0.794],[1.159784569872997e-12,6.265898770303201e-12,5.080265984420509e-10,1.073834074835786e-09,1.921060407592243e-09,2.773380841375334e-09,1.177434528235618e-08,1.995661388228543e-07,3.386210329086582e-07,1.674018450678038e-06,3.997478530516677e-06,5.484484140251845e-06,1.858211834068471e-05,3.060297876352931e-05,5.351657275323213e-05,5.584444275799908e-05,7.837690510616174e-05,0.0001408319879144584,0.0002407149722705208,0.0004250224920217767,0.001060471246181665,0.006607308466943139,0.006856248593655306,0.01231285063432258,0.01702701302863032,0.03334499238881179,0.05141347840827881,0.1137811840249229,0.1300257210623646,0.1300257210623646,0.1775970837066127,0.1806374207559897,0.2057479957079396,0.2137686924244469,0.2750589869627648,0.3027045873980433,0.302778978726189,0.3604130880616613,0.4634030721742747,0.6043404205132957,0.7495111705022854,0.8729083400240246,0.8729083400240246,0.89980315830044,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>p_val<\/th>\n      <th>avg_log2FC<\/th>\n      <th>pct.1<\/th>\n      <th>pct.2<\/th>\n      <th>p_val_adj<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"p_val","targets":1},{"name":"avg_log2FC","targets":2},{"name":"pct.1","targets":3},{"name":"pct.2","targets":4},{"name":"p_val_adj","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
<p>Its quite common practice, especially with so many highly significant
genes (some of which have a small log2FC), to filter our results by fold
change. Common thresholds are 1 (log2(2) - a doubling or halving of
expression), or log2(1.5) for 1.5 fold-changes. Note that threshold is
on ‘absolute’ values, so we don’t accidentally exclude downregulation;
like that Wnt4 being downregulated in its knockout!</p>
<pre class="r"><code>filter(de_result, abs(avg_log2FC) &gt;= log2(2) ) %&gt;% head(100) %&gt;% DT::datatable()</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-02919026514f0c819cc6" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-02919026514f0c819cc6">{"x":{"filter":"none","vertical":false,"data":[["Rps29","Rps27","mt-Nd3","mt-Atp6","mt-Nd2","mt-Co3","Tpm2","Fjx1","Id2","Bex2","Spry1","Wnt4","Sema3d","Cthrc1","Pmp22","Sgcg","Il33","Sufu","Wnk2","Dcx","Hist1h2ap","Pcdh8","Limch1","Plac8","Wdr86","Wfdc2","Anxa2","Id1","Krt36","Bambi","Mycn","Twist1","Scube3","Cdh3","Cd248","Ppm1e","Pgf","Lbx2","Cxcl12","mt-Atp8","Ifitm3","Dusp9","Fam212b","Slain1","Pax8","Olfml3","Ptpn13","Ocm","Mme","Sh3rf3","Dkk1","Gata6","Jag1","Tmem45a","Eva1a","Cox15","Pawr","Btg2","Usp2","Bgn","Pcolce","Lef1","Kdelr3","Fblim1","Sema3c","Rgs5","Sct","Heyl","Tm4sf1","Zcchc24","Rprm","Gucy1a1","Pltp","Fam53c","Pth1r","Wdr66","Cpa1","Dync1i1","Cnn2","Pcp4","Clrn1","Whrn","Fam43a","S100a6","Bmf","Calb1","Npnt","Krt18","Tmem100","Acta2","Alx1","Lmcd1","Igfbp3","Prrx2","Osbpl6","Slc38a4","Lum","Col9a1","Ebf1","Myl9"],[3.739190024415634e-17,1.637897277112716e-14,5.397099818415829e-11,1.288802440763671e-10,9.86651796225596e-10,2.526901541289027e-09,5.725798230216098e-06,7.57381262962068e-05,0.0004401283401219188,0.0006871781829588234,0.0007404749301442385,0.0009533628147512753,0.00144226289251369,0.001823469898389335,0.001823469898389335,0.001823469898389335,0.002261109569970181,0.00272573938021145,0.003462445825637323,0.003880689040435405,0.004065649965075455,0.004210696680334542,0.004744974084182041,0.004835983384043182,0.005134490361044192,0.005918140831154392,0.006693551685765428,0.00687555424142264,0.007489041876428992,0.008292796772711936,0.008723358622398114,0.01052872808001597,0.01228469430847926,0.01228469430847926,0.01228469430847926,0.01228469430847926,0.01238112390127574,0.01276056193863199,0.01276056193863199,0.01281285213919313,0.01295636032508456,0.01359250228568578,0.01370530321700248,0.01384715458004431,0.01543397165069414,0.01561961885276146,0.01598100636485632,0.01705054940634856,0.01723875407266108,0.01789958397385919,0.02305207930283293,0.02305207930283293,0.02305207930283293,0.02305207930283293,0.02305207930283293,0.02384200142297638,0.02508784976794634,0.02592378667180131,0.02592433457471256,0.02766722422459518,0.02900764343993956,0.02900764343993956,0.02900764343993956,0.02911586677996234,0.02951117888647054,0.03010325953517135,0.03047214616689446,0.03163616042252495,0.03226887089949784,0.03641980173961652,0.04208441402292591,0.04334932504632747,0.04506227876095176,0.047845866424879,0.04866278486658943,0.04932313930000851,0.05047114748276797,0.05068192177571,0.06218606018597767,0.06305216456793385,0.06512498965459471,0.07065127676828115,0.07065127676828115,0.07130293957390406,0.0724961498762105,0.07435125075893576,0.07435125075893576,0.08288081248267136,0.09264217341762082,0.09513116341739079,0.09730254635799065,0.09730254635799065,0.09730254635799065,0.09730254635799065,0.1007298234701856,0.1161357674630457,0.1170215596714272,0.1208443558612252,0.1230698051353015,0.1232557544035497],[1.141345377095659,1.131732531636984,1.533841209912128,1.203380090199911,1.181692380444596,1.070890138904474,1.11369090369812,-1.436226039989695,-1.437231234614104,-1.049689170633957,-1.631815147625863,-6.328811135600175,1.053724931418765,-2.571176596043605,-1.839380490465447,1.240415141317815,-1.026849968933194,1.002750076198957,1.108259616654097,1.207664748631319,1.50676590507918,1.410829100749221,-1.397755998294669,-1.902896201731551,-1.053990299841022,-1.266681661366076,-1.35688286450505,-1.221649991385377,1.714799430240826,-1.775345801261402,-1.059283211270084,-1.940826158197121,-2.372780305460654,-2.021722651393042,-1.13623935109037,1.108758899845212,1.160809346665692,-1.299365157677223,-1.016675492953485,1.663987312087541,-1.542543852414661,-1.019423762088479,1.554846087651113,1.418641996903334,-1.681740053902859,-1.269313283711512,-1.319690940793705,1.63263162944905,-1.58241292032349,1.539670907198014,-4.329929097250758,-2.229912498370854,-2.715937701448898,-1.45146072776922,-1.111807407680145,1.058693676327862,-1.693628679506354,1.250909108211322,-1.022489936163666,-2.047912894195685,-1.34594342402705,-1.234331025054695,-1.446983733852258,-1.329390327273341,-1.361080377836946,-3.414386809915026,-1.339659968360467,1.471266146066189,-2.078509672687953,1.390468315297481,-2.839829219052815,-4.116571718234443,1.024073845650707,1.137261337863268,-1.564694566052472,1.082176451118352,-1.22687957292938,1.320567247331956,-1.436056447530588,-1.111730771880389,1.006898150644346,1.392132529931014,1.009054666360824,-1.407800720160705,1.31598041286749,-3.363140122832683,-1.445533126968505,-1.143602567307642,-1.514929272409633,-2.25642393092812,-2.841454384166053,-2.21520276517981,-2.162901350969043,-1.55904361257819,-1.025604080317995,-1.008341548474531,-1.311993320527852,-1.309943880034149,-1.231723470690643,-1.380008050838422],[1,1,0.989,1,0.995,1,0.989,0.157,0.46,0.364,0.191,0.005,0.457,0.04,0.07199999999999999,0.104,0.098,0.245,0.112,0.489,0.383,0.21,0.06900000000000001,0.178,0.056,0.378,0.394,0.481,0.335,0.112,0.266,0.07199999999999999,0.029,0.029,0.04,0.104,0.694,0.154,0.295,0.476,0.146,0.162,0.109,0.152,0.152,0.138,0.056,0.362,0.04,0.128,0.005,0.019,0.019,0.043,0.048,0.152,0.053,0.612,0.056,0.048,0.125,0.059,0.043,0.136,0.082,0.019,0.066,0.455,0.051,0.306,0.037,0.008,0.622,0.191,0.082,0.327,0.051,0.245,0.231,0.08500000000000001,0.306,0.245,0.234,0.165,0.258,0.045,0.059,0.138,0.056,0.07199999999999999,0.019,0.029,0.027,0.048,0.061,0.138,0.074,0.074,0.098,0.274],[0.99,0.972,0.889,0.994,0.965,0.999,0.886,0.32,0.718,0.591,0.349,0.205,0.215,0.173,0.162,0.038,0.173,0.129,0.044,0.254,0.246,0.082,0.143,0.368,0.114,0.598,0.528,0.62,0.115,0.224,0.418,0.194,0.126,0.117,0.107,0.038,0.412,0.289,0.346,0.192,0.311,0.285,0.037,0.058,0.273,0.272,0.142,0.14,0.126,0.044,0.129,0.105,0.101,0.107,0.111,0.083,0.14,0.352,0.105,0.181,0.237,0.143,0.123,0.237,0.18,0.161,0.154,0.19,0.149,0.146,0.177,0.127,0.346,0.089,0.196,0.152,0.104,0.08599999999999999,0.395,0.117,0.158,0.098,0.12,0.311,0.114,0.202,0.114,0.199,0.142,0.183,0.111,0.114,0.107,0.121,0.105,0.218,0.118,0.127,0.183,0.39],[1.159784569872997e-12,5.080265984420509e-10,1.674018450678038e-06,3.997478530516677e-06,3.060297876352931e-05,7.837690510616174e-05,0.1775970837066127,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>p_val<\/th>\n      <th>avg_log2FC<\/th>\n      <th>pct.1<\/th>\n      <th>pct.2<\/th>\n      <th>p_val_adj<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"p_val","targets":1},{"name":"avg_log2FC","targets":2},{"name":"pct.1","targets":3},{"name":"pct.2","targets":4},{"name":"p_val_adj","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="run-differential-expression-for-every-cluster"
class="section level2">
<h2>Run differential expression for every cluster</h2>
<p>We typically want to know about changes in every cluster. This can
make for a lot of contrasts!</p>
<pre><code>2x comparisons * 19x clusters = 38 sets of differential results </code></pre>
<p>We can simply loop through and run them one-by-one. The code below
makes a big table of all the tests - recording which contrast was being
done in which cluster.</p>
<p>NB: Note that the <em>FindMarkers()</em> function returns a table
with feature names as a rowname. Rownames are expected to be unique. So
while that is fine for one comparison, but when we want to put more
together, we need to give it its own column. If we don’t, there’ll be
gene names with ‘.1’ appended to the end!</p>
<pre class="r"><code># Set threhoehsolds
min_cells_per_group &lt;- 20

# Calculate DE across every celltype
# Empty list to collect results
de_result_list     &lt;- list()
de_result_sig_list &lt;- list()

# the contrasts
# If we store them in a list we can give the nice names
contrast_list &lt;-  list( # &#39;Test&#39;      vs  &#39;control&#39;  
                       FloxWnt4KOvsFloxHet  = c(&#39;Wnt4FloxKO&#39;,      &#39;Wnt4FloxHet&#39; ) ,
                       TotalWnt4KOvsHet     = c(&#39;Wnt4KO&#39;,          &#39;Wnt4Het&#39;     ) )

## Or you could autogenerated them
#make_contrast_name &lt;- function(contrast_parts){ paste0(contrast_parts[1],&quot;vs&quot;,contrast_parts[1])}
#contrast_names &lt;- sapply(FUN=make_contrast_name, X=contrast_list) 
#names(contrast_list) &lt;- contrast_names


# Loop through every celltype/cluster
for (the_celltype in levels(so$CelltypeCode)) {
  
  # Subset to one cell type. 
  print(the_celltype)
  
  so.this &lt;- subset(so, CelltypeCode == the_celltype)
  
  # count how many cells within each sample (GTshort)
  # And list which samples have more than the minimum
  cells_per_sample &lt;- table(so.this$GTshort)
  print(cells_per_sample)
  samples_with_enough_cells &lt;- names(cells_per_sample)[cells_per_sample &gt; min_cells_per_group]
  

  ## Loop through every contrast 
  # For each listed contrast, do both sides of the contrast have enough cells?
  # This is of course much simpler if you only have one contrast!
  for (contrast_name in names(contrast_list)) {
    
    # from mycontrastname pull out list of the two samples involved;
    # c(&#39;test&#39;, &#39;control&#39;)
    contrast &lt;- contrast_list[[contrast_name]]
    
    # Only run this contrast if both sides pass!
    if (all(contrast %in% samples_with_enough_cells)) {
      print(contrast_name) 

      # We need to tell Seurat to group by _sample_ not cluster.
      Idents(so.this) &lt;- so.this$GTshort
      
      
      de_result &lt;- FindMarkers(so.this, 
                               ident.1  = contrast[1], 
                               ident.2  = contrast[2],
                               slot     = &#39;data&#39;,
                               test.use = &quot;wilcox&quot;, # the default
                               min.pct  = 0.01,   # Note 
                               logfc.threshold = 0, 
                               max.cells.per.ident = 100 # 1000 # if you have really big clusters, set this to subsample!
                               )
  
      
      # It can be helpful to know the average expression of a gene
      # This will give us the average (per cell) within this celltype.
      avg_expression &lt;- rowMeans(GetAssayData(so.this, assay = &#39;SCT&#39;, layer=&quot;data&quot;))
      
      
      
      de_result.formatted &lt;- de_result %&gt;%
      rownames_to_column(&quot;target&quot;) %&gt;%
      mutate(contrast=contrast_name,
             celltype=the_celltype,
             avg_expression=avg_expression[target]) %&gt;%
      select(celltype,contrast,target,avg_expression, everything()) %&gt;%
      arrange(p_val)
  
      # Filter to just significant results, optionally by log2FC.
      de_result.sig &lt;- filter(de_result.formatted, 
                         p_val_adj &lt; 0.01,
                         abs(avg_log2FC) &gt;= log2(1.5) )

      # Record these results in a list to combine
      full_name &lt;- paste(contrast_name, the_celltype)
      de_result_list[[full_name]] &lt;- de_result.formatted
      de_result_sig_list[[full_name]] &lt;- de_result.sig
      
    }
  }
}</code></pre>
<pre><code>[1] &quot;c0: Blood&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
       1531         567        2597         189 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c1: Stroma&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        553        2014         750         130 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c2: High Mitochondria - Non Wnt4KO&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        868         678        1353          69 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c3: NP&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        587        1225         684         376 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c4: Stroma&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        403        1088         435         283 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c5: Stroma&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        283         587         339         511 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c6: Stroma&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        343         841         391          32 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c7: Stroma&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        326         672         242         330 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c8: Immune&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        408         701         228         144 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c9: Stroma - Wnt4KO&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
         33          41          23        1383 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c10: UE - Cortical&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        325         408         345          35 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c11: UE - Tip&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        251         466         307          47 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c12: Stroma&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        189         658         177          28 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c13: Vasculature&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        295         245         227          22 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c14: High Mitochondria&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        125          16         140         381 
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c15: Podocyte&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
         61          65          49         110 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;TotalWnt4KOvsHet&quot;
[1] &quot;c16: Blood&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
        171          28           8           9 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;c17: Stroma&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
         21          94          13          17 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;c18: Immune&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
         29          59          24           6 
[1] &quot;FloxWnt4KOvsFloxHet&quot;
[1] &quot;c19: UE - Medullary&quot;

Wnt4FloxHet  Wnt4FloxKO     Wnt4Het      Wnt4KO 
         34          58           8           1 
[1] &quot;FloxWnt4KOvsFloxHet&quot;</code></pre>
<pre class="r"><code># Join together results for all celltypes, and pull out those with a singificant adjusted p-value
de_results_all &lt;- bind_rows(de_result_list)
de_results_sig &lt;- bind_rows(de_result_sig_list)</code></pre>
<p>Check out the full set of significant DE genes</p>
<pre class="r"><code>DT::datatable(de_results_sig)</code></pre>
<p>Save the results.</p>
<pre class="r"><code># Save the full set of results as a tab-deliminated text file.
# This is useful for parsing later e.g. for functional enrichment
write_tsv(x = de_results_all, &quot;~/myproject/de_results_all.tsv&quot;)

# If you don&#39;t have too many contrasts and celltypes, 
# you can save it as an excel file, with each contrast in a seprate tab.
# In this canse, excel can&#39;t have &#39;:&#39;symbols in the sheet names (which are drawn from the list names), so we replace them.
# Also consider keeping names short, use a cluster code instead of cell type. 
library(writexl)
de_result_list.excel &lt;- setNames(de_result_list, nm=gsub(&quot;: &quot;,&quot;_&quot;,names(de_result_list)))
de_result_sig_list.excel &lt;- setNames(de_result_sig_list, nm=gsub(&quot;: &quot;,&quot;_&quot;,names(de_result_sig_list)))
write_xlsx(de_result_list.excel,      &quot;~/myproject/de_results_all.xlsx&quot;)
write_xlsx(de_result_sig_list.excel,  &quot;~/myproject/de_results_sig.xlsx&quot;)</code></pre>
</div>
<div id="visualising-de-results" class="section level2">
<h2>Visualising DE results</h2>
<div id="contrast-level-visualisation" class="section level3">
<h3>Contrast level visualisation</h3>
<p>Lets get some contrast-level visualisations of our differential
expression results.</p>
<pre class="r"><code>library(ggrepel) #library for non-overlapping plot text

# Get the results of Wnt4KO vs Het , within the c6 Stromal cluster.
de_result &lt;- filter(de_results_all, celltype == &quot;c3: NP&quot;, contrast == &quot;TotalWnt4KOvsHet&quot;)

# Would like to show a few genes on the plot - the top 5 most significant with 2-fold change.
top5_genes.2fc &lt;- arrange(de_result, p_val_adj) %&gt;% filter( abs(avg_log2FC ) &gt; 1) %&gt;% head(n=5)</code></pre>
<pre class="r"><code>ggplot(de_result, aes(x=avg_expression, y=avg_log2FC, col=p_val_adj &lt; 0.01)) +
  geom_hline(yintercept = 0, col=&#39;grey30&#39;) +
  geom_point(pch=3) +
  geom_point(data=top5_genes.2fc) +
  geom_text_repel(data=top5_genes.2fc, mapping=aes(label=`target`)) + 
  scale_colour_manual(values=c(&#39;FALSE&#39;=&#39;black&#39;, &#39;TRUE&#39;= &#39;red&#39;)) +
  theme_bw() +
  theme(legend.position = &quot;None&quot;) +
  ggtitle(&quot;Differential expression within c3: NP, Wnt4KO vs Het&quot;)</code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-21-1.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>And a volcano plot</p>
<pre class="r"><code>ggplot(de_result, aes(x=avg_log2FC, y=-log10(p_val) ,col=p_val_adj &lt; 0.01)) +
  geom_hline(yintercept = 0, col=&#39;grey30&#39;) +
  geom_point(pch=3) +
  geom_point(data=top5_genes.2fc) +
  geom_text_repel(data=top5_genes.2fc, mapping=aes(label=`target`)) + 
  scale_colour_manual(values=c(&#39;FALSE&#39;=&#39;black&#39;, &#39;TRUE&#39;= &#39;red&#39;)) +
  theme_bw() +
  theme(legend.position = &quot;None&quot;) +
  ggtitle(&quot;DE volcano plot within c3: NP, Wnt4KO vs Het&quot;)</code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-1">
Past versions of unnamed-chunk-22-1.png
</button>
</p>
<div id="fig-unnamed-chunk-22-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-22-1.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/87886e17e98fe7f1db4bfadbcb54ab241f3e60cb/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-22-1.png" target="_blank">87886e1</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="plot-some-individual-genes" class="section level3">
<h3>Plot some individual genes</h3>
<p>For the above differential expression, there were alot of ribosomal
and mitochondrial genes listed as differentially expressed. This can
occur because they generally have high expression, therefore high
statistical power to detect even small changes. These aren’t
neccessarily the most interesting genes in the experiment.</p>
<p>It is worth plotting the expression of individual genes of particular
interest, particuarly if you’re planning downstream work! Plotting two
of these results below.</p>
<ul>
<li>Rps29 : The most significantly DE gene (adjusted pval 6.5 *
10^-28)</li>
<li>Stc1 : At an average expression of 0.0926, log2FC of 1.5 and a
p-value of 0.0057, its much less statiticaly confident, but might be
interesting experimentally</li>
</ul>
<pre class="r"><code># for plotting, make a subset with just that cluster
so.cluster &lt;- subset(so, CelltypeCode == &#39;c3: NP&#39;)

# Subset the results
de_result.sig &lt;- filter(de_results_all, 
                        celltype == &quot;c3: NP&quot;, 
                        contrast == &quot;TotalWnt4KOvsHet&quot;,
                        p_val_adj &lt;= 0.01
                        )

DT::datatable(filter(de_result.sig, p_val_adj &lt; 0.01))</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-33e8fa5db7297a737495" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-33e8fa5db7297a737495">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65"],["c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP","c3: NP"],["TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet","TotalWnt4KOvsHet"],["Rps29","Rpl37a","Rps27","Rpl39","Rpl35","mt-Atp6","Rps28","mt-Nd3","Rps21","mt-Co3","mt-Nd2","Rps19","Rpl38","mt-Co1","mt-Co2","Rpl41","Rpl36","Rps18","Rpl37","Tpm2","Prdx1","Rpl27a","mt-Nd4","mt-Nd1","Rpl36a","Sumo2","Rpl26","Gm42418","Rps23","Oaz1","Rplp1","Rpl17","Marcksl1","Rps9","Eif5a","Eef1a1","Rpl32","Pfn1","Rps26","Pkm","Actb","Ran","Atp5k","mt-Cytb","Ubc","Id2","Btf3","Gapdh","Pdia6","Rpl34","Atp5g3","Ppib","Rps5","Mdk","Tagln2","Rpl3","Rpl12","Calm1","Shisa2","Rplp2","Stc1","Tpt1","Ddx5","mt-Nd5","Actg1"],[2.079715623968678,2.698301363469256,1.750306015176746,2.735460340614041,1.988070193854112,2.819062987898909,1.7513612222394,1.161089199850793,2.278511318678834,3.354271514721082,1.618631363536945,2.431295341149375,2.273623530093516,2.708293496634333,2.75760016817251,3.443191105032489,1.776279148187099,2.570052853263712,2.446455556687055,0.9138073041388981,2.081634924267746,2.967344345408981,2.403571998791985,2.187227959457461,2.17693922364207,2.202180410499853,2.065513618774803,1.223221033473104,2.343764045638861,2.169522191317669,2.783045696557705,2.628414256374549,1.912161787926337,3.102158040896752,1.643768583933485,4.22033687389576,2.88419231884185,1.883300675648638,3.036787836314379,1.6200528311188,3.16660301601941,2.115122529153734,0.6831847891575799,3.03082369711699,0.7550446300534147,0.5571951901495856,2.530564638004793,2.955054817467254,1.001264742528088,2.518413077173242,1.136787220671215,1.538422784355603,3.131529138666562,2.250672794892088,0.168016490247308,2.784705804452274,2.683426613342766,1.268631837986376,0.3121432741966269,2.13931452716356,0.09268256455263139,3.401867238778437,1.33440705979125,0.7828250516016548,2.908671404253379],[2.109640716000917e-32,2.792572876082553e-29,3.315155695118228e-27,3.88902670208015e-27,2.959394419103094e-26,4.087123620191987e-23,5.750106872172456e-23,1.190266547421396e-22,1.51615900869173e-22,4.67925531806146e-22,1.759566457460148e-21,4.090759952277071e-21,2.155672442676362e-20,2.180391758688967e-20,5.055487985503603e-19,2.352929628541086e-17,1.116944625666035e-16,1.002638258297604e-15,3.43923355039114e-14,7.577388754217871e-14,1.700114836183224e-13,3.102985914974568e-13,3.276134037679538e-13,7.227269125394184e-13,9.529165844160837e-13,2.531806532401785e-12,2.667704665350734e-12,3.229675542661777e-12,9.213905768465082e-12,1.250115639824919e-11,1.609744305306939e-11,2.794484358766765e-11,2.577505936196487e-10,2.577505936196487e-10,3.532242831345812e-10,6.192989252972161e-10,6.588538487883259e-10,7.454954479257442e-10,1.293869553121418e-09,1.917708504364778e-09,2.162820959007786e-09,3.48620790591243e-09,7.802223523500829e-09,9.843013198697308e-09,1.131372017205697e-08,1.575733166691073e-08,2.446524700495433e-08,5.321918953413542e-08,5.46767320716746e-08,6.71375767921237e-08,7.534996761738627e-08,8.009979951332802e-08,8.687046057632108e-08,9.168864971968069e-08,1.028266444394451e-07,1.041903810110158e-07,1.167650367002771e-07,1.425902831410597e-07,1.489341889961872e-07,1.650164723744057e-07,1.848112188057907e-07,2.091904454104403e-07,2.116878803213689e-07,2.635935683796153e-07,3.093200551056544e-07],[1.141345377095659,0.869948158976003,1.131732531636984,0.7100687085146395,0.77925762102445,1.203380090199911,0.9988497205912195,1.533841209912128,0.6967146987523241,1.070890138904474,1.181692380444596,0.6016724510897635,0.5699001208678403,0.9848340201209647,0.9624946432396726,0.4373721178260714,0.6956616727799427,0.5227137667650403,0.5479668015859875,1.11369090369812,-0.7793263864593905,0.4248615910802265,0.6925103206238745,0.8356694625503467,0.4320665140807747,-0.4720961191840711,0.4332936105784997,0.7770391352942347,0.380378592624516,-0.5095029995112768,0.4478194443932111,0.3664971297296873,-0.6182550253919761,0.271590353703683,-0.5024201001590591,0.3676695842072863,0.2897749863250993,-0.4790808620970823,0.3467681032182419,-0.4237704378203908,-0.3481638740619317,-0.4712729918516172,0.6394783652773837,0.5211904784011967,0.6895217256711013,-1.437231234614104,-0.1920590480760573,-0.5243573535587656,-0.7543951523049797,0.3148486076651933,-0.461305428986172,-0.3475774236451379,0.1888325715332524,-0.4961901636662605,-1.833586318706401,0.2460276395307899,0.3972780133953515,-0.5105000541285056,0.7473127245458242,0.2725190206364458,1.455966767754698,0.3078896039516534,0.4141502081749122,0.7415625382230455,-0.4191850875206766],[1,1,1,1,1,1,1,0.989,1,1,0.995,1,1,0.997,0.997,1,1,1,1,0.989,0.997,1,0.997,0.995,1,1,1,1,1,1,1,1,1,1,0.997,1,1,1,1,0.997,1,0.997,0.976,1,0.984,0.46,1,1,0.973,1,0.981,0.992,1,1,0.277,1,1,0.981,0.723,1,0.399,1,0.997,0.965,0.997],[0.99,0.997,0.972,1,0.996,0.994,0.985,0.889,0.999,0.999,0.965,1,1,0.993,0.993,1,0.99,0.999,0.997,0.886,1,1,0.987,0.978,0.999,1,1,0.953,1,1,1,0.997,1,1,0.993,1,1,0.997,0.999,0.994,1,1,0.87,0.994,0.858,0.718,1,1,0.965,0.999,0.975,0.993,1,0.999,0.484,1,1,0.985,0.462,0.996,0.178,1,0.9370000000000001,0.841,1],[6.543472608820045e-28,8.661723289745255e-25,1.028261841954821e-22,1.2062594121842e-22,9.179153669732066e-22,1.267703133274948e-18,1.783510648541731e-18,3.691849750136943e-18,4.70267039725914e-18,1.451364622003123e-17,5.457647281104142e-17,1.268831014397779e-16,6.686249215449273e-16,6.762921117925569e-16,1.568060708463652e-14,7.298081828845886e-13,3.464427145428341e-12,3.10988308576168e-11,1.06674707032482e-09,2.350278669895757e-09,5.273246187389506e-09,9.624531412476618e-09,1.016158494467062e-08,2.241682064623514e-08,2.955661369883367e-08,7.852904321550616e-08,8.27441956051837e-08,1.001748463067403e-07,2.857877152204814e-07,3.877483680044952e-07,4.992943911770534e-07,8.667652135586874e-07,7.994650162300643e-06,7.994650162300643e-06,1.09559575899853e-05,1.920879476594375e-05,2.043566982786751e-05,2.312303230831281e-05,4.013195192916702e-05,5.948156467988234e-05,6.708421768554449e-05,0.0001081317106176858,0.0002420015670284252,0.0003053007403839944,0.0003509176585766909,0.00048874515631257,0.0007588385663526685,0.001650699601780278,0.001695908198667131,0.002082406219361301,0.00233712994558847,0.002484455481504895,0.002694461075695751,0.002843906848355336,0.003189374030578269,0.003231673047818676,0.003621701143332495,0.00442272281218625,0.004619491740094738,0.005118315923636943,0.00573228957369921,0.006488460045295626,0.0065659229839279,0.008175881710430528,0.009594180149212083]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>celltype<\/th>\n      <th>contrast<\/th>\n      <th>target<\/th>\n      <th>avg_expression<\/th>\n      <th>p_val<\/th>\n      <th>avg_log2FC<\/th>\n      <th>pct.1<\/th>\n      <th>pct.2<\/th>\n      <th>p_val_adj<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[4,5,6,7,8,9]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"celltype","targets":1},{"name":"contrast","targets":2},{"name":"target","targets":3},{"name":"avg_expression","targets":4},{"name":"p_val","targets":5},{"name":"avg_log2FC","targets":6},{"name":"pct.1","targets":7},{"name":"pct.2","targets":8},{"name":"p_val_adj","targets":9}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
<p>First step would be a violin plot. Rps29 has very high expresison,
and is in practically every cell. Stc1 is more typical low expression,
but still quite informative. The black line at the bottom of these
violin plots is typical - that’s cells with zero copies of the gene
dominating the plot.</p>
<pre class="r"><code>VlnPlot(so.cluster, features=c(&#39;Rps29&#39;, &#39;Stc1&#39;))</code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-24-1.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Normalised expression across the umap can give a good overview of how
specific your DE gene is expressed. Can see that there’s alot more
expression of Rps29 overall, and its pretty ubiquitous.</p>
<pre class="r"><code>FeaturePlot(so, features=&#39;Rps29&#39;) </code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-25-1">
Past versions of unnamed-chunk-25-1.png
</button>
</p>
<div id="fig-unnamed-chunk-25-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-25-1.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/87886e17e98fe7f1db4bfadbcb54ab241f3e60cb/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-25-1.png" target="_blank">87886e1</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>FeaturePlot(so, features=&#39;Stc1&#39;)</code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-25-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-25-2">
Past versions of unnamed-chunk-25-2.png
</button>
</p>
<div id="fig-unnamed-chunk-25-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-25-2.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>It can also be worth checking how high counts are - note that Stc1
maxes out at 8 copies in a cell. Rps29 looks more specific now - but
many of the low counts are in the grey (This isn’t a good visualisation,
its just for checking counts).</p>
<pre class="r"><code>FeaturePlot(so, features=&#39;Rps29&#39;, slot = &#39;counts&#39;)</code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-26-1">
Past versions of unnamed-chunk-26-1.png
</button>
</p>
<div id="fig-unnamed-chunk-26-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-26-1.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/87886e17e98fe7f1db4bfadbcb54ab241f3e60cb/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-26-1.png" target="_blank">87886e1</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>FeaturePlot(so, features=&#39;Stc1&#39;, slot = &#39;counts&#39;)</code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-26-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-26-2">
Past versions of unnamed-chunk-26-2.png
</button>
</p>
<div id="fig-unnamed-chunk-26-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-26-2.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/87886e17e98fe7f1db4bfadbcb54ab241f3e60cb/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-26-2.png" target="_blank">87886e1</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>If the Seurat default plots aren’t giving enough control for plotting
your experiment; The ‘fetchData’ function builds a table from the
various sources in the seurat data (since gene expression or UMAP
coordinates arne’t in the metadata). You can build your own pretty plots
from there.</p>
<pre class="r"><code>plottable &lt;- 
  FetchData(so, c(&#39;GTshort&#39;, &#39;CelltypeCode&#39;, # From metadata table
                  &#39;UMAP_1&#39;,&#39;UMAP_2&#39;,         # From dimensional reductions
                  &#39;Rps29&#39;,&#39;Stc1&#39;             # Genes from default assay 
                ),
          slot = &#39;data&#39; #layer to get (slot is pre seurat 5 termnology)
  )
head(plottable)</code></pre>
<pre><code>                       GTshort                       CelltypeCode      UMAP_1
S1_AAACCCACAAGGTTGG Wnt4FloxKO c2: High Mitochondria - Non Wnt4KO  -4.5556949
S1_AAACCCACACCTCTGT Wnt4FloxKO                         c8: Immune -12.0080126
S1_AAACCCACAGGGCTTC Wnt4FloxKO                         c4: Stroma   5.9303346
S1_AAACCCAGTTAGGACG Wnt4FloxKO                         c1: Stroma   0.8406711
S1_AAACCCAGTTGACTGT Wnt4FloxKO                      c11: UE - Tip  -1.4513678
S1_AAACCCAGTTGTCCCT Wnt4FloxKO                         c1: Stroma   2.0917211
                          UMAP_2    Rps29 Stc1
S1_AAACCCACAAGGTTGG  -3.57434729 2.095264    0
S1_AAACCCACACCTCTGT  -0.05614393 1.972429    0
S1_AAACCCACAGGGCTTC   3.14180413 1.385021    0
S1_AAACCCAGTTAGGACG   2.14208197 2.134845    0
S1_AAACCCAGTTGACTGT -10.38371904 2.696576    0
S1_AAACCCAGTTGTCCCT   1.97180467 2.897872    0</code></pre>
<p>Get some finer control on a UMAP plot.</p>
<pre class="r"><code>ggplot(plottable, aes(x=UMAP_1, y=UMAP_2, col=Stc1)) +
  geom_point(pch=3) +
  theme_bw() +
  scale_colour_viridis_c(direction=-1) +
  facet_wrap(~GTshort)</code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-28-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-28-1">
Past versions of unnamed-chunk-28-1.png
</button>
</p>
<div id="fig-unnamed-chunk-28-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-28-1.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>A compact view of gene expression across all clusters (is the change
specific to a cluster, or global?)</p>
<pre class="r"><code>ggplot(plottable, aes(x=CelltypeCode, y=Stc1, fill=GTshort)) +
  geom_boxplot( outlier.alpha = .5, outlier.shape = 3) + 
  xlab(&quot;&quot;) +
  coord_flip() + 
  theme_bw() </code></pre>
<p><img src="figure/e_DEWithoutReps.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-29-1">
Past versions of unnamed-chunk-29-1.png
</button>
</p>
<div id="fig-unnamed-chunk-29-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/f4435807e2deed7f8ea260143e754b83f3ee9123/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-29-1.png" target="_blank">f443580</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/87886e17e98fe7f1db4bfadbcb54ab241f3e60cb/docs/figure/e_DEWithoutReps.Rmd/unnamed-chunk-29-1.png" target="_blank">87886e1</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-08
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
<div id="code-snippet" class="section level1">
<h1>Code Snippet</h1>
<p>The following code assumes your seruat object is called so, and has
the following metadata columns;</p>
<ul>
<li>cluster : The cluster or celltype</li>
<li>samplename: The samples to compare (n=1). For the snippet, these are
called ‘Test’ and ‘Control’</li>
</ul>
<p>Update as needed.</p>
<pre class="r"><code># Libraries
library(Seurat)
library(tidyverse)
library(writexl)

# Ensure that relevant metadata are set as factors
so$samplename  &lt;- factor(so$samplename) # levels are &#39;Test&#39; and &#39;Control&#39;
so$cluster     &lt;- factor(so$cluster) 

## Set default assay
DefaultAssay(so) &lt;- &quot;RNA&quot;


## Set thresholds
min_cells_per_group &lt;- 20 # Skip comparisons, when fewer than this many cells in one side of the comparison.
p_value_threshold   &lt;- 0.01 # threshold for significance (after correction)
log_fc_threshold    &lt;- 0    # threshold for sitnificange (0=none), log2(1.5), log2(2) also options.



## Define contrasts
# Note this is a list of pairs of samplenames from &#39;samplename&#39; column.
# There can be multiple contrasts.
# This will build names for each contrast
contrast_list &lt;-  list(c(&#39;Test&#39;,&#39;Control&#39;))
make_contrast_name &lt;- function(contrast_parts){paste0(contrast_parts[1],&quot;vs&quot;,contrast_parts[1])}
contrast_names &lt;- sapply(FUN=make_contrast_name, X=contrast_list) 
names(contrast_list) &lt;- contrast_names



## Prepare empty lists to store results.
de_result_list     &lt;- list()
de_result_sig_list &lt;- list()

## Loop through each cluster / celltype
for (the_celltype in levels(so$cluster)) {
  
  ## Subset to the cluster/celltype
  print(the_celltype)
  so.this &lt;- subset(so, CelltypeCode == the_celltype)

  # Count how many cells within each sample
  # And list which samples have more than the minimum
  cells_per_sample &lt;- table(so.this$samplename)
  print(cells_per_sample)
  samples_with_enough_cells &lt;- names(cells_per_sample)[cells_per_sample &gt; min_cells_per_group]

  ## Loop through every contrast 
  # For each listed contrast, do both sides of the contrast have enough cells?
  # This is of course much simpler if you only have one contrast!
  for (contrast_name in names(contrast_list)) {
    
    # from contrast list pull out list of the two samples involved;
    # contrast_name is &#39;TestvsControl&#39;  
    # c(&#39;Test&#39;, &#39;Control&#39;)
    contrast &lt;- contrast_list[[contrast_name]]
    
    # Only run this contrast if both sides pass!
    if (all(contrast %in% samples_with_enough_cells)) {
      print(contrast_name) 

      # We need to tell Seurat to group by _sample_ not cluster.
      Idents(so.this) &lt;- so.this$samplename
  
      # Finally, calculate the DE!      
      de_result &lt;- FindMarkers(so.this, 
                               ident.1  = contrast[1], 
                               ident.2  = contrast[2],
                               slot     = &#39;data&#39;,
                               test.use = &quot;wilcox&quot;, # the default
                               min.pct  = 0.01,     # Note 
                               logfc.threshold     = 0,  # Report all, filter later 
                               max.cells.per.ident = 500 # Optional - if you have really big clusters, set this to subsample!
                               )
  
      
      # It can be helpful to know the average expression of a gene
      # This will give us the average (per cell) within this cluster.
      avg_expression &lt;- rowMeans(GetAssayData(so.this, assay = &#39;RNA&#39;, layer=&quot;data&quot;))
      
      # Format the results with extra info, record the contrast run.
      de_result.formatted &lt;- de_result %&gt;%
        rownames_to_column(&quot;target&quot;) %&gt;%
        mutate(contrast=contrast_name,
               celltype=the_celltype,
               avg_expression=avg_expression[target]) %&gt;%
        select(celltype,contrast,target,avg_expression, everything()) %&gt;%
        arrange(p_val)
  
      # Filter to just significant results, optionally by log2FC.
      de_result.sig &lt;- filter(de_result.formatted, 
                         p_val_adj       &lt;  p_value_threshold,
                         abs(avg_log2FC) &gt;= log_fc_threshold )

      # Record these results in a list 
      full_name &lt;- paste(contrast_name, the_celltype)
      de_result_list[[full_name]]     &lt;- de_result.formatted
      de_result_sig_list[[full_name]] &lt;- de_result.sig
      
    }
  }
}
# And combine all and significant results into two big tables.
de_results_all &lt;- bind_rows(de_result_list)
de_results_sig &lt;- bind_rows(de_result_sig_list)


# Save the full set of results as a tab-deliminated text file
# for processing, and excel files for browsing.
write_tsv(de_results_all,       &quot;~/myproject/de_results_all.tsv&quot;)
write_xlsx(de_result_list,      &quot;~/myproject/de_results_all.xlsx&quot;)
write_xlsx(de_result_sig_list,  &quot;~/myproject/de_results_sig.xlsx&quot;)</code></pre>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<pre class="r"><code>DT::datatable(de_result.sig)</code></pre>
<ul>
<li><strong>celltype</strong>: The cluster or celltype the contrat is
within. Added by running code.</li>
<li><strong>contrast</strong>: The name of what is being compared. Added
by running code.</li>
<li><strong>target</strong> : Gene name. (NB: Added by our running code,
in the direct output of FindMarkers(), these are row names instead)</li>
<li><strong>avg_expression</strong> : Average expression of target
within celltype group. (Added by running code)</li>
<li><strong>p_val</strong> : P value without multiple hypothesis
correction - see <em>p_val_adj</em></li>
<li><strong>avg_log2FC</strong> : Average log 2 fold change of the
comparison. Is calculated as: <em>log2(test expression) - log2(control
expression)</em> A value of 0 represents no change, +1 is a doubling,
and -1 is a halving of expression.</li>
<li><strong>pct.1</strong> : Percent of cells in group 1 (generally
test) that express the gene at all.</li>
<li><strong>pct.2</strong> : Percent of cells in group 2 (generally
control/WT/reference) that express the gene at all.</li>
<li><strong>p_val_adj</strong> : Pvalue with multiple hypothesis
correction. This may be used for filtering.</li>
</ul>
</div>
<div id="more-information" class="section level1">
<h1>More information</h1>
<p>List of useful resources. Papers, vignettes, pertinent forum
posts</p>
<ul>
<li><a href="d_Wnt4KO.html">Wnt4 KO in developing mouse kidney - 10X
Chromium scRNAseq</a>: <a
href="https://www.biorxiv.org/content/10.1101/2023.11.19.567482v1.full">Forming
nephrons promote nephron progenitor maintenance and branching
morphogenesis via paracrine BMP4 signalling under the control of
Wnt4</a> <span
class="citation">(<strong>moreauFormingNephronsPromote2023a?</strong>)</span>
: The paper with this data.</li>
<li><a
href="https://www.nature.com/articles/s41467-021-25960-2">Confronting
false discoveries in single-cell differential expression</a><span
class="citation">(<strong>squairConfrontingFalseDiscoveries2021?</strong>)</span>
: This is an excellent paper benchmarking and exploring differental
expression methods for single cell expression data.</li>
<li><a href="https://satijalab.org/seurat/articles/de_vignette">Seurat
Differential Expression Vignette</a> : How to do differential expression
with the seurat package. Tells you how to use the multiple statistical
tests that seurat offers.</li>
<li><a
href="https://bioconductor.org/books/3.19/OSCA.multisample/multi-sample-comparisons.html">OSCA
Differntial Expression</a>: The excellent book <a
href="https://bioconductor.org/books/release/OSCA/book-contents.html#multi-sample">Orchestrating
single cell analysis</a> includes a section on differential expression,
it focusses on pseuboulk approaches with replicates (which can’t be used
for a 1 vs 1) using the bioconductor toolkit, but provides useful
background.</li>
<li><a
href="https://swbioinf.github.io/scRNAseqInR_Doco/preprocessing.html#the-seurat-object">‘The
Seurat Object’ in scRNAseq analysis with R</a> Small section of this
document ‘the Seurat object’ has some explanation of the seurat object
format.</li>
</ul>
</div>
<div id="refereneces" class="section level1">
<h1>Refereneces</h1>
<!-- nothing here, autopopulated from bibliography: in json -->
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-moreau_forming_2023" class="csl-entry">
Moreau, Julie L. M., Sarah Williams, Jihane Homman-Ludiye, Andrew J.
Mallett, and Alexander N. Combes. 2023. <span>“Forming Nephrons Promote
Nephron Progenitor Maintenance and Branching Morphogenesis via Paracrine
<span>BMP4</span> Signalling Under the Control of <em>Wnt4</em>.”</span>
<a
href="https://doi.org/10.1101/2023.11.19.567482">https://doi.org/10.1101/2023.11.19.567482</a>.
</div>
<div id="ref-squair_confronting_2021" class="csl-entry">
Squair, Jordan W., Matthieu Gautier, Claudia Kathe, Mark A. Anderson,
Nicholas D. James, Thomas H. Hutson, Rémi Hudelle, et al. 2021.
<span>“Confronting False Discoveries in Single-Cell Differential
Expression.”</span> <em>Nature Communications</em> 12 (1): 5692. <a
href="https://doi.org/10.1038/s41467-021-25960-2">https://doi.org/10.1038/s41467-021-25960-2</a>.
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] ggrepel_0.9.5      lubridate_1.9.3    forcats_1.0.0      stringr_1.5.1     
 [5] dplyr_1.1.4        purrr_1.0.2        readr_2.1.5        tidyr_1.3.1       
 [9] tibble_3.2.1       ggplot2_3.5.0      tidyverse_2.0.0    DT_0.33           
[13] Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-3           workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.16.0      jsonlite_1.8.8        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   farver_2.1.1          
  [7] rmarkdown_2.26         fs_1.6.3               vctrs_0.6.5           
 [10] ROCR_1.0-11            spatstat.explore_3.2-7 htmltools_0.5.8.1     
 [13] sass_0.4.9             sctransform_0.4.1      parallelly_1.37.1     
 [16] KernSmooth_2.23-22     bslib_0.7.0            htmlwidgets_1.6.4     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         
 [22] zoo_1.8-12             cachem_1.0.8           whisker_0.4.1         
 [25] igraph_2.0.3           mime_0.12              lifecycle_1.0.4       
 [28] pkgconfig_2.0.3        Matrix_1.6-5           R6_2.5.1              
 [31] fastmap_1.1.1          fitdistrplus_1.1-11    future_1.33.2         
 [34] shiny_1.8.1.1          digest_0.6.35          colorspace_2.1-0      
 [37] patchwork_1.2.0        ps_1.7.6               rprojroot_2.0.4       
 [40] tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1         
 [43] crosstalk_1.2.1        labeling_0.4.3         progressr_0.14.0      
 [46] timechange_0.3.0       fansi_1.0.6            spatstat.sparse_3.0-3 
 [49] httr_1.4.7             polyclip_1.10-6        abind_1.4-5           
 [52] compiler_4.3.2         withr_3.0.0            fastDummies_1.7.3     
 [55] highr_0.10             MASS_7.3-60.0.1        tools_4.3.2           
 [58] lmtest_0.9-40          httpuv_1.6.15          future.apply_1.11.2   
 [61] goftest_1.2-3          glue_1.7.0             callr_3.7.6           
 [64] nlme_3.1-164           promises_1.2.1         grid_4.3.2            
 [67] Rtsne_0.17             getPass_0.2-4          cluster_2.1.6         
 [70] reshape2_1.4.4         generics_0.1.3         gtable_0.3.4          
 [73] spatstat.data_3.0-4    tzdb_0.4.0             hms_1.1.3             
 [76] data.table_1.15.4      utf8_1.2.4             spatstat.geom_3.2-9   
 [79] RcppAnnoy_0.0.22       RANN_2.6.1             pillar_1.9.0          
 [82] limma_3.58.1           spam_2.10-0            RcppHNSW_0.6.0        
 [85] later_1.3.2            splines_4.3.2          lattice_0.22-6        
 [88] renv_1.0.5             survival_3.5-8         deldir_2.0-4          
 [91] tidyselect_1.2.1       miniUI_0.1.1.1         pbapply_1.7-2         
 [94] knitr_1.45             git2r_0.33.0           gridExtra_2.3         
 [97] scattermore_1.2        xfun_0.43              statmod_1.5.0         
[100] matrixStats_1.2.0      stringi_1.8.3          lazyeval_0.2.2        
[103] yaml_2.3.8             evaluate_0.23          codetools_0.2-20      
[106] BiocManager_1.30.22    cli_3.6.2              uwot_0.1.16           
[109] xtable_1.8-4           reticulate_1.35.0      munsell_0.5.1         
[112] processx_3.8.4         jquerylib_0.1.4        Rcpp_1.0.12           
[115] globals_0.16.3         spatstat.random_3.2-3  png_0.1-8             
[118] parallel_4.3.2         presto_1.0.0           dotCall64_1.1-1       
[121] listenv_0.9.1          viridisLite_0.4.2      scales_1.3.0          
[124] ggridges_0.5.6         leiden_0.4.3.1         rlang_1.1.3           
[127] cowplot_1.1.3         </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
