<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Sarah Williams" />


<title>Differential expression between groups using pseudobulk</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/htmltools-fill-0.5.8/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.33/datatables.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">spatialsnippets</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tests</a>
</li>
<li>
  <a href="index_data.html">Data</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/swbioinf/spatialsnippets">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Differential expression between groups
using pseudobulk</h1>
<h4 class="author">Sarah Williams</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-05-22
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>spatialsnippets/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20231017code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20231017)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20231017code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20231017)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree36dd228c42b376d577f528d7fbb8c4021242a570targetblank36dd228a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/swbioinf/spatialsnippets/tree/36dd228c42b376d577f528d7fbb8c4021242a570" target="_blank">36dd228</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree36dd228c42b376d577f528d7fbb8c4021242a570targetblank36dd228a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/swbioinf/spatialsnippets/tree/36dd228c42b376d577f528d7fbb8c4021242a570" target="_blank">36dd228</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    renv/library/
    Ignored:    renv/staging/

Untracked files:
    Untracked:  code/generate_diagram_plots.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/e_DEPseudobulk_insitu.Rmd</code>)
and HTML (<code>docs/e_DEPseudobulk_insitu.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/36dd228c42b376d577f528d7fbb8c4021242a570/analysis/e_DEPseudobulk_insitu.Rmd" target="_blank">36dd228</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-22
</td>
<td>
wflow_publish("analysis/e_DEPseudobulk_insitu.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/5a9d7e9002faa026640cfe522a79bcf86b1ad8bf/docs/e_DEPseudobulk_insitu.html" target="_blank">5a9d7e9</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/c67ef8694b603f4410d6e1957dd9794c7ccfbb0f/analysis/e_DEPseudobulk_insitu.Rmd" target="_blank">c67ef86</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
<td>
wflow_publish("analysis/e_DEPseudobulk_insitu.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/c77f76cecb74337321d877773789d8703b9bdb89/docs/e_DEPseudobulk_insitu.html" target="_blank">c77f76c</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/7b0be9369f7a0cb0fcae8df07cef531fb2fbe732/analysis/e_DEPseudobulk_insitu.Rmd" target="_blank">7b0be93</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-07
</td>
<td>
wflow_publish(c("analysis/index.Rmd",
"analysis/e_DEPseudobulk_insitu.Rmd",
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<!-- keywords

insituspatial
seurat
de
test

-->
<div id="overview" class="section level1">
<h1>Overview</h1>
<p><img src="assets/insitu_spatial_pseudobulk_de.png" /></p>
<p>This requires:</p>
<ul>
<li>Biological replicates for each group</li>
<li>Assigned cell types</li>
<li>[Optionally] Multiple fovs measured per sample</li>
</ul>
</div>
<div id="worked-example" class="section level1">
<h1>Worked example</h1>
<p>How does gene expression change within each cell type between
Ulcerative colitis or Crohn’s disease, and Healthy controls?</p>
<div id="load-libraries-and-data" class="section level2">
<h2>Load libraries and data</h2>
<pre class="r"><code>library(Seurat)</code></pre>
<pre><code>Loading required package: SeuratObject</code></pre>
<pre><code>Loading required package: sp</code></pre>
<pre><code>&#39;SeuratObject&#39; was built under R 4.3.0 but the current version is
4.3.2; it is recomended that you reinstall &#39;SeuratObject&#39; as the ABI
for R may have changed</code></pre>
<pre><code>
Attaching package: &#39;SeuratObject&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    intersect</code></pre>
<pre class="r"><code>library(speckle)</code></pre>
<pre><code>Warning: replacing previous import &#39;S4Arrays::makeNindexFromArrayViewport&#39; by
&#39;DelayedArray::makeNindexFromArrayViewport&#39; when loading &#39;SummarizedExperiment&#39;</code></pre>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.0     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     </code></pre>
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<pre class="r"><code>library(limma)
library(DT)</code></pre>
<pre><code>
Attaching package: &#39;DT&#39;

The following object is masked from &#39;package:Seurat&#39;:

    JS

The following object is masked from &#39;package:SeuratObject&#39;:

    JS</code></pre>
<pre class="r"><code>library(edgeR)</code></pre>
<pre class="r"><code>data_dir              &lt;- file.path(&quot;~/projects/spatialsnippets/datasets/GSE234713_IBDcosmx_GarridoTrigo2023/processed_data&quot;) 
seurat_file_01_loaded &lt;- file.path(data_dir, &quot;GSE234713_CosMx_IBD_seurat_01_loaded.RDS&quot;)</code></pre>
<pre class="r"><code>so &lt;- readRDS(seurat_file_01_loaded)</code></pre>
</div>
<div id="experimental-design" class="section level2">
<h2>Experimental design</h2>
<!-- Move this to own page? --->
<p>There are three individuals per condition (one tissue sample from
each individual). With multiple fovs on each physical tissue sample.</p>
<pre class="r"><code>sample_table &lt;- select(as_tibble(so@meta.data), condition, individual_code, fov_name) %&gt;%
  unique() %&gt;% 
  group_by(condition, individual_code) %&gt;% 
  summarise(n_fovs= n(), item = str_c(fov_name, collapse = &quot;, &quot;))</code></pre>
<pre><code>`summarise()` has grouped output by &#39;condition&#39;. You can override using the
`.groups` argument.</code></pre>
<pre class="r"><code>DT::datatable(sample_table)</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-8839f08f82529a5247be" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8839f08f82529a5247be">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9"],["Crohn's disease","Crohn's disease","Crohn's disease","Healthy controls","Healthy controls","Healthy controls","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis"],["CD_a","CD_b","CD_c","HC_a","HC_b","HC_c","UC_a","UC_b","UC_c"],[19,19,16,19,20,16,19,22,21],["CD_a_001, CD_a_002, CD_a_003, CD_a_004, CD_a_005, CD_a_006, CD_a_007, CD_a_008, CD_a_009, CD_a_010, CD_a_011, CD_a_012, CD_a_013, CD_a_014, CD_a_015, CD_a_016, CD_a_017, CD_a_018, CD_a_019","CD_b_002, CD_b_003, CD_b_004, CD_b_005, CD_b_006, CD_b_007, CD_b_008, CD_b_009, CD_b_010, CD_b_011, CD_b_012, CD_b_013, CD_b_014, CD_b_015, CD_b_016, CD_b_017, CD_b_018, CD_b_019, CD_b_020","CD_c_001, CD_c_002, CD_c_003, CD_c_004, CD_c_005, CD_c_006, CD_c_007, CD_c_008, CD_c_009, CD_c_010, CD_c_011, CD_c_012, CD_c_013, CD_c_014, CD_c_015, CD_c_017","HC_a_001, HC_a_002, HC_a_003, HC_a_004, HC_a_005, HC_a_006, HC_a_007, HC_a_008, HC_a_009, HC_a_010, HC_a_011, HC_a_012, HC_a_013, HC_a_014, HC_a_015, HC_a_016, HC_a_018, HC_a_019, HC_a_020","HC_b_001, HC_b_002, HC_b_003, HC_b_004, HC_b_005, HC_b_006, HC_b_007, HC_b_008, HC_b_009, HC_b_010, HC_b_011, HC_b_012, HC_b_013, HC_b_014, HC_b_015, HC_b_016, HC_b_017, HC_b_018, HC_b_019, HC_b_020","HC_c_001, HC_c_002, HC_c_003, HC_c_004, HC_c_005, HC_c_006, HC_c_007, HC_c_008, HC_c_009, HC_c_010, HC_c_011, HC_c_012, HC_c_013, HC_c_014, HC_c_015, HC_c_016","UC_a_001, UC_a_002, UC_a_005, UC_a_006, UC_a_007, UC_a_008, UC_a_009, UC_a_010, UC_a_012, UC_a_013, UC_a_014, UC_a_015, UC_a_016, UC_a_017, UC_a_018, UC_a_019, UC_a_020, UC_a_021, UC_a_022","UC_b_001, UC_b_002, UC_b_003, UC_b_004, UC_b_005, UC_b_006, UC_b_007, UC_b_008, UC_b_009, UC_b_010, UC_b_011, UC_b_012, UC_b_014, UC_b_015, UC_b_016, UC_b_017, UC_b_018, UC_b_020, UC_b_021, UC_b_022, UC_b_023, UC_b_024","UC_c_001, UC_c_002, UC_c_003, UC_c_004, UC_c_005, UC_c_006, UC_c_007, UC_c_008, UC_c_009, UC_c_010, UC_c_011, UC_c_012, UC_c_013, UC_c_014, UC_c_015, UC_c_016, UC_c_017, UC_c_018, UC_c_019, UC_c_020, UC_c_021"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>condition<\/th>\n      <th>individual_code<\/th>\n      <th>n_fovs<\/th>\n      <th>item<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":3},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"condition","targets":1},{"name":"individual_code","targets":2},{"name":"n_fovs","targets":3},{"name":"item","targets":4}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="count-how-many-cells-of-each-type-in-your-data"
class="section level2">
<h2>Count how many cells of each type in your data</h2>
<p>Using a pseudobulk approach.</p>
<ul>
<li>Need at least x reads in a cell to include it</li>
<li>Need at least x cells of a celltype within an fov to include
that</li>
<li>Can only test where we have at least 2 samples on each side of a
contrast.</li>
</ul>
<pre class="r"><code>min_reads_per_cell &lt;- 200

ggplot(so@meta.data, aes(x=nCount_RNA)) +
  geom_density() +
  geom_vline(xintercept = min_reads_per_cell, lty=3) +
  scale_x_log10() +
  theme_bw()+
  ggtitle(&quot;How many reads per cell?&quot;)</code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/c77f76cecb74337321d877773789d8703b9bdb89/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-5-1.png" target="_blank">c77f76c</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>so&lt;- so[,so$nCount_RNA &gt;= min_reads_per_cell]</code></pre>
<p>We will pool each celltype within each fov (cluster_group). But there
needs to be a certain number of cells for that to work.</p>
<p>Note there are much fewer t-cells overall, but given that we have a
high number of samples, there should still be enough to include. Its
typical that some of the less common cell types are difficult or
impossible to reliably test.</p>
<pre class="r"><code>min_cells_per_fovcluster &lt;- 20

so$fov_cluster &lt;- paste0(so$fov_name,&quot;_&quot;, so$celltype_subset)

celltype_summary_table &lt;- so@meta.data %&gt;% 
  group_by(condition, group, individual_code, fov_name, celltype_subset, fov_cluster) %&gt;%
  summarise(cells=n(), .groups = &#39;drop&#39;)
DT::datatable(celltype_summary_table)</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-a4d51c6334d8450be088" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-a4d51c6334d8450be088">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304","305","306","307","308","309","310","311","312","313","314","315","316","317","318","319","320","321","322","323","324","325","326","327","328","329","330","331","332","333","334","335","336","337","338","339","340","341","342","343","344","345","346","347","348","349","350","351","352","353","354","355","356","357","358","359","360","361","362","363","364","365","366","367","368","369","370","371","372","373","374","375","376","377","378","379","380","381","382","383","384","385","386","387","388","389","390","391","392","393","394","395","396","397","398","399","400","401","402","403","404","405","406","407","408","409","410","411","412","413","414","415","416","417","418","419","420","421","422","423","424","425","426","427","428","429","430","431","432","433","434","435","436","437","438","439","440","441","442","443","444","445","446","447","448","449","450","451","452","453","454","455","456","457","458","459","460","461","462","463","464","465","466","467","468","469","470","471","472","473","474","475","476","477","478","479","480","481","482","483","484","485","486","487","488","489","490","491","492","493","494","495","496","497","498","499","500","501","502","503","504","505","506","507","508","509","510","511","512","513","514","515","516","517","518","519","520","521","522","523","524","525","526","527","528","529","530","531","532","533","534","535","536","537","538","539","540","541","542","543","544","545","546","547","548","549","550","551","552","553","554","555","556","557","558","559","560","561","562","563","564","565","566","567","568","569","570","571","572","573","574","575","576","577","578","579","580","581","582","583","584","585","586","587","588","589","590","591","592","593","594","595","596","597","598","599","600","601","602","603","604","605","606","607","608","609","610","611","612","613","614","615","616","617","618","619","620","621","622","623","624","625","626","627","628","629","630","631","632","633","634","635","636","637","638","639","640","641","642","643","644","645","646","647","648","649","650","651","652","653","654","655","656","657","658","659","660","661","662","663","664","665","666","667","668","669","670","671","672","673","674","675","676","677","678","679","680","681","682","683","684","685","686","687","688","689","690","691","692","693","694","695","696","697","698","699","700","701","702","703","704","705","706","707","708","709","710","711","712","713","714","715","716","717","718","719","720","721","722","723","724","725","726","727","728","729","730","731","732","733","734","735","736","737","738","739","740","741","742","743","744","745","746","747","748","749","750","751","752","753","754","755","756","757","758","759","760","761","762","763","764","765","766","767","768","769","770","771","772","773","774","775","776","777","778","779","780","781","782","783","784","785","786","787","788","789","790","791"],["Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Crohn's disease","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Healthy controls","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis","Ulcerative colitis"],["CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","CD","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","HC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC","UC"],["CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_a","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_b","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","CD_c","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_a","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_b","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","HC_c","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_a","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_b","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c","UC_c"],["CD_a_001","CD_a_001","CD_a_001","CD_a_001","CD_a_001","CD_a_002","CD_a_002","CD_a_002","CD_a_002","CD_a_002","CD_a_003","CD_a_003","CD_a_003","CD_a_003","CD_a_004","CD_a_004","CD_a_004","CD_a_004","CD_a_004","CD_a_005","CD_a_005","CD_a_005","CD_a_005","CD_a_006","CD_a_006","CD_a_006","CD_a_006","CD_a_006","CD_a_007","CD_a_007","CD_a_007","CD_a_007","CD_a_007","CD_a_008","CD_a_008","CD_a_008","CD_a_008","CD_a_008","CD_a_009","CD_a_009","CD_a_009","CD_a_009","CD_a_010","CD_a_010","CD_a_010","CD_a_010","CD_a_011","CD_a_011","CD_a_011","CD_a_011","CD_a_011","CD_a_012","CD_a_012","CD_a_012","CD_a_012","CD_a_013","CD_a_013","CD_a_013","CD_a_013","CD_a_013","CD_a_014","CD_a_014","CD_a_014","CD_a_014","CD_a_014","CD_a_015","CD_a_015","CD_a_015","CD_a_015","CD_a_015","CD_a_016","CD_a_016","CD_a_016","CD_a_017","CD_a_017","CD_a_017","CD_a_017","CD_a_017","CD_a_018","CD_a_018","CD_a_018","CD_a_018","CD_a_018","CD_a_019","CD_a_019","CD_a_019","CD_a_019","CD_b_002","CD_b_002","CD_b_002","CD_b_002","CD_b_003","CD_b_003","CD_b_003","CD_b_003","CD_b_004","CD_b_004","CD_b_004","CD_b_004","CD_b_005","CD_b_005","CD_b_005","CD_b_005","CD_b_006","CD_b_006","CD_b_006","CD_b_006","CD_b_007","CD_b_007","CD_b_007","CD_b_007","CD_b_008","CD_b_008","CD_b_008","CD_b_008","CD_b_009","CD_b_009","CD_b_009","CD_b_009","CD_b_009","CD_b_010","CD_b_010","CD_b_010","CD_b_010","CD_b_011","CD_b_011","CD_b_011","CD_b_011","CD_b_011","CD_b_012","CD_b_012","CD_b_012","CD_b_012","CD_b_012","CD_b_013","CD_b_013","CD_b_013","CD_b_013","CD_b_013","CD_b_014","CD_b_014","CD_b_014","CD_b_014","CD_b_014","CD_b_015","CD_b_015","CD_b_015","CD_b_015","CD_b_015","CD_b_016","CD_b_016","CD_b_016","CD_b_016","CD_b_017","CD_b_017","CD_b_017","CD_b_017","CD_b_018","CD_b_018","CD_b_018","CD_b_018","CD_b_019","CD_b_019","CD_b_019","CD_b_019","CD_b_020","CD_b_020","CD_b_020","CD_b_020","CD_c_001","CD_c_001","CD_c_001","CD_c_001","CD_c_001","CD_c_002","CD_c_002","CD_c_002","CD_c_002","CD_c_002","CD_c_003","CD_c_003","CD_c_003","CD_c_003","CD_c_003","CD_c_004","CD_c_004","CD_c_004","CD_c_004","CD_c_004","CD_c_005","CD_c_005","CD_c_005","CD_c_005","CD_c_005","CD_c_006","CD_c_006","CD_c_006","CD_c_006","CD_c_006","CD_c_007","CD_c_007","CD_c_007","CD_c_007","CD_c_007","CD_c_008","CD_c_008","CD_c_008","CD_c_008","CD_c_008","CD_c_009","CD_c_009","CD_c_009","CD_c_009","CD_c_009","CD_c_010","CD_c_010","CD_c_010","CD_c_010","CD_c_010","CD_c_011","CD_c_011","CD_c_011","CD_c_011","CD_c_011","CD_c_012","CD_c_012","CD_c_012","CD_c_012","CD_c_013","CD_c_013","CD_c_013","CD_c_013","CD_c_013","CD_c_014","CD_c_014","CD_c_014","CD_c_014","CD_c_014","CD_c_015","CD_c_015","CD_c_015","CD_c_015","CD_c_017","CD_c_017","CD_c_017","CD_c_017","CD_c_017","HC_a_001","HC_a_001","HC_a_001","HC_a_001","HC_a_001","HC_a_002","HC_a_002","HC_a_002","HC_a_002","HC_a_002","HC_a_003","HC_a_003","HC_a_003","HC_a_003","HC_a_003","HC_a_004","HC_a_004","HC_a_004","HC_a_004","HC_a_004","HC_a_005","HC_a_005","HC_a_005","HC_a_005","HC_a_005","HC_a_006","HC_a_006","HC_a_006","HC_a_006","HC_a_006","HC_a_007","HC_a_007","HC_a_007","HC_a_007","HC_a_007","HC_a_008","HC_a_008","HC_a_008","HC_a_008","HC_a_008","HC_a_009","HC_a_009","HC_a_009","HC_a_009","HC_a_009","HC_a_010","HC_a_010","HC_a_010","HC_a_010","HC_a_010","HC_a_011","HC_a_011","HC_a_011","HC_a_011","HC_a_011","HC_a_012","HC_a_012","HC_a_012","HC_a_012","HC_a_012","HC_a_013","HC_a_013","HC_a_013","HC_a_013","HC_a_013","HC_a_014","HC_a_014","HC_a_014","HC_a_014","HC_a_014","HC_a_015","HC_a_015","HC_a_015","HC_a_015","HC_a_015","HC_a_016","HC_a_016","HC_a_016","HC_a_016","HC_a_016","HC_a_018","HC_a_018","HC_a_018","HC_a_018","HC_a_018","HC_a_019","HC_a_019","HC_a_019","HC_a_019","HC_a_019","HC_a_020","HC_a_020","HC_a_020","HC_a_020","HC_a_020","HC_b_001","HC_b_001","HC_b_001","HC_b_001","HC_b_001","HC_b_002","HC_b_002","HC_b_002","HC_b_002","HC_b_002","HC_b_003","HC_b_003","HC_b_003","HC_b_003","HC_b_003","HC_b_004","HC_b_004","HC_b_004","HC_b_004","HC_b_004","HC_b_005","HC_b_005","HC_b_005","HC_b_005","HC_b_005","HC_b_006","HC_b_006","HC_b_006","HC_b_006","HC_b_006","HC_b_007","HC_b_007","HC_b_007","HC_b_007","HC_b_007","HC_b_008","HC_b_008","HC_b_008","HC_b_008","HC_b_008","HC_b_009","HC_b_009","HC_b_009","HC_b_009","HC_b_009","HC_b_010","HC_b_010","HC_b_010","HC_b_010","HC_b_010","HC_b_011","HC_b_011","HC_b_011","HC_b_011","HC_b_011","HC_b_012","HC_b_012","HC_b_012","HC_b_012","HC_b_012","HC_b_013","HC_b_013","HC_b_013","HC_b_013","HC_b_013","HC_b_014","HC_b_014","HC_b_014","HC_b_014","HC_b_014","HC_b_015","HC_b_015","HC_b_015","HC_b_015","HC_b_015","HC_b_016","HC_b_016","HC_b_016","HC_b_016","HC_b_016","HC_b_017","HC_b_017","HC_b_017","HC_b_017","HC_b_018","HC_b_018","HC_b_018","HC_b_019","HC_b_019","HC_b_019","HC_b_020","HC_b_020","HC_b_020","HC_b_020","HC_b_020","HC_c_001","HC_c_001","HC_c_001","HC_c_001","HC_c_001","HC_c_002","HC_c_002","HC_c_002","HC_c_002","HC_c_002","HC_c_003","HC_c_003","HC_c_003","HC_c_003","HC_c_004","HC_c_004","HC_c_004","HC_c_004","HC_c_005","HC_c_005","HC_c_005","HC_c_005","HC_c_006","HC_c_006","HC_c_006","HC_c_007","HC_c_007","HC_c_007","HC_c_007","HC_c_008","HC_c_008","HC_c_008","HC_c_008","HC_c_008","HC_c_009","HC_c_009","HC_c_009","HC_c_009","HC_c_010","HC_c_010","HC_c_011","HC_c_011","HC_c_011","HC_c_011","HC_c_012","HC_c_012","HC_c_012","HC_c_013","HC_c_013","HC_c_013","HC_c_013","HC_c_013","HC_c_014","HC_c_014","HC_c_014","HC_c_014","HC_c_014","HC_c_015","HC_c_015","HC_c_015","HC_c_015","HC_c_016","HC_c_016","HC_c_016","HC_c_016","UC_a_001","UC_a_001","UC_a_001","UC_a_001","UC_a_001","UC_a_002","UC_a_002","UC_a_002","UC_a_002","UC_a_002","UC_a_005","UC_a_005","UC_a_005","UC_a_005","UC_a_005","UC_a_006","UC_a_006","UC_a_006","UC_a_006","UC_a_006","UC_a_007","UC_a_007","UC_a_007","UC_a_007","UC_a_007","UC_a_008","UC_a_008","UC_a_008","UC_a_008","UC_a_008","UC_a_009","UC_a_009","UC_a_009","UC_a_009","UC_a_010","UC_a_010","UC_a_010","UC_a_010","UC_a_010","UC_a_012","UC_a_012","UC_a_012","UC_a_012","UC_a_012","UC_a_013","UC_a_013","UC_a_013","UC_a_013","UC_a_014","UC_a_014","UC_a_014","UC_a_014","UC_a_015","UC_a_015","UC_a_015","UC_a_015","UC_a_016","UC_a_016","UC_a_016","UC_a_017","UC_a_017","UC_a_017","UC_a_017","UC_a_018","UC_a_018","UC_a_018","UC_a_018","UC_a_018","UC_a_019","UC_a_019","UC_a_019","UC_a_019","UC_a_020","UC_a_020","UC_a_020","UC_a_020","UC_a_020","UC_a_021","UC_a_021","UC_a_021","UC_a_021","UC_a_022","UC_a_022","UC_a_022","UC_a_022","UC_b_001","UC_b_001","UC_b_001","UC_b_001","UC_b_001","UC_b_002","UC_b_002","UC_b_002","UC_b_002","UC_b_002","UC_b_003","UC_b_003","UC_b_003","UC_b_003","UC_b_003","UC_b_004","UC_b_004","UC_b_004","UC_b_004","UC_b_004","UC_b_005","UC_b_005","UC_b_005","UC_b_005","UC_b_006","UC_b_006","UC_b_006","UC_b_006","UC_b_006","UC_b_007","UC_b_007","UC_b_007","UC_b_007","UC_b_007","UC_b_008","UC_b_008","UC_b_008","UC_b_008","UC_b_008","UC_b_009","UC_b_009","UC_b_009","UC_b_009","UC_b_010","UC_b_010","UC_b_010","UC_b_010","UC_b_011","UC_b_011","UC_b_011","UC_b_011","UC_b_012","UC_b_012","UC_b_012","UC_b_012","UC_b_014","UC_b_014","UC_b_014","UC_b_014","UC_b_014","UC_b_015","UC_b_015","UC_b_015","UC_b_015","UC_b_015","UC_b_016","UC_b_016","UC_b_016","UC_b_016","UC_b_016","UC_b_017","UC_b_017","UC_b_017","UC_b_017","UC_b_017","UC_b_018","UC_b_018","UC_b_018","UC_b_018","UC_b_018","UC_b_020","UC_b_020","UC_b_020","UC_b_020","UC_b_021","UC_b_021","UC_b_021","UC_b_021","UC_b_021","UC_b_022","UC_b_022","UC_b_022","UC_b_022","UC_b_022","UC_b_023","UC_b_023","UC_b_023","UC_b_023","UC_b_024","UC_b_024","UC_b_024","UC_b_024","UC_c_001","UC_c_001","UC_c_001","UC_c_001","UC_c_001","UC_c_002","UC_c_002","UC_c_002","UC_c_002","UC_c_002","UC_c_003","UC_c_003","UC_c_003","UC_c_003","UC_c_003","UC_c_004","UC_c_004","UC_c_004","UC_c_005","UC_c_005","UC_c_005","UC_c_005","UC_c_006","UC_c_006","UC_c_006","UC_c_006","UC_c_006","UC_c_007","UC_c_007","UC_c_007","UC_c_007","UC_c_007","UC_c_008","UC_c_008","UC_c_008","UC_c_008","UC_c_008","UC_c_009","UC_c_009","UC_c_009","UC_c_009","UC_c_009","UC_c_010","UC_c_010","UC_c_010","UC_c_010","UC_c_010","UC_c_011","UC_c_011","UC_c_011","UC_c_011","UC_c_011","UC_c_012","UC_c_012","UC_c_012","UC_c_012","UC_c_012","UC_c_013","UC_c_013","UC_c_013","UC_c_013","UC_c_013","UC_c_014","UC_c_014","UC_c_014","UC_c_014","UC_c_014","UC_c_015","UC_c_015","UC_c_015","UC_c_015","UC_c_015","UC_c_016","UC_c_016","UC_c_016","UC_c_016","UC_c_016","UC_c_017","UC_c_017","UC_c_017","UC_c_017","UC_c_017","UC_c_018","UC_c_018","UC_c_018","UC_c_018","UC_c_018","UC_c_019","UC_c_019","UC_c_019","UC_c_019","UC_c_019","UC_c_020","UC_c_020","UC_c_020","UC_c_020","UC_c_020","UC_c_021","UC_c_021","UC_c_021","UC_c_021","UC_c_021"],["epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","myeloids","plasmas","stroma","myeloids","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","epi","plasmas","stroma","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","myeloids","stroma","epi","myeloids","plasmas","stroma","epi","myeloids","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","stroma","epi","myeloids","plasmas","stroma","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells","epi","myeloids","plasmas","stroma","tcells"],["CD_a_001_epi","CD_a_001_myeloids","CD_a_001_plasmas","CD_a_001_stroma","CD_a_001_tcells","CD_a_002_epi","CD_a_002_myeloids","CD_a_002_plasmas","CD_a_002_stroma","CD_a_002_tcells","CD_a_003_myeloids","CD_a_003_plasmas","CD_a_003_stroma","CD_a_003_tcells","CD_a_004_epi","CD_a_004_myeloids","CD_a_004_plasmas","CD_a_004_stroma","CD_a_004_tcells","CD_a_005_myeloids","CD_a_005_plasmas","CD_a_005_stroma","CD_a_005_tcells","CD_a_006_epi","CD_a_006_myeloids","CD_a_006_plasmas","CD_a_006_stroma","CD_a_006_tcells","CD_a_007_epi","CD_a_007_myeloids","CD_a_007_plasmas","CD_a_007_stroma","CD_a_007_tcells","CD_a_008_epi","CD_a_008_myeloids","CD_a_008_plasmas","CD_a_008_stroma","CD_a_008_tcells","CD_a_009_epi","CD_a_009_myeloids","CD_a_009_plasmas","CD_a_009_stroma","CD_a_010_epi","CD_a_010_myeloids","CD_a_010_plasmas","CD_a_010_stroma","CD_a_011_epi","CD_a_011_myeloids","CD_a_011_plasmas","CD_a_011_stroma","CD_a_011_tcells","CD_a_012_epi","CD_a_012_myeloids","CD_a_012_plasmas","CD_a_012_stroma","CD_a_013_epi","CD_a_013_myeloids","CD_a_013_plasmas","CD_a_013_stroma","CD_a_013_tcells","CD_a_014_epi","CD_a_014_myeloids","CD_a_014_plasmas","CD_a_014_stroma","CD_a_014_tcells","CD_a_015_epi","CD_a_015_myeloids","CD_a_015_plasmas","CD_a_015_stroma","CD_a_015_tcells","CD_a_016_myeloids","CD_a_016_plasmas","CD_a_016_stroma","CD_a_017_epi","CD_a_017_myeloids","CD_a_017_plasmas","CD_a_017_stroma","CD_a_017_tcells","CD_a_018_epi","CD_a_018_myeloids","CD_a_018_plasmas","CD_a_018_stroma","CD_a_018_tcells","CD_a_019_epi","CD_a_019_myeloids","CD_a_019_plasmas","CD_a_019_stroma","CD_b_002_myeloids","CD_b_002_plasmas","CD_b_002_stroma","CD_b_002_tcells","CD_b_003_myeloids","CD_b_003_plasmas","CD_b_003_stroma","CD_b_003_tcells","CD_b_004_myeloids","CD_b_004_plasmas","CD_b_004_stroma","CD_b_004_tcells","CD_b_005_myeloids","CD_b_005_plasmas","CD_b_005_stroma","CD_b_005_tcells","CD_b_006_myeloids","CD_b_006_plasmas","CD_b_006_stroma","CD_b_006_tcells","CD_b_007_myeloids","CD_b_007_plasmas","CD_b_007_stroma","CD_b_007_tcells","CD_b_008_myeloids","CD_b_008_plasmas","CD_b_008_stroma","CD_b_008_tcells","CD_b_009_epi","CD_b_009_myeloids","CD_b_009_plasmas","CD_b_009_stroma","CD_b_009_tcells","CD_b_010_myeloids","CD_b_010_plasmas","CD_b_010_stroma","CD_b_010_tcells","CD_b_011_epi","CD_b_011_myeloids","CD_b_011_plasmas","CD_b_011_stroma","CD_b_011_tcells","CD_b_012_epi","CD_b_012_myeloids","CD_b_012_plasmas","CD_b_012_stroma","CD_b_012_tcells","CD_b_013_epi","CD_b_013_myeloids","CD_b_013_plasmas","CD_b_013_stroma","CD_b_013_tcells","CD_b_014_epi","CD_b_014_myeloids","CD_b_014_plasmas","CD_b_014_stroma","CD_b_014_tcells","CD_b_015_epi","CD_b_015_myeloids","CD_b_015_plasmas","CD_b_015_stroma","CD_b_015_tcells","CD_b_016_myeloids","CD_b_016_plasmas","CD_b_016_stroma","CD_b_016_tcells","CD_b_017_myeloids","CD_b_017_plasmas","CD_b_017_stroma","CD_b_017_tcells","CD_b_018_myeloids","CD_b_018_plasmas","CD_b_018_stroma","CD_b_018_tcells","CD_b_019_myeloids","CD_b_019_plasmas","CD_b_019_stroma","CD_b_019_tcells","CD_b_020_myeloids","CD_b_020_plasmas","CD_b_020_stroma","CD_b_020_tcells","CD_c_001_epi","CD_c_001_myeloids","CD_c_001_plasmas","CD_c_001_stroma","CD_c_001_tcells","CD_c_002_epi","CD_c_002_myeloids","CD_c_002_plasmas","CD_c_002_stroma","CD_c_002_tcells","CD_c_003_epi","CD_c_003_myeloids","CD_c_003_plasmas","CD_c_003_stroma","CD_c_003_tcells","CD_c_004_epi","CD_c_004_myeloids","CD_c_004_plasmas","CD_c_004_stroma","CD_c_004_tcells","CD_c_005_epi","CD_c_005_myeloids","CD_c_005_plasmas","CD_c_005_stroma","CD_c_005_tcells","CD_c_006_epi","CD_c_006_myeloids","CD_c_006_plasmas","CD_c_006_stroma","CD_c_006_tcells","CD_c_007_epi","CD_c_007_myeloids","CD_c_007_plasmas","CD_c_007_stroma","CD_c_007_tcells","CD_c_008_epi","CD_c_008_myeloids","CD_c_008_plasmas","CD_c_008_stroma","CD_c_008_tcells","CD_c_009_epi","CD_c_009_myeloids","CD_c_009_plasmas","CD_c_009_stroma","CD_c_009_tcells","CD_c_010_epi","CD_c_010_myeloids","CD_c_010_plasmas","CD_c_010_stroma","CD_c_010_tcells","CD_c_011_epi","CD_c_011_myeloids","CD_c_011_plasmas","CD_c_011_stroma","CD_c_011_tcells","CD_c_012_epi","CD_c_012_myeloids","CD_c_012_plasmas","CD_c_012_stroma","CD_c_013_epi","CD_c_013_myeloids","CD_c_013_plasmas","CD_c_013_stroma","CD_c_013_tcells","CD_c_014_epi","CD_c_014_myeloids","CD_c_014_plasmas","CD_c_014_stroma","CD_c_014_tcells","CD_c_015_epi","CD_c_015_myeloids","CD_c_015_plasmas","CD_c_015_stroma","CD_c_017_epi","CD_c_017_myeloids","CD_c_017_plasmas","CD_c_017_stroma","CD_c_017_tcells","HC_a_001_epi","HC_a_001_myeloids","HC_a_001_plasmas","HC_a_001_stroma","HC_a_001_tcells","HC_a_002_epi","HC_a_002_myeloids","HC_a_002_plasmas","HC_a_002_stroma","HC_a_002_tcells","HC_a_003_epi","HC_a_003_myeloids","HC_a_003_plasmas","HC_a_003_stroma","HC_a_003_tcells","HC_a_004_epi","HC_a_004_myeloids","HC_a_004_plasmas","HC_a_004_stroma","HC_a_004_tcells","HC_a_005_epi","HC_a_005_myeloids","HC_a_005_plasmas","HC_a_005_stroma","HC_a_005_tcells","HC_a_006_epi","HC_a_006_myeloids","HC_a_006_plasmas","HC_a_006_stroma","HC_a_006_tcells","HC_a_007_epi","HC_a_007_myeloids","HC_a_007_plasmas","HC_a_007_stroma","HC_a_007_tcells","HC_a_008_epi","HC_a_008_myeloids","HC_a_008_plasmas","HC_a_008_stroma","HC_a_008_tcells","HC_a_009_epi","HC_a_009_myeloids","HC_a_009_plasmas","HC_a_009_stroma","HC_a_009_tcells","HC_a_010_epi","HC_a_010_myeloids","HC_a_010_plasmas","HC_a_010_stroma","HC_a_010_tcells","HC_a_011_epi","HC_a_011_myeloids","HC_a_011_plasmas","HC_a_011_stroma","HC_a_011_tcells","HC_a_012_epi","HC_a_012_myeloids","HC_a_012_plasmas","HC_a_012_stroma","HC_a_012_tcells","HC_a_013_epi","HC_a_013_myeloids","HC_a_013_plasmas","HC_a_013_stroma","HC_a_013_tcells","HC_a_014_epi","HC_a_014_myeloids","HC_a_014_plasmas","HC_a_014_stroma","HC_a_014_tcells","HC_a_015_epi","HC_a_015_myeloids","HC_a_015_plasmas","HC_a_015_stroma","HC_a_015_tcells","HC_a_016_epi","HC_a_016_myeloids","HC_a_016_plasmas","HC_a_016_stroma","HC_a_016_tcells","HC_a_018_epi","HC_a_018_myeloids","HC_a_018_plasmas","HC_a_018_stroma","HC_a_018_tcells","HC_a_019_epi","HC_a_019_myeloids","HC_a_019_plasmas","HC_a_019_stroma","HC_a_019_tcells","HC_a_020_epi","HC_a_020_myeloids","HC_a_020_plasmas","HC_a_020_stroma","HC_a_020_tcells","HC_b_001_epi","HC_b_001_myeloids","HC_b_001_plasmas","HC_b_001_stroma","HC_b_001_tcells","HC_b_002_epi","HC_b_002_myeloids","HC_b_002_plasmas","HC_b_002_stroma","HC_b_002_tcells","HC_b_003_epi","HC_b_003_myeloids","HC_b_003_plasmas","HC_b_003_stroma","HC_b_003_tcells","HC_b_004_epi","HC_b_004_myeloids","HC_b_004_plasmas","HC_b_004_stroma","HC_b_004_tcells","HC_b_005_epi","HC_b_005_myeloids","HC_b_005_plasmas","HC_b_005_stroma","HC_b_005_tcells","HC_b_006_epi","HC_b_006_myeloids","HC_b_006_plasmas","HC_b_006_stroma","HC_b_006_tcells","HC_b_007_epi","HC_b_007_myeloids","HC_b_007_plasmas","HC_b_007_stroma","HC_b_007_tcells","HC_b_008_epi","HC_b_008_myeloids","HC_b_008_plasmas","HC_b_008_stroma","HC_b_008_tcells","HC_b_009_epi","HC_b_009_myeloids","HC_b_009_plasmas","HC_b_009_stroma","HC_b_009_tcells","HC_b_010_epi","HC_b_010_myeloids","HC_b_010_plasmas","HC_b_010_stroma","HC_b_010_tcells","HC_b_011_epi","HC_b_011_myeloids","HC_b_011_plasmas","HC_b_011_stroma","HC_b_011_tcells","HC_b_012_epi","HC_b_012_myeloids","HC_b_012_plasmas","HC_b_012_stroma","HC_b_012_tcells","HC_b_013_epi","HC_b_013_myeloids","HC_b_013_plasmas","HC_b_013_stroma","HC_b_013_tcells","HC_b_014_epi","HC_b_014_myeloids","HC_b_014_plasmas","HC_b_014_stroma","HC_b_014_tcells","HC_b_015_epi","HC_b_015_myeloids","HC_b_015_plasmas","HC_b_015_stroma","HC_b_015_tcells","HC_b_016_epi","HC_b_016_myeloids","HC_b_016_plasmas","HC_b_016_stroma","HC_b_016_tcells","HC_b_017_epi","HC_b_017_myeloids","HC_b_017_plasmas","HC_b_017_stroma","HC_b_018_myeloids","HC_b_018_plasmas","HC_b_018_stroma","HC_b_019_myeloids","HC_b_019_stroma","HC_b_019_tcells","HC_b_020_epi","HC_b_020_myeloids","HC_b_020_plasmas","HC_b_020_stroma","HC_b_020_tcells","HC_c_001_epi","HC_c_001_myeloids","HC_c_001_plasmas","HC_c_001_stroma","HC_c_001_tcells","HC_c_002_epi","HC_c_002_myeloids","HC_c_002_plasmas","HC_c_002_stroma","HC_c_002_tcells","HC_c_003_epi","HC_c_003_myeloids","HC_c_003_plasmas","HC_c_003_stroma","HC_c_004_epi","HC_c_004_myeloids","HC_c_004_plasmas","HC_c_004_stroma","HC_c_005_epi","HC_c_005_myeloids","HC_c_005_plasmas","HC_c_005_stroma","HC_c_006_epi","HC_c_006_plasmas","HC_c_006_stroma","HC_c_007_epi","HC_c_007_myeloids","HC_c_007_plasmas","HC_c_007_stroma","HC_c_008_epi","HC_c_008_myeloids","HC_c_008_plasmas","HC_c_008_stroma","HC_c_008_tcells","HC_c_009_epi","HC_c_009_myeloids","HC_c_009_plasmas","HC_c_009_stroma","HC_c_010_myeloids","HC_c_010_stroma","HC_c_011_epi","HC_c_011_myeloids","HC_c_011_plasmas","HC_c_011_stroma","HC_c_012_epi","HC_c_012_myeloids","HC_c_012_stroma","HC_c_013_epi","HC_c_013_myeloids","HC_c_013_plasmas","HC_c_013_stroma","HC_c_013_tcells","HC_c_014_epi","HC_c_014_myeloids","HC_c_014_plasmas","HC_c_014_stroma","HC_c_014_tcells","HC_c_015_epi","HC_c_015_myeloids","HC_c_015_plasmas","HC_c_015_stroma","HC_c_016_epi","HC_c_016_myeloids","HC_c_016_plasmas","HC_c_016_stroma","UC_a_001_epi","UC_a_001_myeloids","UC_a_001_plasmas","UC_a_001_stroma","UC_a_001_tcells","UC_a_002_epi","UC_a_002_myeloids","UC_a_002_plasmas","UC_a_002_stroma","UC_a_002_tcells","UC_a_005_epi","UC_a_005_myeloids","UC_a_005_plasmas","UC_a_005_stroma","UC_a_005_tcells","UC_a_006_epi","UC_a_006_myeloids","UC_a_006_plasmas","UC_a_006_stroma","UC_a_006_tcells","UC_a_007_epi","UC_a_007_myeloids","UC_a_007_plasmas","UC_a_007_stroma","UC_a_007_tcells","UC_a_008_epi","UC_a_008_myeloids","UC_a_008_plasmas","UC_a_008_stroma","UC_a_008_tcells","UC_a_009_myeloids","UC_a_009_plasmas","UC_a_009_stroma","UC_a_009_tcells","UC_a_010_epi","UC_a_010_myeloids","UC_a_010_plasmas","UC_a_010_stroma","UC_a_010_tcells","UC_a_012_epi","UC_a_012_myeloids","UC_a_012_plasmas","UC_a_012_stroma","UC_a_012_tcells","UC_a_013_myeloids","UC_a_013_plasmas","UC_a_013_stroma","UC_a_013_tcells","UC_a_014_myeloids","UC_a_014_plasmas","UC_a_014_stroma","UC_a_014_tcells","UC_a_015_myeloids","UC_a_015_plasmas","UC_a_015_stroma","UC_a_015_tcells","UC_a_016_myeloids","UC_a_016_plasmas","UC_a_016_stroma","UC_a_017_myeloids","UC_a_017_plasmas","UC_a_017_stroma","UC_a_017_tcells","UC_a_018_epi","UC_a_018_myeloids","UC_a_018_plasmas","UC_a_018_stroma","UC_a_018_tcells","UC_a_019_myeloids","UC_a_019_plasmas","UC_a_019_stroma","UC_a_019_tcells","UC_a_020_epi","UC_a_020_myeloids","UC_a_020_plasmas","UC_a_020_stroma","UC_a_020_tcells","UC_a_021_myeloids","UC_a_021_plasmas","UC_a_021_stroma","UC_a_021_tcells","UC_a_022_myeloids","UC_a_022_plasmas","UC_a_022_stroma","UC_a_022_tcells","UC_b_001_epi","UC_b_001_myeloids","UC_b_001_plasmas","UC_b_001_stroma","UC_b_001_tcells","UC_b_002_epi","UC_b_002_myeloids","UC_b_002_plasmas","UC_b_002_stroma","UC_b_002_tcells","UC_b_003_epi","UC_b_003_myeloids","UC_b_003_plasmas","UC_b_003_stroma","UC_b_003_tcells","UC_b_004_epi","UC_b_004_myeloids","UC_b_004_plasmas","UC_b_004_stroma","UC_b_004_tcells","UC_b_005_myeloids","UC_b_005_plasmas","UC_b_005_stroma","UC_b_005_tcells","UC_b_006_epi","UC_b_006_myeloids","UC_b_006_plasmas","UC_b_006_stroma","UC_b_006_tcells","UC_b_007_epi","UC_b_007_myeloids","UC_b_007_plasmas","UC_b_007_stroma","UC_b_007_tcells","UC_b_008_epi","UC_b_008_myeloids","UC_b_008_plasmas","UC_b_008_stroma","UC_b_008_tcells","UC_b_009_myeloids","UC_b_009_plasmas","UC_b_009_stroma","UC_b_009_tcells","UC_b_010_myeloids","UC_b_010_plasmas","UC_b_010_stroma","UC_b_010_tcells","UC_b_011_myeloids","UC_b_011_plasmas","UC_b_011_stroma","UC_b_011_tcells","UC_b_012_myeloids","UC_b_012_plasmas","UC_b_012_stroma","UC_b_012_tcells","UC_b_014_epi","UC_b_014_myeloids","UC_b_014_plasmas","UC_b_014_stroma","UC_b_014_tcells","UC_b_015_epi","UC_b_015_myeloids","UC_b_015_plasmas","UC_b_015_stroma","UC_b_015_tcells","UC_b_016_epi","UC_b_016_myeloids","UC_b_016_plasmas","UC_b_016_stroma","UC_b_016_tcells","UC_b_017_epi","UC_b_017_myeloids","UC_b_017_plasmas","UC_b_017_stroma","UC_b_017_tcells","UC_b_018_epi","UC_b_018_myeloids","UC_b_018_plasmas","UC_b_018_stroma","UC_b_018_tcells","UC_b_020_myeloids","UC_b_020_plasmas","UC_b_020_stroma","UC_b_020_tcells","UC_b_021_epi","UC_b_021_myeloids","UC_b_021_plasmas","UC_b_021_stroma","UC_b_021_tcells","UC_b_022_epi","UC_b_022_myeloids","UC_b_022_plasmas","UC_b_022_stroma","UC_b_022_tcells","UC_b_023_myeloids","UC_b_023_plasmas","UC_b_023_stroma","UC_b_023_tcells","UC_b_024_myeloids","UC_b_024_plasmas","UC_b_024_stroma","UC_b_024_tcells","UC_c_001_epi","UC_c_001_myeloids","UC_c_001_plasmas","UC_c_001_stroma","UC_c_001_tcells","UC_c_002_epi","UC_c_002_myeloids","UC_c_002_plasmas","UC_c_002_stroma","UC_c_002_tcells","UC_c_003_epi","UC_c_003_myeloids","UC_c_003_plasmas","UC_c_003_stroma","UC_c_003_tcells","UC_c_004_epi","UC_c_004_myeloids","UC_c_004_stroma","UC_c_005_epi","UC_c_005_myeloids","UC_c_005_plasmas","UC_c_005_stroma","UC_c_006_epi","UC_c_006_myeloids","UC_c_006_plasmas","UC_c_006_stroma","UC_c_006_tcells","UC_c_007_epi","UC_c_007_myeloids","UC_c_007_plasmas","UC_c_007_stroma","UC_c_007_tcells","UC_c_008_epi","UC_c_008_myeloids","UC_c_008_plasmas","UC_c_008_stroma","UC_c_008_tcells","UC_c_009_epi","UC_c_009_myeloids","UC_c_009_plasmas","UC_c_009_stroma","UC_c_009_tcells","UC_c_010_epi","UC_c_010_myeloids","UC_c_010_plasmas","UC_c_010_stroma","UC_c_010_tcells","UC_c_011_epi","UC_c_011_myeloids","UC_c_011_plasmas","UC_c_011_stroma","UC_c_011_tcells","UC_c_012_epi","UC_c_012_myeloids","UC_c_012_plasmas","UC_c_012_stroma","UC_c_012_tcells","UC_c_013_epi","UC_c_013_myeloids","UC_c_013_plasmas","UC_c_013_stroma","UC_c_013_tcells","UC_c_014_epi","UC_c_014_myeloids","UC_c_014_plasmas","UC_c_014_stroma","UC_c_014_tcells","UC_c_015_epi","UC_c_015_myeloids","UC_c_015_plasmas","UC_c_015_stroma","UC_c_015_tcells","UC_c_016_epi","UC_c_016_myeloids","UC_c_016_plasmas","UC_c_016_stroma","UC_c_016_tcells","UC_c_017_epi","UC_c_017_myeloids","UC_c_017_plasmas","UC_c_017_stroma","UC_c_017_tcells","UC_c_018_epi","UC_c_018_myeloids","UC_c_018_plasmas","UC_c_018_stroma","UC_c_018_tcells","UC_c_019_epi","UC_c_019_myeloids","UC_c_019_plasmas","UC_c_019_stroma","UC_c_019_tcells","UC_c_020_epi","UC_c_020_myeloids","UC_c_020_plasmas","UC_c_020_stroma","UC_c_020_tcells","UC_c_021_epi","UC_c_021_myeloids","UC_c_021_plasmas","UC_c_021_stroma","UC_c_021_tcells"],[279,12,99,78,2,2,28,2,230,21,8,1,129,5,309,25,156,69,9,19,2,189,3,23,4,9,64,3,270,18,76,69,2,338,8,100,21,1,93,6,37,22,49,1,6,1,68,2,14,2,2,57,3,6,2,85,3,7,3,1,256,6,56,35,1,31,3,6,4,2,1,2,8,25,5,6,17,1,33,6,11,14,1,328,8,30,16,807,1445,932,224,988,1694,993,206,910,891,701,172,982,447,426,294,1074,1025,532,350,684,526,797,53,755,386,587,185,3,931,1159,743,154,577,968,724,483,624,468,1051,619,140,432,778,1733,486,148,717,446,1505,568,139,606,1044,1548,560,244,212,1790,1761,816,448,919,645,1380,236,1002,2217,765,181,1336,1347,538,305,817,398,528,130,415,281,801,38,314,92,982,142,16,458,153,1163,175,35,299,26,382,96,6,8,10,194,36,2,804,186,901,209,22,290,95,436,74,2,681,98,615,142,15,438,37,279,82,11,378,88,393,134,7,497,42,359,120,2,476,38,225,112,3,141,9,77,56,680,89,681,175,3,411,119,766,196,8,163,64,133,43,105,14,138,22,3,423,91,348,220,25,499,61,217,276,36,247,30,136,72,13,238,94,733,398,185,241,74,273,188,17,277,61,271,212,39,251,51,210,115,19,297,33,185,171,11,204,74,631,173,109,238,64,407,144,51,268,71,279,146,14,240,55,192,122,5,173,56,316,184,26,285,45,268,137,15,233,109,329,252,38,323,60,402,254,51,301,42,147,95,31,459,40,141,178,31,867,150,439,332,50,1263,133,552,446,330,1222,133,333,448,69,1199,87,293,398,48,1319,96,524,438,123,1224,96,182,244,8,1299,64,170,265,2,556,41,134,290,4,1103,131,976,332,106,1382,119,317,390,11,1203,50,196,307,12,1071,78,152,152,7,793,165,1276,458,463,1183,54,182,162,4,1063,53,153,263,9,663,61,431,188,200,290,39,107,225,1,11,2,4,88,27,1,255,11,65,1,1,31,1,206,3,20,3,42,3,1,496,47,125,120,1,157,23,77,25,437,40,123,92,240,21,88,95,112,5,12,272,19,102,48,466,76,185,150,4,339,48,169,97,6,31,304,26,97,63,2,1,14,153,4,8,92,1,415,64,167,91,1,5,3,1,20,1,1,1,8,377,94,1170,711,13,186,89,666,188,1,112,207,1810,884,14,749,95,591,206,2,707,302,1765,1018,17,1339,164,1126,325,5,67,1138,441,25,1,48,353,132,4,5,197,2659,986,23,235,1477,1094,54,174,85,347,1,120,387,211,1,93,634,189,292,773,621,11,1,143,286,451,11,124,8,920,6,1,175,26,522,10,108,21,820,10,193,35,729,16,8,259,1100,610,45,52,433,1419,852,81,12,458,1444,628,80,15,280,898,404,32,263,182,349,27,3,237,336,352,17,1,274,686,393,23,4,360,769,421,25,162,127,191,12,146,141,160,15,179,48,235,7,16,140,104,4,6,314,1016,521,49,2,257,540,328,26,11,312,1802,552,90,1,161,442,314,21,4,334,676,469,36,66,434,99,13,10,293,1585,503,40,4,178,982,369,17,373,1147,531,64,340,1466,505,85,374,14,48,52,2,257,8,60,47,3,425,30,89,132,7,4,3,41,2,13,1,54,750,88,96,182,12,825,71,156,255,16,728,38,119,240,14,558,40,115,134,5,942,51,85,171,11,1085,24,127,64,1,1278,116,97,150,5,754,48,95,110,6,1244,49,108,79,3,970,62,107,139,4,1085,128,118,187,10,998,41,148,217,10,1062,45,107,126,12,254,4,42,80,2,319,17,80,108,7,240,19,40,144,4]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>condition<\/th>\n      <th>group<\/th>\n      <th>individual_code<\/th>\n      <th>fov_name<\/th>\n      <th>celltype_subset<\/th>\n      <th>fov_cluster<\/th>\n      <th>cells<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":7},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"condition","targets":1},{"name":"group","targets":2},{"name":"individual_code","targets":3},{"name":"fov_name","targets":4},{"name":"celltype_subset","targets":5},{"name":"fov_cluster","targets":6},{"name":"cells","targets":7}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>ggplot(celltype_summary_table, aes(x=cells, col=celltype_subset)) +
  geom_density() +
  geom_vline(xintercept=min_cells_per_fovcluster, lty=3) +
  geom_rug() +
  scale_x_log10() +
  theme_bw() +
  ggtitle(&quot;How many cells per fov-cluster?&quot;)</code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-7-1">
Past versions of unnamed-chunk-7-1.png
</button>
</p>
<div id="fig-unnamed-chunk-7-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/c77f76cecb74337321d877773789d8703b9bdb89/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-7-1.png" target="_blank">c77f76c</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>passed_fov_clusters &lt;- celltype_summary_table$fov_cluster[celltype_summary_table$cells &gt;= min_cells_per_fovcluster]</code></pre>
</div>
<div id="calculate-pseudobulk" class="section level2">
<h2>Calculate pseudobulk</h2>
<pre class="r"><code>pseudobulk_counts &lt;- PseudobulkExpression(so, assays = &quot;RNA&quot;, layer=&quot;counts&quot;,  method = &#39;aggregate&#39;, group.by = &#39;fov_cluster&#39;)</code></pre>
<pre><code>Names of identity class contain underscores (&#39;_&#39;), replacing with dashes (&#39;-&#39;)
This message is displayed once every 8 hours.</code></pre>
<pre class="r"><code>pseudobulk_counts_matrix &lt;- pseudobulk_counts[[&quot;RNA&quot;]]

# CHange - back to _. Ideally we&#39;d have neither, but - will cause problems later
colnames(pseudobulk_counts_matrix)&lt;-gsub(&quot;-&quot;,&quot;_&quot;,colnames(pseudobulk_counts_matrix))</code></pre>
<p>Keep only the passed fovs</p>
<pre class="r"><code>pseudobulk_counts_matrix &lt;- pseudobulk_counts_matrix[,passed_fov_clusters]

# pull in relevant annotation in a matched order
pseudobulk_anno_table &lt;- celltype_summary_table
match_order &lt;- match(passed_fov_clusters, pseudobulk_anno_table$fov_cluster)
pseudobulk_anno_table &lt;- pseudobulk_anno_table[match_order,]
stopifnot(all(colnames(pseudobulk_counts_matrix) == pseudobulk_anno_table$fov_cluster  ))</code></pre>
</div>
<div id="calculate-differential-expression" class="section level2">
<h2>Calculate Differential Expression</h2>
<pre class="r"><code>min_samples_to_calc &lt;- 2 # require 2 samples on on either side of contrast

de_result_list &lt;- list()

# celltype_subset is a matrix
for (the_celltype in levels(so$celltype_subset)) {
  
  anno_table.this   &lt;- pseudobulk_anno_table[pseudobulk_anno_table$celltype_subset == the_celltype,]
  count_matrix.this &lt;- pseudobulk_counts_matrix[,anno_table.this$fov_cluster]
 
  print(the_celltype)
  
  # skip clusters with nothing 
  if( nrow(anno_table.this) &lt; 1 ) {next}
  
  # Setup objects for limma
  dge &lt;- DGEList(count_matrix.this)
  dge &lt;- calcNormFactors(dge)
  
  # Build model
  group           &lt;- anno_table.this$group
  individual_code &lt;- anno_table.this$individual_code

  # To do any calculations, we need at least 2 pseudobulk groups per contrast.
  # there are plenty in this experiemnt, but with less replicates and rare cell types
  # it may be neccesary to check and skip certain contrasts

  # Model design 
  design    &lt;- model.matrix( ~0 + group)

  vm  &lt;- voom(dge, design = design, plot = FALSE)
  
  # Adding dupliate correlation to use individual fovs, rather than pooled per biosample
  corrfit &lt;- duplicateCorrelation(vm, design, block=individual_code)
  fit     &lt;- lmFit(vm, design, correlation = corrfit$consensus, block=individual_code) 

  # Then fit contrasts and run ebayes
  contrasts &lt;- makeContrasts(UCvHC  = groupUC - groupHC, 
                             CDvHC  = groupCD - groupHC,
                           levels=coef(fit))

  fit &lt;- contrasts.fit(fit, contrasts)
  fit &lt;- eBayes(fit)

  
  for ( the_coef in colnames(contrasts) ) {
    de_result.this &lt;- topTable(fit, n = Inf, adjust.method = &quot;BH&quot;, coef = the_coef) %&gt;%
      rownames_to_column(&quot;target&quot;) %&gt;%
      mutate(contrast=the_coef,
             celltype=the_celltype) %&gt;%
      select(celltype,contrast,target,everything()) %&gt;%
      arrange(P.Value)
    
    
      de_result_list[[paste(the_celltype, the_coef, sep=&quot;_&quot;)]] &lt;- de_result.this
    
  }

  

 
}</code></pre>
<pre><code>[1] &quot;epi&quot;
[1] &quot;myeloids&quot;
[1] &quot;plasmas&quot;
[1] &quot;stroma&quot;
[1] &quot;tcells&quot;</code></pre>
<pre class="r"><code>de_results_all &lt;- bind_rows(de_result_list)
de_results_sig &lt;- filter(de_results_all, adj.P.Val &lt; 0.01)</code></pre>
<p>Table of significant results.</p>
<pre class="r"><code>DT::datatable(de_results_sig)</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-0b50cf44ed0c70fb0261" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-0b50cf44ed0c70fb0261">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90"],["epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","epi","myeloids","myeloids","myeloids","myeloids","myeloids","myeloids","myeloids","myeloids","myeloids","myeloids","myeloids","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","plasmas","stroma","stroma","stroma","stroma","stroma","stroma","stroma","stroma","stroma","stroma","stroma","tcells","tcells","tcells","tcells","tcells","tcells","tcells"],["UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","CDvHC","CDvHC","CDvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","UCvHC","CDvHC","CDvHC","CDvHC","UCvHC","UCvHC","CDvHC","CDvHC","CDvHC","CDvHC","CDvHC"],["LYZ","IGHG2","IGHG1","HLA-DRB1","SAA1","B3GNT7","LCN2","CD74","HLA-DQB1","HLA-DRB5","DUOXA2","FZD5","OLFM4","CXCL3","RPS4Y1","SERPINA1","S100P","MT1X","HLA-DPA1","CHGA","BST2","CIITA","IGHG2","S100P","IGHG1","HLA-DRB1","HLA-DPA1","NFKBIA","JCHAIN","IGHG1","PIGR","IGHA1","EZR","FCGR3A","IGHG2","STAT1","IGHG2","IGHG1","TNFRSF18","IGHG2","ADAMDEC1","PIGR","IGHG1","TNFRSF13B","CXCL14","CCL3","KRT8","BCL2","MPO","AQP8","CSF2RB","NFKBIA","IL10RA","BMP6","TSLP","IGHA1","DUSP5","CLDN4","IGFBP5","GSN","CD9","ADAMDEC1","PIGR","CFD","CCL3","BMP6","TYK2","TNFRSF18","NFKBIA","CSF2RB","TNFRSF13B","IGHG2","CHI3L1","IGHG2","PIGR","IGHG1","RGS2","ADIRF","NEAT1","THBS2","IGHG1","IGHG2","GSN","IGHG2","IGHG1","IGHG2","IGHG1","NKG7","IGFBP3","CCL5"],[3.567564119296202,2.795073130682136,2.879391845164468,2.361549337874527,1.631196105042839,-2.22902616410666,2.986098025933204,2.677411189005053,1.512103586119469,2.350496535521083,1.946177981090452,-1.3765127842276,3.225954573501639,1.43996969891476,1.61826043428817,1.523583054716656,1.647556307515806,-1.605970493589,1.635540055819042,-2.428388704219707,1.187321072748105,1.776041545678071,1.830887485617209,1.638926426762776,1.731910236980951,1.679511886277687,1.723530709567912,-1.323530859914316,-3.055135338194182,2.626831673261172,-1.656182816008096,-2.504257536956217,-1.182089886382563,1.259345661385609,2.490625965927146,1.444024189469216,2.959214486069962,2.895197538495122,-1.551875547459231,2.801221266031003,-2.042545759264552,-1.719333849950313,2.436133826847907,-1.377993100947934,-1.56173539484846,-1.312671899149015,-1.247059668779755,-1.003574298926768,1.035748284455588,0.8638284635710001,-0.9813245550919021,-1.070030637058935,-0.881687075173776,-1.235671222969335,1.032733391529867,-1.946599175585204,-1.157763846524961,-0.9386019983812162,1.017711609577507,-0.8306380731087852,-0.925595441149385,-2.231047465768352,-1.736983082599837,-1.234176073759773,-1.308053769356771,-1.426543407705825,-1.553999828624507,-0.8592175719326107,-1.12644069219124,-0.966821425094798,-1.00741190505738,1.571530944842946,1.717449126530921,2.392657245708063,-1.53413582799122,2.291347247631435,-1.052363849724966,-1.127049405740587,1.031574678330074,1.072829965767276,2.670784285849066,2.583315979828216,-1.468737864512629,3.762142871006782,3.756961086904459,3.13556379060284,2.824195244217366,1.754341898812724,-1.672256462920759,1.495756943713651],[10.03158050151357,10.08822340923212,10.64183684123074,10.49204862077934,9.280586762672716,10.21965706852791,9.819188792751048,11.0739699195851,9.717027813777538,8.99914742966644,9.266166959434551,10.99377115544704,11.58920055217826,10.53460812999346,8.163485904047038,10.90036096185432,10.90299566469834,11.03824378594429,9.383216924156001,9.831014457379553,8.675779119784458,9.006292693527007,10.08822340923212,10.90299566469834,10.64183684123074,10.49204862077934,9.383216924156001,10.75163886811029,11.57605197770097,12.7038035812418,9.492965372705058,13.40836899907759,9.528723480348708,10.53797953929835,12.17485116407275,10.44487553969692,12.17485116407275,12.7038035812418,9.860769732625343,14.2916407265515,9.643947460803444,9.370663859874281,14.72519515171609,9.334726352946706,9.790280729272755,9.250379846644684,10.75498269117512,9.838659184830153,8.976412435826091,9.213255382182449,9.723124920028235,10.19017774981531,9.798257678334519,9.227690542438465,8.432198190201893,15.9808948117601,11.86807935687188,8.724556749287968,9.523617410582409,9.826316187992772,8.852990530101223,9.643947460803444,9.370663859874281,9.841842559522757,9.250379846644684,9.227690542438465,11.99171417660733,9.860769732625343,10.19017774981531,9.723124920028235,9.334726352946706,14.2916407265515,9.671695326977723,11.97946364454942,9.514133607886265,12.5494684237014,10.45861719404181,9.755847063282133,12.82855389950085,9.650613590274645,12.5494684237014,11.97946364454942,11.72216107222022,12.79161423293002,13.3295094037593,12.79161423293002,13.3295094037593,10.15096070274869,9.003218595093237,11.91857001409501],[7.232630638451168,6.888271579353455,6.887306136142595,5.945282521192369,5.812176889783621,-5.724733455355884,5.527039987802286,5.32880465187132,5.252396917911148,4.989584057967216,4.944140327455846,-4.902413902067449,4.732053747074156,4.714972685239418,4.136114108992563,4.0737112807222,4.069453665490661,-4.044220391168595,3.933040304155463,-3.918784481221275,5.071872764037372,5.067494067382227,4.783752650890804,4.59884163539882,4.467468770218849,4.431291113475372,4.360751530459121,-5.664051396209842,-5.501482098933411,4.870486500417138,-4.791720222585978,-4.673835062816096,-4.650899541306694,4.628228862388813,4.627912594019214,5.269825402163731,5.112629150754543,5.001376066502649,-8.342410693179668,7.650725444382537,-6.840051126430498,-6.46716858606015,5.942123538428466,-5.613563748244217,-5.44597778741044,-5.251556323549513,-4.7182180517685,-4.711822851816427,4.643079099309246,4.414038038421214,-4.263384275158596,-4.251331626358502,-4.240852169118441,-4.13714078259534,4.070008981040085,-4.065412449100157,-3.961458382163322,-3.92204639909292,3.895802882582506,-3.839226383242579,-3.794647651305515,-7.340891481654241,-6.439885374118395,-5.440016834988488,-5.161958935233134,-4.716422979007784,-4.700432727125553,-4.450000894176343,-4.278749156708985,-4.061898975405736,-4.018194835834734,3.968111458363705,5.357873038996134,5.05009507952037,-4.887808232124166,4.745493940782008,-4.417545690936533,-4.317017441317242,4.124337621073577,4.042312497634022,5.497519322715465,5.394030116228833,-4.415274391617404,7.463696590874045,7.187127921254906,6.318604563809319,5.515429374855586,4.82932218512822,-4.617368950041155,4.518477246358165],[5.119345362818246e-11,2.923327921764283e-10,2.937509365918907e-10,2.867917974831155e-08,5.340761465147169e-08,8.004303439047451e-08,1.974629258294922e-07,4.800461154677258e-07,6.728368802398447e-07,2.10386543842307e-06,2.553471734465412e-06,3.047640514526322e-06,6.217291010092443e-06,6.67238606426986e-06,6.642101698668688e-05,8.410747025990232e-05,8.5465821427275e-05,9.39573091199037e-05,0.0001419576130961226,0.0001495885172048369,1.477635308545976e-06,1.505806696010763e-06,5.015716447518878e-06,1.074140524982886e-05,1.824342074241814e-05,2.107266228911371e-05,2.785302822961781e-05,7.460552570915575e-08,1.614780646718673e-07,2.829217322718362e-06,3.981371170412786e-06,6.593078966209273e-06,7.265816867494595e-06,7.995703791539758e-06,8.006370600827041e-06,4.738810765097689e-07,9.677902767881571e-07,1.590896834363102e-06,2.765227677948197e-14,1.578767989065736e-12,1.471273293748189e-10,1.084752512387892e-09,1.622924796387145e-08,8.225925434564765e-08,1.841084148031835e-07,4.596824050086569e-07,5.045689633084174e-06,5.187241600245973e-06,6.972488496149413e-06,1.827357293861324e-05,3.379662361007394e-05,3.54774352721205e-05,3.700369415638532e-05,5.591569449144431e-05,7.275762334522323e-05,7.407271255893418e-05,0.0001106361522926812,0.000128557207087295,0.0001419865433407018,0.0001756108556596315,0.0002072924249020821,9.193075078771764e-12,1.252485514488161e-09,1.894065844719342e-07,6.956932297969378e-07,5.085040644519933e-06,5.448909196048362e-06,1.574370012663578e-05,3.176434347661988e-05,7.509319816405989e-05,8.896234888494423e-05,0.0001078559343641328,2.689463529182307e-07,1.123802473376827e-06,2.334262482354535e-06,4.37187652158974e-06,1.766366039686707e-05,2.671610314376168e-05,5.791938663492277e-05,7.989021471140958e-05,1.380324423234782e-07,2.265297032982009e-07,1.783089780383879e-05,8.923785374155373e-11,3.067727843391495e-10,1.388536492392634e-08,4.130867582205837e-07,6.462233343889036e-06,1.459806440178121e-05,2.121715275998987e-05],[5.114226017455428e-08,9.781906188509959e-08,9.781906188509959e-08,7.16262514214081e-06,1.067084140736404e-05,1.332716522601401e-05,2.818078041480897e-05,5.994575866903226e-05,7.468489370662276e-05,0.0002101761572984647,0.0002319016602482679,0.0002537160728343163,0.0004761224055861136,0.0004761224055861136,0.004423639731313346,0.005022373859167513,0.005022373859167513,0.005214630656154655,0.007463981867527711,0.007471946434381603,0.0007521504446573761,0.0007521504446573761,0.001670233577023787,0.002682665961144758,0.003508598271137432,0.003508598271137432,0.003975025028769742,7.45309201834466e-05,8.065829330359769e-05,0.0009421293684652145,0.0009943474498105933,0.0009997955287782767,0.0009997955287782767,0.0009997955287782767,0.0009997955287782767,0.0004734071954332591,0.0004834112432556845,0.000529768645842913,2.762462450270249e-11,7.885946105383352e-10,4.899340068181468e-08,2.70916939968876e-07,3.242603743181516e-06,1.369616584855033e-05,2.627490091262576e-05,5.740284032545603e-05,0.0005182054358645727,0.0005182054358645727,0.0006332287279684785,0.001521274947139552,0.002464446030815263,0.002464446030815263,0.002464446030815263,0.003491236174809554,0.004111035547020847,0.004111035547020847,0.005817132428441504,0.006421432494010387,0.006754502704636244,0.007974329309271447,0.009003701412051306,9.183882003692993e-09,6.256165144868365e-07,6.307239262915409e-05,0.0001737493841417852,0.0009072433811420522,0.0009072433811420522,0.002246850918072734,0.003966572391642907,0.008335344996210648,0.00888733865360593,0.009795279857251695,0.0002686774065653124,0.000561339335451725,0.0007773094066240603,0.001091876161267038,0.003529199347294042,0.00444823117343632,0.00826592389261255,0.009976290562087271,0.0001131515867974513,0.0001131515867974513,0.005937688968678316,8.914861588781218e-08,1.532330057774052e-07,1.387147955900241e-05,0.0002063368357311816,0.002151923703515049,0.003645866584344857,0.004239187121445977],[14.69238513777382,13.05866703432359,13.03426921103388,8.711641595720597,8.122865305782181,7.742637380062929,6.889559057450681,6.061015340769459,5.730799204135004,4.655580287827374,4.473216650960913,4.341798489594714,3.687744622424388,3.595972418520791,1.433395664294009,1.255963818716237,1.255198044727718,1.172041888665698,0.7186912027124155,0.7035498422638957,5.005988546317318,4.984556703757729,3.878042928402239,3.180597548504576,2.702724355458039,2.57265959438836,2.319766632195723,7.761673649497091,7.029726280155962,4.390598048975406,4.080853389832622,3.613684445805871,3.528843621310983,3.441393063190576,3.439589433225058,5.970577512724047,5.315577637595614,4.872578075782341,21.99065253911652,17.89611038923422,13.66974618507669,11.73356785593951,9.179991820443474,7.562097595374857,6.79836567128377,5.912201834893696,3.657444798762384,3.600245919587826,3.336104021128083,2.417443217162118,1.821949844552977,1.792073907387681,1.736661548689145,1.351426512005018,1.13947269211905,1.42360631324742,0.7998729521525005,0.5804395727746758,0.484153114362714,0.2745748219458379,0.1313779753277968,16.15852386325345,11.57643422857783,6.893013674084593,5.686575781080462,3.845449121719792,3.746863078424248,2.809122901290561,2.171007089183513,1.377709590428573,1.220779333341551,1.011380972078205,6.56029786225487,5.191980346737862,4.563084975134908,3.935151225880965,2.706318061824186,2.317915840505318,1.625317407348885,1.318442192834512,6.933194157133171,6.568393313884926,2.680102978119456,13.74603279856178,12.63176971324722,9.323864530290809,6.207635434093853,3.682200629419371,2.935436333483019,2.599751602209489]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>celltype<\/th>\n      <th>contrast<\/th>\n      <th>target<\/th>\n      <th>logFC<\/th>\n      <th>AveExpr<\/th>\n      <th>t<\/th>\n      <th>P.Value<\/th>\n      <th>adj.P.Val<\/th>\n      <th>B<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[4,5,6,7,8,9]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"celltype","targets":1},{"name":"contrast","targets":2},{"name":"target","targets":3},{"name":"logFC","targets":4},{"name":"AveExpr","targets":5},{"name":"t","targets":6},{"name":"P.Value","targets":7},{"name":"adj.P.Val","targets":8},{"name":"B","targets":9}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="de-plots" class="section level2">
<h2>DE plots</h2>
<pre class="r"><code>library(ggrepel) # gg_repel, For non-overlapping gene labels


make_ma_style_plot &lt;- function(res_table, pval_threshold = 0.01, n_genes_to_label = 10) {
  p &lt;- ggplot(res_table, aes(x=AveExpr, y=logFC, col=adj.P.Val &lt; pval_threshold) ) +
    geom_hline(yintercept = c(0), col=&#39;grey80&#39;) +
    geom_point(pch=3) +
    geom_text_repel(data    = head(arrange(filter(res_table , adj.P.Val &lt; pval_threshold ), P.Value), n=5),
                    mapping = aes(label=target), col=&quot;red&quot; ) +
    theme_bw() +
    geom_hline(yintercept = c(-1,1), lty=3) +
    scale_colour_manual(values = c(&#39;FALSE&#39;=&quot;black&quot;, &#39;TRUE&#39;=&quot;red&quot;)) +
    theme(legend.position = &#39;none&#39;)
  return(p)
}</code></pre>
<pre class="r"><code>#res_table.UCvHC.epi &lt;- filter(de_results_all, contrast == &quot;UCvHC&quot;, celltype==&quot;epi&quot;)

make_ma_style_plot(res_table = filter(de_results_all, contrast == &quot;UCvHC&quot;, celltype==&quot;epi&quot;))</code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-1">
Past versions of unnamed-chunk-14-1.png
</button>
</p>
<div id="fig-unnamed-chunk-14-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/5a9d7e9002faa026640cfe522a79bcf86b1ad8bf/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-14-1.png" target="_blank">5a9d7e9</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>make_ma_style_plot(res_table = filter(de_results_all, contrast == &quot;UCvHC&quot;, celltype==&quot;tcells&quot;))</code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-14-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-2">
Past versions of unnamed-chunk-14-2.png
</button>
</p>
<div id="fig-unnamed-chunk-14-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/5a9d7e9002faa026640cfe522a79bcf86b1ad8bf/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-14-2.png" target="_blank">5a9d7e9</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>make_ma_style_plot(res_table = filter(de_results_all, contrast == &quot;UCvHC&quot;, celltype==&quot;stroma&quot;))</code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-14-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-3">
Past versions of unnamed-chunk-14-3.png
</button>
</p>
<div id="fig-unnamed-chunk-14-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/5a9d7e9002faa026640cfe522a79bcf86b1ad8bf/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-14-3.png" target="_blank">5a9d7e9</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="check-some-examples" class="section level2">
<h2>Check some examples</h2>
<p>Its always worth visualising how the expression of your
differentially expressed genes really looks, with respect to your
experimental design. How best to do this depends on your experiment.</p>
<p>The results suggests that TNFRSF18 was significantly DE between
individuals with Ulcerative Colitis and Healthy Controls in plasma
cells. There’s some very convenient seurat plots below;</p>
<pre class="r"><code>VlnPlot(subset(so, celltype_subset == &quot;plasmas&quot;), features = &quot;TNFRSF18&quot;, group.by = &#39;group&#39;, alpha = 0.1)</code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-15-1">
Past versions of unnamed-chunk-15-1.png
</button>
</p>
<div id="fig-unnamed-chunk-15-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/5a9d7e9002faa026640cfe522a79bcf86b1ad8bf/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-15-1.png" target="_blank">5a9d7e9</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>FeaturePlot(so, &quot;TNFRSF18&quot;, split.by = &quot;group&quot;)</code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-15-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-15-2">
Past versions of unnamed-chunk-15-2.png
</button>
</p>
<div id="fig-unnamed-chunk-15-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/5a9d7e9002faa026640cfe522a79bcf86b1ad8bf/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-15-2.png" target="_blank">5a9d7e9</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>But it gets difficult to summarise data at the single cell level. We
can also use the the normalised pseudobulk expression to see how gene
expression varies within each fov,individual,celltype and condition -
The plot below shows an overview of TNFRSF18 across the entire
experiment.</p>
<p>NB: We have to plot normalised expression instead of the raw counts
as there are vastly different numbers of cells in each fov+celltype
grouping.</p>
<pre class="r"><code># Get tmm normalised coutns for all pseudobulk
# WHen we did the DE we calculated this a celltype at a time, so values might differ slightly!
dge &lt;- DGEList(pseudobulk_counts_matrix)
dge &lt;- calcNormFactors(dge)
norm_pseudobulk &lt;- cpm(dge , log=TRUE) # uses tmm normalisation

# Plot expression for TNFRSF18
plottable &lt;- cbind(pseudobulk_anno_table, expression = norm_pseudobulk[&quot;TNFRSF18&quot;,])
ggplot(plottable, aes(x=individual_code, y=expression, col=condition )) +
  geom_boxplot(outlier.shape = NA) +
  geom_point() + 
  theme_bw() + 
  theme(axis.text.x=element_text(angle = -90, hjust = 0)) + 
  facet_wrap(~celltype_subset) </code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-16-1">
Past versions of unnamed-chunk-16-1.png
</button>
</p>
<div id="fig-unnamed-chunk-16-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/5a9d7e9002faa026640cfe522a79bcf86b1ad8bf/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-16-1.png" target="_blank">5a9d7e9</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>And you can compare that with IGHG1, which was flagged at
differentially expressed across multiple cell types.</p>
<pre class="r"><code>plottable &lt;- cbind(pseudobulk_anno_table, expression = norm_pseudobulk[&quot;IGHG1&quot;,])
ggplot(plottable, aes(x=individual_code, y=expression, col=condition )) +
  geom_boxplot(outlier.shape = NA) +
  geom_point() + 
  theme_bw() + 
  theme(axis.text.x=element_text(angle = -90, hjust = 0)) + 
  facet_wrap(~celltype_subset) </code></pre>
<p><img src="figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-17-1">
Past versions of unnamed-chunk-17-1.png
</button>
</p>
<div id="fig-unnamed-chunk-17-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/5a9d7e9002faa026640cfe522a79bcf86b1ad8bf/docs/figure/e_DEPseudobulk_insitu.Rmd/unnamed-chunk-17-1.png" target="_blank">5a9d7e9</a>
</td>
<td>
swbioinf
</td>
<td>
2024-05-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="code-snippet" class="section level1">
<h1>Code Snippet</h1>
<p>Todo</p>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<pre class="r"><code>head(de_results_all)</code></pre>
<pre><code>  celltype contrast   target     logFC   AveExpr         t      P.Value
1      epi    UCvHC      LYZ  3.567564 10.031581  7.232631 5.119345e-11
2      epi    UCvHC    IGHG2  2.795073 10.088223  6.888272 2.923328e-10
3      epi    UCvHC    IGHG1  2.879392 10.641837  6.887306 2.937509e-10
4      epi    UCvHC HLA-DRB1  2.361549 10.492049  5.945283 2.867918e-08
5      epi    UCvHC     SAA1  1.631196  9.280587  5.812177 5.340761e-08
6      epi    UCvHC   B3GNT7 -2.229026 10.219657 -5.724733 8.004303e-08
     adj.P.Val         B
1 5.114226e-08 14.692385
2 9.781906e-08 13.058667
3 9.781906e-08 13.034269
4 7.162625e-06  8.711642
5 1.067084e-05  8.122865
6 1.332717e-05  7.742637</code></pre>
<p>This table is the typical output of <em>limma</em> tests; With a
couple of extra columns added by our code.</p>
<ul>
<li><strong>celltype</strong>: The celltype being tested (Added by
example code)</li>
<li><strong>contrast</strong>: The contrast being tested (Added by
example code)</li>
<li><strong>target</strong> : The gene name (Added by example code, is
the rowname in limma output)</li>
<li><strong>rownames</strong> : The tested cell types</li>
<li><strong>logFC</strong> : Log 2 fold change between tested groups.
For a test of Test-Con;
<ul>
<li>At logFC +1, A is doubled B.</li>
<li>At logFC -1, A is half of B.<br />
</li>
<li>A logFC 0 indicates no change.</li>
</ul></li>
<li><strong>AveExpr</strong> : Average expression of a gene across all
replicates.</li>
<li><strong>t</strong> : Moderated T-statistic. See Limma
documentation.</li>
<li><strong>P.Value</strong> : P.value</li>
<li><strong>adj.P.Val</strong> : A multiple-hypothesis corrected
p-value</li>
<li><strong>B</strong> : B statistic (rarely used). See Limma
documentation.</li>
</ul>
</div>
<div id="more-information" class="section level1">
<h1>More Information</h1>
<p>Todo</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] ggrepel_0.9.5      edgeR_4.0.16       DT_0.33            limma_3.58.1      
 [5] lubridate_1.9.3    forcats_1.0.0      stringr_1.5.1      dplyr_1.1.4       
 [9] purrr_1.0.2        readr_2.1.5        tidyr_1.3.1        tibble_3.2.1      
[13] ggplot2_3.5.0      tidyverse_2.0.0    speckle_1.2.0      Seurat_5.0.3      
[17] SeuratObject_5.0.1 sp_2.1-3           workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22            splines_4.3.2              
  [3] later_1.3.2                 bitops_1.0-7               
  [5] polyclip_1.10-6             fastDummies_1.7.3          
  [7] lifecycle_1.0.4             rprojroot_2.0.4            
  [9] globals_0.16.3              processx_3.8.4             
 [11] lattice_0.22-6              MASS_7.3-60.0.1            
 [13] crosstalk_1.2.1             magrittr_2.0.3             
 [15] plotly_4.10.4               sass_0.4.9                 
 [17] rmarkdown_2.26              jquerylib_0.1.4            
 [19] yaml_2.3.8                  httpuv_1.6.15              
 [21] sctransform_0.4.1           spam_2.10-0                
 [23] spatstat.sparse_3.0-3       reticulate_1.35.0          
 [25] cowplot_1.1.3               pbapply_1.7-2              
 [27] RColorBrewer_1.1-3          abind_1.4-5                
 [29] zlibbioc_1.48.2             Rtsne_0.17                 
 [31] GenomicRanges_1.54.1        BiocGenerics_0.48.1        
 [33] RCurl_1.98-1.14             git2r_0.33.0               
 [35] GenomeInfoDbData_1.2.11     IRanges_2.36.0             
 [37] S4Vectors_0.40.2            irlba_2.3.5.1              
 [39] listenv_0.9.1               spatstat.utils_3.0-4       
 [41] goftest_1.2-3               RSpectra_0.16-1            
 [43] spatstat.random_3.2-3       fitdistrplus_1.1-11        
 [45] parallelly_1.37.1           leiden_0.4.3.1             
 [47] codetools_0.2-20            DelayedArray_0.28.0        
 [49] tidyselect_1.2.1            farver_2.1.1               
 [51] matrixStats_1.2.0           stats4_4.3.2               
 [53] spatstat.explore_3.2-7      jsonlite_1.8.8             
 [55] progressr_0.14.0            ggridges_0.5.6             
 [57] survival_3.5-8              tools_4.3.2                
 [59] ica_1.0-3                   Rcpp_1.0.12                
 [61] glue_1.7.0                  gridExtra_2.3              
 [63] SparseArray_1.2.4           xfun_0.43                  
 [65] MatrixGenerics_1.14.0       GenomeInfoDb_1.38.8        
 [67] withr_3.0.0                 BiocManager_1.30.22        
 [69] fastmap_1.1.1               fansi_1.0.6                
 [71] callr_3.7.6                 digest_0.6.35              
 [73] timechange_0.3.0            R6_2.5.1                   
 [75] mime_0.12                   colorspace_2.1-0           
 [77] scattermore_1.2             tensor_1.5                 
 [79] spatstat.data_3.0-4         utf8_1.2.4                 
 [81] generics_0.1.3              renv_1.0.5                 
 [83] data.table_1.15.4           httr_1.4.7                 
 [85] htmlwidgets_1.6.4           S4Arrays_1.2.1             
 [87] whisker_0.4.1               uwot_0.1.16                
 [89] pkgconfig_2.0.3             gtable_0.3.4               
 [91] lmtest_0.9-40               SingleCellExperiment_1.24.0
 [93] XVector_0.42.0              htmltools_0.5.8            
 [95] dotCall64_1.1-1             scales_1.3.0               
 [97] Biobase_2.62.0              png_0.1-8                  
 [99] knitr_1.45                  rstudioapi_0.16.0          
[101] tzdb_0.4.0                  reshape2_1.4.4             
[103] nlme_3.1-164                cachem_1.0.8               
[105] zoo_1.8-12                  KernSmooth_2.23-22         
[107] parallel_4.3.2              miniUI_0.1.1.1             
[109] pillar_1.9.0                grid_4.3.2                 
[111] vctrs_0.6.5                 RANN_2.6.1                 
[113] promises_1.2.1              xtable_1.8-4               
[115] cluster_2.1.6               evaluate_0.23              
[117] cli_3.6.2                   locfit_1.5-9.9             
[119] compiler_4.3.2              rlang_1.1.3                
[121] crayon_1.5.2                future.apply_1.11.2        
[123] labeling_0.4.3              ps_1.7.6                   
[125] getPass_0.2-4               plyr_1.8.9                 
[127] fs_1.6.3                    stringi_1.8.3              
[129] viridisLite_0.4.2           deldir_2.0-4               
[131] munsell_0.5.1               lazyeval_0.2.2             
[133] spatstat.geom_3.2-9         Matrix_1.6-5               
[135] RcppHNSW_0.6.0              hms_1.1.3                  
[137] patchwork_1.2.0             future_1.33.2              
[139] statmod_1.5.0               shiny_1.8.1.1              
[141] highr_0.10                  SummarizedExperiment_1.32.0
[143] ROCR_1.0-11                 igraph_2.0.3               
[145] bslib_0.7.0                </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
