<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Sarah Williams" />


<title>Spatially resticted genes</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.33/datatables.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Spatial Sampler</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tests</a>
</li>
<li>
  <a href="index_data.html">Data</a>
</li>
<li>
  <a href="contributing.html">Contributing</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/swbioinf/spatialsnippets">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Spatially resticted genes</h1>
<h4 class="author">Sarah Williams</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-01
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>spatialsnippets/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20231017code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20231017)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20231017code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20231017)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstreeda09a66f0eaa5bf6bbb0dd4a4c73d417dca6b8a2targetblankda09a66a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/swbioinf/spatialsnippets/tree/da09a66f0eaa5bf6bbb0dd4a4c73d417dca6b8a2" target="_blank">da09a66</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstreeda09a66f0eaa5bf6bbb0dd4a4c73d417dca6b8a2targetblankda09a66a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/swbioinf/spatialsnippets/tree/da09a66f0eaa5bf6bbb0dd4a4c73d417dca6b8a2" target="_blank">da09a66</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/contributions.nb.html
    Ignored:    analysis/e_neighbourcellchanges.nb.html
    Ignored:    analysis/e_spatiallyRestrictedGenesVoyager.nb.html
    Ignored:    analysis/g_toolkits.nb.html
    Ignored:    analysis/glossary.nb.html
    Ignored:    renv/library/
    Ignored:    renv/staging/

Untracked files:
    Untracked:  code/scwat-at_betweenClusters_heterotypic_score_functions.R
    Untracked:  other/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/e_spatiallyRestrictedGenes.Rmd</code>) and HTML
(<code>docs/e_spatiallyRestrictedGenes.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/da09a66f0eaa5bf6bbb0dd4a4c73d417dca6b8a2/analysis/e_spatiallyRestrictedGenes.Rmd" target="_blank">da09a66</a>
</td>
<td>
swbioinf
</td>
<td>
2024-11-01
</td>
<td>
wflow_publish("analysis")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/0ce06a8fe9e3d0354b4cca882c8978ff95d0f1bf/docs/e_spatiallyRestrictedGenes.html" target="_blank">0ce06a8</a>
</td>
<td>
swbioinf
</td>
<td>
2024-11-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/6cf65ff5b2702f1ec8a3209fffb576099129462a/analysis/e_spatiallyRestrictedGenes.Rmd" target="_blank">6cf65ff</a>
</td>
<td>
swbioinf
</td>
<td>
2024-11-01
</td>
<td>
wflow_publish("analysis/e_spatiallyRestrictedGenes.Rmd")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/45cde50463b2e404d922760981ea284008526769/analysis/e_spatiallyRestrictedGenes.Rmd" target="_blank">45cde50</a>
</td>
<td>
swbioinf
</td>
<td>
2024-10-28
</td>
<td>
adding halfbuilt files
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/swbioinf/spatialsnippets/4e21b4e777c143f6fb9229d92b13c68a7af1e1ff/docs/e_spatiallyRestrictedGenes.html" target="_blank">4e21b4e</a>
</td>
<td>
swbioinf
</td>
<td>
2024-07-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/a10809a605c50889944f135d7f139d35f3acbde0/analysis/e_spatiallyRestrictedGenes.Rmd" target="_blank">a10809a</a>
</td>
<td>
swbioinf
</td>
<td>
2024-07-03
</td>
<td>
wflow_publish("analysis/e_spatiallyRestrictedGenes.Rmd")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/fd64e3009d2a3284f3a5e43d572c953497246b13/analysis/e_spatiallyRestrictedGenes.Rmd" target="_blank">fd64e30</a>
</td>
<td>
swbioinf
</td>
<td>
2024-06-18
</td>
<td>
new plots
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>It is possible to test for genes that are expressed in a spatially
non-random pattern. These might be restricted to regions of a tissu
(e.g. epithelia), with a very punctate expression in selected cells only
(e.g. immunoglobulins in plasma cells).</p>
<p>One popular approach to find these genes is the MoransI test of
spatial autocorrelation.</p>
<p><img src="assets/spatiallyrestricted.png" /></p>
<p>This example will show how to use the morans test within seurat to
find spatially restricted genes.</p>
<p>This requires:</p>
<ul>
<li>X,Y coordinates of individual transcripts, or their cells.</li>
<li>Annotation of the tissue samples in the cell metadata (if multiple
samples)</li>
</ul>
<p>There is no need for celltype annotation.</p>
<p>For example:</p>
<ul>
<li>What cells are saptially restricted in this tissue, are they in a
layer of epithelial cells, some structure or elsewhere.</li>
<li>Are some cancer-related genes showing a restricted spatial
expression (and does this align with tumour-dense regions?)</li>
</ul>
<p>Steps:</p>
<ol style="list-style-type: decimal">
<li>Subset each tissue</li>
<li>Calculate moransI for each sample</li>
<li>Join and inspect results.</li>
</ol>
</div>
<div id="worked-example" class="section level1">
<h1>Worked example</h1>
<p>Paper <a
href="https://www.sciencedirect.com/science/article/pii/S2211124724005448?via%3Dihub">Microglia-astrocyte
crosstalk in the amyloid plaque niche of an Alzheimer’s disease mouse
model, as revealed by spatial transcriptomics</a><span
class="citation">(Mallach et al. 2024)</span> <span
class="citation">(Mallach et al. 2024)</span> explores the spatial
transcritome of amaloid plaques in a mouse model.</p>
<p>Their work includes an analysis of cosMX samples from of 4 mouse
brain samples.</p>
<p>Those sections show a very distinct spatial patterning of the
different brain structures. This example will test which genes are
expressed in a spatially restricted pattern, independently of any
celltype annotations. e.g.</p>
<ul>
<li>Present within or nearby a given structure</li>
<li>Existing in ‘clumps’ e.g. high expression on a small subset of
cells</li>
</ul>
<div id="load-libraries-and-data" class="section level3">
<h3>Load libraries and data</h3>
<p>Load relevant libraries</p>
<pre class="r"><code># NB: The Rfast2 and ape packages may need to be installed
# to use the moransI test (in addition to Seurat)
# The GSL system library might also need to be installed, if it isn&#39;t already.
# install.packages(&#39;Rfast2&#39;)
# install.packages(&#39;ape&#39;)

library(Seurat)
library(tidyverse)
library(DT)</code></pre>
<p>And load the preprocessed Seurat object.</p>
<pre class="r"><code>dataset_dir      &lt;- &#39;~/projects/spatialsnippets/datasets/GSE263793_Mallach2024_AlzPlaque/processed_data/&#39;
seurat_file_01_preprocessed &lt;- file.path(dataset_dir,  &quot;GSE263793_AlzPlaque_seurat_01_preprocessed.RDS&quot;)

so &lt;- readRDS(seurat_file_01_preprocessed)</code></pre>
</div>
<div id="spatially-variable-features" class="section level3">
<h3>Spatially variable features</h3>
<p>Morans test can be slow to run, so save time by only running it on
variable features. Variable features are those with a non-even
distribution across cells, and are routinely deteced with
<em>FindVariableFeatures()</em> during preprocessing. Non variable
features are unlikely to be spatially restricted.</p>
<p>For the purpose of this demo, recalculate just the top 10 variable
features. The actual number for a real experiment could be judged from
the variable features plot below, e.g. 100-200-2000 (it really depends
on your panel!).</p>
<pre class="r"><code>num_variable_features = 10 # Test only, Would be 100+ in real life
so &lt;-  FindVariableFeatures(so, nfeatures=num_variable_features)
VariableFeaturePlot(so)</code></pre>
<p><img src="figure/e_spatiallyRestrictedGenes.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We will look for <em>spatially</em> variable features on each of our
tissue samples indenedantly; in this case 4 samples across 2 slides. But
first, just test one.</p>
<p>We do this because all probes are essentially going to be restricted
to the tissue itself, not the surrounding empty slide. This might be
particularly noticeable on a panel of many small cores.</p>
<p>First, subset to just one tissue sample.</p>
<pre class="r"><code>so.sample &lt;- subset( so, subset= sample == &#39;sample1&#39;)</code></pre>
<p>Then find the spatially variable genes with
<em>FindSpatiallyVariableFeatures()</em> function.</p>
<p>That code <em>should</em> look like this:</p>
<pre class="r"><code>so.sample &lt;- FindSpatiallyVariableFeatures(
  so.sample, 
  assay = &quot;RNA&quot;, 
  features = VariableFeatures(so.sample),    
  selection.method = &quot;moransi&quot;,
  layer = &quot;counts&quot;)</code></pre>
<p>However, right now, there is a bug with the current
FindSpatiallyVariableFeatures function, described <a
href="https://github.com/satijalab/seurat/issues/8226">here</a>. Likely
that issue will be closed when this is fixed!</p>
<p>If you see the following error - try a workaround, using a customised
version of this that avoids the issue in just this dataset. The custom
version, may be found <a
href="https://github.com/swbioinf/spatialsnippets/blob/main/code/spatially_variable_features_code.R">here</a></p>
<pre><code>Error in `LayerData&lt;-`:
! &#39;layer&#39; must be a single non-empty string</code></pre>
<p>The adjusted function is called as:</p>
<pre class="r"><code># Workaround 
# Available: https://github.com/swbioinf/spatialsnippets/blob/main/code/spatially_variable_features_code.R
source(&quot;code/spatially_variable_features_code.R&quot;)
so.sample &lt;- FindSpatiallyVariableFeatures.Seurat_EDITED(
  so.sample, 
  assay = &quot;RNA&quot;, 
  features = VariableFeatures(so.sample),    
  selection.method = &quot;moransi&quot;,
  layer = &quot;counts&quot;)</code></pre>
<pre><code>[1] &quot;&gt;&gt;&gt;&gt; USING EDITED FUNCTION!!!! &lt;&lt;&lt;&quot;</code></pre>
<p><em>FindSpatiallyVariableFeatures()</em> should return a seurat
object with the moransI scores embedded in the feature metatdata of the
‘RNA’ assay.</p>
<pre class="r"><code>gene_metadata &lt;- so.sample[[&quot;RNA&quot;]]@meta.data
#NB: This is geme metadata, different to the usual *cell* metadata found at so.sample@meta.data
# so.sample[[&#39;RNA&#39;]]  retreives the &#39;RNA&#39; assay.
DT::datatable(head(gene_metadata), width=&#39;100%&#39;)</code></pre>
<div id="htmlwidget-90b70f07a5c348cdd0a6" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-90b70f07a5c348cdd0a6">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["Chrna4","Slc6a1","Cd109","Ldha","Aldoc","Drd1"],["RNAProbe","RNAProbe","RNAProbe","RNAProbe","RNAProbe","RNAProbe"],[20278,41055,8140,35695,188086,8103],[0.4703017371338451,0.9521766356657467,0.1887886448500591,0.8278637196465432,4.362223716863418,0.1879305146461952],[0.5198235818514522,1.07159949326702,0.1891803124853376,0.9039084127058603,5.650448083329423,0.1967343874630507],[1.199130749655941,5.429370269015958,0.2345732441844675,1.546603060680439,82.24698940600416,0.3816672632022888],[0.9334658184326288,2.594828661199474,0.2615785654487797,2.099963124092061,45.11653708903471,0.2750117479755442],[1.284600599162089,2.092381030856228,0.8967601905072201,0.7364905806853842,1.822989854999171,1.387821669481007],[false,false,false,false,false,false],[null,null,null,null,null,null],[0.3660380970613573,0.856705274241409,0.1915861012730928,0.7687852972145114,3.182684024121758,0.1817746721546856],[0.6582367603347562,4.07107526880193,0.2393184720530114,1.187479144729904,24.537390278082,0.2636107689253243],[0.6134307422177253,1.927255109491727,0.2619282154153541,1.710812015747839,17.2546218750321,0.245481816291253],[1.073041690012283,2.112369685129758,0.9136796189502528,0.6941026447087406,1.422076383695691,1.07385049087524],[false,false,false,false,false,false],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>feature<\/th>\n      <th>CodeClass<\/th>\n      <th>total_count_unfiltered<\/th>\n      <th>mean_count_unfiltered<\/th>\n      <th>vf_vst_counts.AD2.AD3.CosMx_mean<\/th>\n      <th>vf_vst_counts.AD2.AD3.CosMx_variance<\/th>\n      <th>vf_vst_counts.AD2.AD3.CosMx_variance.expected<\/th>\n      <th>vf_vst_counts.AD2.AD3.CosMx_variance.standardized<\/th>\n      <th>vf_vst_counts.AD2.AD3.CosMx_variable<\/th>\n      <th>vf_vst_counts.AD2.AD3.CosMx_rank<\/th>\n      <th>vf_vst_counts.AD4.AD5.CosMx_mean<\/th>\n      <th>vf_vst_counts.AD4.AD5.CosMx_variance<\/th>\n      <th>vf_vst_counts.AD4.AD5.CosMx_variance.expected<\/th>\n      <th>vf_vst_counts.AD4.AD5.CosMx_variance.standardized<\/th>\n      <th>vf_vst_counts.AD4.AD5.CosMx_variable<\/th>\n      <th>vf_vst_counts.AD4.AD5.CosMx_rank<\/th>\n      <th>var.features<\/th>\n      <th>var.features.rank<\/th>\n      <th>MoransI_observed<\/th>\n      <th>MoransI_p.value<\/th>\n      <th>moransi.spatially.variable<\/th>\n      <th>moransi.spatially.variable.rank<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7,8,10,11,12,13,14,16,18,19,20,22]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"feature","targets":1},{"name":"CodeClass","targets":2},{"name":"total_count_unfiltered","targets":3},{"name":"mean_count_unfiltered","targets":4},{"name":"vf_vst_counts.AD2.AD3.CosMx_mean","targets":5},{"name":"vf_vst_counts.AD2.AD3.CosMx_variance","targets":6},{"name":"vf_vst_counts.AD2.AD3.CosMx_variance.expected","targets":7},{"name":"vf_vst_counts.AD2.AD3.CosMx_variance.standardized","targets":8},{"name":"vf_vst_counts.AD2.AD3.CosMx_variable","targets":9},{"name":"vf_vst_counts.AD2.AD3.CosMx_rank","targets":10},{"name":"vf_vst_counts.AD4.AD5.CosMx_mean","targets":11},{"name":"vf_vst_counts.AD4.AD5.CosMx_variance","targets":12},{"name":"vf_vst_counts.AD4.AD5.CosMx_variance.expected","targets":13},{"name":"vf_vst_counts.AD4.AD5.CosMx_variance.standardized","targets":14},{"name":"vf_vst_counts.AD4.AD5.CosMx_variable","targets":15},{"name":"vf_vst_counts.AD4.AD5.CosMx_rank","targets":16},{"name":"var.features","targets":17},{"name":"var.features.rank","targets":18},{"name":"MoransI_observed","targets":19},{"name":"MoransI_p.value","targets":20},{"name":"moransi.spatially.variable","targets":21},{"name":"moransi.spatially.variable.rank","targets":22}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
<p>The whole gene-metadata includes other columns, and in fact the
columns we are interested in only have values for the ‘variable’ genes
that we tested. So, make a summary table with just the relevant
data.</p>
<pre class="r"><code>gene_metadata_morans &lt;- 
  filter(gene_metadata, !is.na(moransi.spatially.variable.rank)) %&gt;%
  select(feature, 
         MoransI_observed, MoransI_p.value, moransi.spatially.variable,moransi.spatially.variable.rank) %&gt;% 
  arrange(moransi.spatially.variable.rank)
  
DT::datatable(gene_metadata_morans, width = &#39;100%&#39;)</code></pre>
<div id="htmlwidget-fcbdfde5eeb925643261" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-fcbdfde5eeb925643261">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10"],["Ptgds","Penk","Drd4","Lilra5","Vtn","Acta2","Cnr1","Ttr","Sst","Npy"],[0.3229695055639554,0.1590561524926949,0.08720462499499981,0.08718933646160711,0.05282752140827351,0.02569336530082773,0.02362886716393527,0.01266743839515891,0.04844601413882734,0.01902939103935176],[0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.001951219512195122,0.001951219512195122],[true,true,true,true,true,true,true,true,true,true],[1,2,3,4,5,6,7,8,9,10]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>feature<\/th>\n      <th>MoransI_observed<\/th>\n      <th>MoransI_p.value<\/th>\n      <th>moransi.spatially.variable<\/th>\n      <th>moransi.spatially.variable.rank<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"feature","targets":1},{"name":"MoransI_observed","targets":2},{"name":"MoransI_p.value","targets":3},{"name":"moransi.spatially.variable","targets":4},{"name":"moransi.spatially.variable.rank","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="plot-results" class="section level3">
<h3>Plot results</h3>
<p>We can pull out the most significant genes from that table.</p>
<pre class="r"><code>top_genes = gene_metadata_morans$feature[1:3]
top_genes</code></pre>
<pre><code>[1] &quot;Ptgds&quot; &quot;Penk&quot;  &quot;Drd4&quot; </code></pre>
<p>Here we are plotting the top 3 genes on that slide. Each has a
different but clear reasons for being spatially restricted. Ptgds and
Penk seem to be restricted to specific regions of the tissue. Wherease
Drd4 seems to have high expression in a subset of cells - its proximity
to itself also triggers the significance n the moransI test.</p>
<p>NB: Genes without any sort of spatial pattern (e.g negative controls)
might still have some sort of moransI test significance - since they’re
still restricted to the tissue itself, it isn’t random.</p>
<pre class="r"><code>ImageDimPlot(so.sample,  fov = &quot;AD2.AD3.CosMx&quot;,
             molecules = top_genes, 
             group.by = &#39;sample&#39;, cols = c(&quot;grey30&quot;), # Make all cells grey.
             boundaries = &quot;segmentation&quot;,
             border.color = &#39;black&#39;, axes = T, crop=TRUE)</code></pre>
<p><img src="figure/e_spatiallyRestrictedGenes.Rmd/unnamed-chunk-10-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/0ce06a8fe9e3d0354b4cca882c8978ff95d0f1bf/docs/figure/e_spatiallyRestrictedGenes.Rmd/unnamed-chunk-10-1.png" target="_blank">0ce06a8</a>
</td>
<td>
swbioinf
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<!-- ```{r} -->
<!-- # get averages of these 3 genes in  -->
<!-- norm_matrix_means <- FetchData(so.sample, c('cluster_name', 'Ptgds', 'Penk', 'Drd4')) %>% -->
<!--   group_by(cluster_name) %>% -->
<!--   summarise(Ptgds = mean(Ptgds), -->
<!--             Penk  = mean(Penk), -->
<!--             Drd4  = mean(Drd4)) %>% -->
<!--   column_to_rownames(var = 'cluster_name') %>% # turn it into a matrix -->
<!--   as.matrix() -->
<!-- ``` -->
<!-- ```{r} -->
<!-- heatmap(norm_matrix_means, -->
<!--         scale = "none",     # plot normalised expression, no scaling. -->
<!--         Rowv = NA, Colv=NA, # no heirachical clustering -->
<!--         cexCol = 1.5,       # adjust gene name sizes  -->
<!--         margin = c(5,10)    # make window wide enough for labels. -->
<!--         ) -->
<!-- ``` -->
</div>
<div id="run-across-all-samples" class="section level2">
<h2>Run across all samples</h2>
<p>Realistically, we would want to test all samples. Here we run the
test on each tissue sample separately.</p>
<pre class="r"><code>samples &lt;- levels(so@meta.data$sample)
results_list &lt;- list()

for (the_sample in samples) {
  
  so.sample &lt;- subset( so, subset= sample == the_sample)
  
  # Again, this should be: 
  #so.sample &lt;- FindSpatiallyVariableFeatures(
  so.sample &lt;- FindSpatiallyVariableFeatures.Seurat_EDITED(
    so.sample, 
    assay = &quot;RNA&quot;, 
    features = VariableFeatures(so.sample),    
    selection.method = &quot;moransi&quot;,
    layer = &quot;counts&quot;
  )

  gene_metadata &lt;- so.sample[[&quot;RNA&quot;]]@meta.data
  results &lt;- 
  select(gene_metadata, 
         feature, 
         MoransI_observed, 
         MoransI_p.value, 
         moransi.spatially.variable,
         moransi.spatially.variable.rank) %&gt;% 
    filter(!is.na(moransi.spatially.variable.rank)) %&gt;% # only tested
    arrange(moransi.spatially.variable.rank) %&gt;%
    mutate(sample = the_sample) %&gt;%
    select(sample, everything())

  
  results_list[[the_sample]] &lt;- results
}</code></pre>
<pre><code>[1] &quot;&gt;&gt;&gt;&gt; USING EDITED FUNCTION!!!! &lt;&lt;&lt;&quot;
[1] &quot;&gt;&gt;&gt;&gt; USING EDITED FUNCTION!!!! &lt;&lt;&lt;&quot;
[1] &quot;&gt;&gt;&gt;&gt; USING EDITED FUNCTION!!!! &lt;&lt;&lt;&quot;
[1] &quot;&gt;&gt;&gt;&gt; USING EDITED FUNCTION!!!! &lt;&lt;&lt;&quot;</code></pre>
<pre class="r"><code>results_all &lt;- bind_rows(results_list)</code></pre>
<p>Display results for variable genes</p>
<pre class="r"><code>DT::datatable(results_all, width=&#39;100%&#39;) </code></pre>
<div id="htmlwidget-423a02c40977061eb4d3" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-423a02c40977061eb4d3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40"],["sample1","sample1","sample1","sample1","sample1","sample1","sample1","sample1","sample1","sample1","sample2","sample2","sample2","sample2","sample2","sample2","sample2","sample2","sample2","sample2","sample3","sample3","sample3","sample3","sample3","sample3","sample3","sample3","sample3","sample3","sample4","sample4","sample4","sample4","sample4","sample4","sample4","sample4","sample4","sample4"],["Ptgds","Penk","Drd4","Lilra5","Vtn","Acta2","Cnr1","Ttr","Sst","Npy","Ttr","Penk","Ptgds","Lilra5","Drd4","Acta2","Cnr1","Sst","Npy","Vtn","Ttr","Penk","Ptgds","Lilra5","Drd4","Acta2","Sst","Vtn","Cnr1","Npy","Ttr","Ptgds","Drd4","Penk","Lilra5","Acta2","Sst","Npy","Cnr1","Vtn"],[0.3229695055639554,0.1590561524926949,0.08720462499499981,0.08718933646160711,0.05282752140827351,0.02569336530082773,0.02362886716393527,0.01266743839515891,0.04844601413882734,0.01902939103935176,0.3728454799388959,0.1466038049287352,0.1224925545450483,0.09091210714592868,0.08363449281579936,0.03678956364445252,0.0277001335719391,0.02115759343775087,0.01988063216026358,0.01278082783592561,0.5195494965118608,0.1511118241235551,0.1425199586828255,0.09155962979698748,0.08776290200400122,0.05324483809413939,0.0414617262356598,0.03179844426917366,0.03079903927019299,0.02070981737900819,0.6042436378674948,0.1254127115631302,0.1087605254637382,0.1034361004316954,0.1017928004816937,0.05254541191339321,0.033958104426764,0.03103475845295824,0.0243497538943351,0.02054798872420385],[0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.001951219512195122,0.001951219512195122,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561],[true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],[1,2,3,4,5,6,7,8,9,10,1,2,3,4,5,6,7,8,9,10,1,2,3,4,5,6,7,8,9,10,1,2,3,4,5,6,7,8,9,10]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>feature<\/th>\n      <th>MoransI_observed<\/th>\n      <th>MoransI_p.value<\/th>\n      <th>moransi.spatially.variable<\/th>\n      <th>moransi.spatially.variable.rank<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[3,4,6]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"feature","targets":2},{"name":"MoransI_observed","targets":3},{"name":"MoransI_p.value","targets":4},{"name":"moransi.spatially.variable","targets":5},{"name":"moransi.spatially.variable.rank","targets":6}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
<p>Ttr had a higher MoransI score in sample4 than sample1. Plotting its
distribution in both demonstrates the difference - a very high
expression region in sample4.</p>
<pre class="r"><code>ImageDimPlot(
  subset( so, subset = sample == &#39;sample1&#39;),  
             fov          = &quot;AD2.AD3.CosMx&quot;,
             molecules    = &#39;Ttr&#39;, 
             group.by     = &#39;sample&#39;, cols = c(&quot;grey30&quot;), # Make all cells grey.
             boundaries   = &quot;segmentation&quot;,
             border.color = &#39;black&#39;, axes = T, crop=TRUE)</code></pre>
<p><img src="figure/e_spatiallyRestrictedGenes.Rmd/unnamed-chunk-13-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ImageDimPlot(
  subset( so, subset = sample == &#39;sample4&#39;),  
             fov          = &quot;AD4.AD5.CosMx&quot;, # note the slide it is on.
             molecules    = &#39;Ttr&#39;, 
             group.by     = &#39;sample&#39;, cols = c(&quot;grey30&quot;), # Make all cells grey.
             boundaries   = &quot;segmentation&quot;,
             border.color = &#39;black&#39;, axes = T, crop=TRUE)</code></pre>
<p><img src="figure/e_spatiallyRestrictedGenes.Rmd/unnamed-chunk-13-2.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="code-snippet" class="section level1">
<h1>Code Snippet</h1>
<p>Assumes that tissue samples are in a metadata column data called
‘sample’. If there are multiple slides, it may be neccessary to call
joinlayers.</p>
<pre class="r"><code>library(Seurat)
library(tidyverse)
library(DT)

# Load edited function, see https://github.com/satijalab/seurat/issues/8226
# Available here: https://github.com/swbioinf/spatialsnippets/blob/main/code/spatially_variable_features_code.R
source(&quot;spatially_variable_features_code.R&quot;)


# If not alread run, find variable features
num_variable_features = 1000 # Choose based on likely results and acceptable runtime
so &lt;-  FindVariableFeatures(so, nfeatures=num_variable_features)


# Record moransI results fore ach sample one by one.
samples &lt;- levels(so@meta.data$sample)
results_list &lt;- list()
for (the_sample in samples) {
  
  so.sample &lt;- subset( so, subset= sample == the_sample)
  
  # Again, this should be: 
  #so.sample &lt;- FindSpatiallyVariableFeatures(
  so.sample &lt;- FindSpatiallyVariableFeatures.Seurat_EDITED(
    so.sample, 
    assay = &quot;RNA&quot;, 
    features = VariableFeatures(so.sample),    
    selection.method = &quot;moransi&quot;,
    layer = &quot;counts&quot;
  )
  
  # Format output table
  gene_metadata &lt;- so.sample[[&quot;RNA&quot;]]@meta.data
  results &lt;- 
  select(gene_metadata, 
         feature, 
         MoransI_observed, 
         MoransI_p.value, 
         moransi.spatially.variable,
         moransi.spatially.variable.rank) %&gt;% 
    filter(!is.na(moransi.spatially.variable.rank)) %&gt;% # only tested
    arrange(moransi.spatially.variable.rank) %&gt;%
    mutate(sample = the_sample) %&gt;%
    select(sample, everything())

  
  results_list[[the_sample]] &lt;- results
}

# Collect output result 
results_all &lt;- bind_rows(results_list)</code></pre>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<pre class="r"><code>DT::datatable(results_all, width=&#39;100%&#39;)</code></pre>
<div id="htmlwidget-07a7c656fb218458b363" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-07a7c656fb218458b363">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40"],["sample1","sample1","sample1","sample1","sample1","sample1","sample1","sample1","sample1","sample1","sample2","sample2","sample2","sample2","sample2","sample2","sample2","sample2","sample2","sample2","sample3","sample3","sample3","sample3","sample3","sample3","sample3","sample3","sample3","sample3","sample4","sample4","sample4","sample4","sample4","sample4","sample4","sample4","sample4","sample4"],["Ptgds","Penk","Drd4","Lilra5","Vtn","Acta2","Cnr1","Ttr","Sst","Npy","Ttr","Penk","Ptgds","Lilra5","Drd4","Acta2","Cnr1","Sst","Npy","Vtn","Ttr","Penk","Ptgds","Lilra5","Drd4","Acta2","Sst","Vtn","Cnr1","Npy","Ttr","Ptgds","Drd4","Penk","Lilra5","Acta2","Sst","Npy","Cnr1","Vtn"],[0.3229695055639554,0.1590561524926949,0.08720462499499981,0.08718933646160711,0.05282752140827351,0.02569336530082773,0.02362886716393527,0.01266743839515891,0.04844601413882734,0.01902939103935176,0.3728454799388959,0.1466038049287352,0.1224925545450483,0.09091210714592868,0.08363449281579936,0.03678956364445252,0.0277001335719391,0.02115759343775087,0.01988063216026358,0.01278082783592561,0.5195494965118608,0.1511118241235551,0.1425199586828255,0.09155962979698748,0.08776290200400122,0.05324483809413939,0.0414617262356598,0.03179844426917366,0.03079903927019299,0.02070981737900819,0.6042436378674948,0.1254127115631302,0.1087605254637382,0.1034361004316954,0.1017928004816937,0.05254541191339321,0.033958104426764,0.03103475845295824,0.0243497538943351,0.02054798872420385],[0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.001951219512195122,0.001951219512195122,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561,0.000975609756097561],[true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],[1,2,3,4,5,6,7,8,9,10,1,2,3,4,5,6,7,8,9,10,1,2,3,4,5,6,7,8,9,10,1,2,3,4,5,6,7,8,9,10]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>feature<\/th>\n      <th>MoransI_observed<\/th>\n      <th>MoransI_p.value<\/th>\n      <th>moransi.spatially.variable<\/th>\n      <th>moransi.spatially.variable.rank<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[3,4,6]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"feature","targets":2},{"name":"MoransI_observed","targets":3},{"name":"MoransI_p.value","targets":4},{"name":"moransi.spatially.variable","targets":5},{"name":"moransi.spatially.variable.rank","targets":6}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
<ul>
<li><strong>sample</strong>: <em>(not a default column, added by
code)</em>: What tissue sample the test was run on.</li>
<li><strong>feature</strong> : The gene being tested</li>
<li><strong>MoransI_observed</strong> : The moransI statistic
calculated. Higher values indicate more spatial correlation, 0 is
completely random, and negative values indicate anti-correlation (ie
repulsion).</li>
<li><strong>MoransI_p.value</strong> : P-value for the moransI
test.</li>
<li><strong>moransi.spatially.variable</strong> : Is this gene spatially
restricted? True or false value.</li>
<li><strong>moransi.spatially.variable.rank</strong> : Ranking of the
genes by spatial correlation, where 1 is the most distincly spatially
restricted.</li>
</ul>
</div>
<div id="more-information" class="section level1">
<h1>More information</h1>
<ul>
<li><a
href="https://www.sciencedirect.com/science/article/pii/S2211124724005448?via%3Dihub">Microglia-astrocyte
crosstalk in the amyloid plaque niche of an Alzheimer’s disease mouse
model, as revealed by spatial transcriptomics</a>: Data used in this
example. <span class="citation">(Mallach et al. 2024)</span></li>
<li><a href="https://en.wikipedia.org/wiki/Moran%27s_I">MoransI
wikipedia</a>: What is MoransI test, with pictures.</li>
<li><a
href="https://satijalab.org/seurat/articles/spatial_vignette">Seurat
Spatially variable features</a>: This is actually the sequencing based
technology vignette (for visium data), but it covers the
FindSpatiallyVariableFeatures function</li>
<li><a
href="https://github.com/satijalab/seurat/issues/8226">FindSpatiallyVariableFeatures()
Bug report</a> : Link to the bug report on the Seurat repo in github.
Can check the status of this issue here.</li>
<li><a
href="https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#local-spatial-statistics-of-marker-genes">Voyager
toolkit - spatial statistics for cosmx vignette</a> : Here is a more
focussed and detailed vignette on calculating spatial statistics for
gene expression using the Voyager package. It does not use Seurat
objects, it is rather using the bioconductor-compatible
SpatialExperiment class.</li>
</ul>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-mallachMicrogliaastrocyteCrosstalkAmyloid2024"
class="csl-entry">
Mallach, Anna, Magdalena Zielonka, Veerle van Lieshout, Yanru An, Jia
Hui Khoo, Marisa Vanheusden, Wei-Ting Chen, et al. 2024.
<span>“Microglia-Astrocyte Crosstalk in the Amyloid Plaque Niche of an
<span>Alzheimer</span>’s Disease Mouse Model, as Revealed by Spatial
Transcriptomics.”</span> <em>Cell Reports</em> 43 (6): 114216. <a
href="https://doi.org/10.1016/j.celrep.2024.114216">https://doi.org/10.1016/j.celrep.2024.114216</a>.
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] DT_0.33            lubridate_1.9.3    forcats_1.0.0      stringr_1.5.1     
 [5] dplyr_1.1.4        purrr_1.0.2        readr_2.1.5        tidyr_1.3.1       
 [9] tibble_3.2.1       ggplot2_3.5.0      tidyverse_2.0.0    Seurat_5.1.0      
[13] SeuratObject_5.0.2 sp_2.1-3           workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.16.0      jsonlite_1.8.8        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   farver_2.1.1          
  [7] rmarkdown_2.26         fs_1.6.3               vctrs_0.6.5           
 [10] ROCR_1.0-11            spatstat.explore_3.2-7 htmltools_0.5.8.1     
 [13] sass_0.4.9             sctransform_0.4.1      parallelly_1.37.1     
 [16] KernSmooth_2.23-22     bslib_0.7.0            htmlwidgets_1.6.4     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         
 [22] zoo_1.8-12             cachem_1.0.8           whisker_0.4.1         
 [25] igraph_2.0.3           mime_0.12              lifecycle_1.0.4       
 [28] pkgconfig_2.0.3        Matrix_1.6-5           R6_2.5.1              
 [31] fastmap_1.1.1          fitdistrplus_1.1-11    future_1.33.2         
 [34] shiny_1.8.1.1          digest_0.6.35          colorspace_2.1-0      
 [37] patchwork_1.2.0        ps_1.7.6               rprojroot_2.0.4       
 [40] tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1         
 [43] crosstalk_1.2.1        labeling_0.4.3         RcppZiggurat_0.1.6    
 [46] progressr_0.14.0       timechange_0.3.0       fansi_1.0.6           
 [49] spatstat.sparse_3.0-3  httr_1.4.7             polyclip_1.10-6       
 [52] abind_1.4-5            compiler_4.3.2         proxy_0.4-27          
 [55] withr_3.0.0            DBI_1.2.2              fastDummies_1.7.3     
 [58] highr_0.10             MASS_7.3-60.0.1        classInt_0.4-10       
 [61] units_0.8-5            tools_4.3.2            lmtest_0.9-40         
 [64] httpuv_1.6.15          future.apply_1.11.2    Rfast2_0.1.5.2        
 [67] goftest_1.2-3          glue_1.7.0             callr_3.7.6           
 [70] nlme_3.1-164           promises_1.2.1         sf_1.0-16             
 [73] grid_4.3.2             Rtsne_0.17             getPass_0.2-4         
 [76] cluster_2.1.6          reshape2_1.4.4         generics_0.1.3        
 [79] gtable_0.3.4           spatstat.data_3.0-4    tzdb_0.4.0            
 [82] class_7.3-22           hms_1.1.3              data.table_1.15.4     
 [85] utf8_1.2.4             spatstat.geom_3.2-9    RcppAnnoy_0.0.22      
 [88] ggrepel_0.9.5          RANN_2.6.1             pillar_1.9.0          
 [91] spam_2.10-0            RcppHNSW_0.6.0         later_1.3.2           
 [94] splines_4.3.2          lattice_0.22-6         renv_1.0.5            
 [97] survival_3.5-8         deldir_2.0-4           tidyselect_1.2.1      
[100] Rnanoflann_0.0.3       miniUI_0.1.1.1         pbapply_1.7-2         
[103] knitr_1.45             git2r_0.33.0           gridExtra_2.3         
[106] scattermore_1.2        xfun_0.43              matrixStats_1.2.0     
[109] stringi_1.8.3          lazyeval_0.2.2         yaml_2.3.8            
[112] evaluate_0.23          codetools_0.2-20       BiocManager_1.30.22   
[115] cli_3.6.2              RcppParallel_5.1.7     uwot_0.1.16           
[118] xtable_1.8-4           reticulate_1.35.0      munsell_0.5.1         
[121] processx_3.8.4         jquerylib_0.1.4        Rcpp_1.0.12           
[124] globals_0.16.3         spatstat.random_3.2-3  png_0.1-8             
[127] Rfast_2.1.0            parallel_4.3.2         dotCall64_1.1-1       
[130] listenv_0.9.1          viridisLite_0.4.2      e1071_1.7-14          
[133] scales_1.3.0           ggridges_0.5.6         leiden_0.4.3.1        
[136] rlang_1.1.3            cowplot_1.1.3         </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
