<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Home</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/core-js-2.5.3/shim.min.js"></script>
<script src="site_libs/react-18.2.0/react.min.js"></script>
<script src="site_libs/react-18.2.0/react-dom.min.js"></script>
<script src="site_libs/reactwidget-2.0.0/react-tools.js"></script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/reactable-0.4.4/reactable.css" rel="stylesheet" />
<script src="site_libs/reactable-binding-0.4.4/reactable.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Spatial Sampler</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tests</a>
</li>
<li>
  <a href="index_data.html">Data</a>
</li>
<li>
  <a href="contributing.html">Contributing</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/swbioinf/spatialsnippets">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Home</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-21
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>spatialsnippets/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20231017code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20231017)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20231017code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20231017)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it
difficult for you and others to run your code on a different machine.
Change the absolute path(s) below to the suggested relative path(s) to
make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
/home/s.williams/projects/spatialsnippets/spatialsnippets/data/howto_info.tsv
</td>
<td style="text-align:left;">
data/howto_info.tsv
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree074825743f8d7b5ad0d33f4af584b44bf2d65d73targetblank0748257a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/swbioinf/spatialsnippets/tree/074825743f8d7b5ad0d33f4af584b44bf2d65d73" target="_blank">0748257</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomswbioinfspatialsnippetstree074825743f8d7b5ad0d33f4af584b44bf2d65d73targetblank0748257a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/swbioinf/spatialsnippets/tree/074825743f8d7b5ad0d33f4af584b44bf2d65d73" target="_blank">0748257</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/contributions.nb.html
    Ignored:    analysis/d_cosmxIBD_sfe.nb.html
    Ignored:    analysis/d_test.nb.html
    Ignored:    analysis/e_neighbourcellchanges.nb.html
    Ignored:    analysis/e_spatiallyRestrictedGenesVoyager.nb.html
    Ignored:    analysis/g_toolkits.nb.html
    Ignored:    analysis/glossary.nb.html
    Ignored:    renv/library/
    Ignored:    renv/staging/

Untracked files:
    Untracked:  analysis/e_DEPseudobulk_insitu_sfe.Rmd
    Untracked:  data/howto_info.tsv
    Untracked:  other/

Unstaged changes:
    Modified:   analysis/d_cosmxIBD_sfe.Rmd
    Modified:   analysis/e_DEPseudobulk_insitu.Rmd
    Modified:   renv.lock
    Modified:   renv/activate.R
    Modified:   spatialsnippets.Rproj

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/index_update.Rmd</code>) and HTML
(<code>docs/index_update.html</code>) files. If you’ve configured a
remote Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/swbioinf/spatialsnippets/blob/074825743f8d7b5ad0d33f4af584b44bf2d65d73/analysis/index_update.Rmd" target="_blank">0748257</a>
</td>
<td>
swbioinf
</td>
<td>
2025-03-21
</td>
<td>
wflow_publish("analysis/index_update.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>These examples are a sampler of tests that can be performed on an
in-situ single cell spatial datasets. Initial examples use cosMX, though
approaches should be valid on other platforms like Xenium.</p>
<p>Further details on the papers, tools, statistics and methods are
linked from each page.</p>
<p>Each example presents just one way of doing things (and currently
only with Seurat) - there are many, many other approaches! These are
intended as a starting point.</p>
<p>These examples assume that the data has been loaded, preprocessed
(e.g. QC-ed, UMAP) and annotated (e.g. sample names, cell types, cell
niches). For information on how each dataset was prepared, see the data
tab.</p>
<div class="reactable html-widget html-fill-item" id="htmlwidget-8839f08f82529a5247be" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8839f08f82529a5247be">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"image":["assets/cluster_prop_diff.png","assets/insitu_spatial_pseudobulk_de_fovs.png","assets/spatiallyrestricted.png","assets/de_1v1.png"],"howto":["Test for different cell type composition by group","Differential Expression by celltypes, between groups","Find genes with spatial pattern","Differential expression without replicates with scRNAseq"],"Description":["Looking at proportions of celltypes, within in-situ spatial single cell datasets ","Looking at differential expression, celltype by celltype, within in-situ spatial single cell datasets","Find genes that are expressed in distinct patterns on the samples","For scRNAseq data, compare gene expression cluster by cluster, between two samples, where no replicates are available."],"Technology":["single cell spatial transcriptomics","single cell spatial transcriptomics","single cell spatial transcriptomics","single cell RNAseq"],"Toolkit":["seurat","seurat","seurat","seurat"],"url":["e_CompositionChange.html","e_DEPseudobulk_insitu.html","e_spatiallyRestrictedGenes.html","e_DEWithoutReps.html"],"SpecificTech":["cosmx","cosmx","cosmx","10X chromium"]},"columns":[{"id":"image","name":"","type":"character","cell":[{"name":"Fragment","attribs":[],"children":[{"name":"a","attribs":{"style":{"font-weight":"600","color":"black"},"href":"e_CompositionChange.html","target":"_blank"},"children":["Test for different cell type composition by group"]},{"name":"div","attribs":{"style":{"display":"inline-block"}},"children":[{"name":"a","attribs":{"href":"e_CompositionChange.html"},"children":[{"name":"img","attribs":{"src":"assets/cluster_prop_diff.png","alt":"assets/cluster_prop_diff.png"},"children":[]}]}]}]},{"name":"Fragment","attribs":[],"children":[{"name":"a","attribs":{"style":{"font-weight":"600","color":"black"},"href":"e_DEPseudobulk_insitu.html","target":"_blank"},"children":["Differential Expression by celltypes, between groups"]},{"name":"div","attribs":{"style":{"display":"inline-block"}},"children":[{"name":"a","attribs":{"href":"e_DEPseudobulk_insitu.html"},"children":[{"name":"img","attribs":{"src":"assets/insitu_spatial_pseudobulk_de_fovs.png","alt":"assets/insitu_spatial_pseudobulk_de_fovs.png"},"children":[]}]}]}]},{"name":"Fragment","attribs":[],"children":[{"name":"a","attribs":{"style":{"font-weight":"600","color":"black"},"href":"e_spatiallyRestrictedGenes.html","target":"_blank"},"children":["Find genes with spatial pattern"]},{"name":"div","attribs":{"style":{"display":"inline-block"}},"children":[{"name":"a","attribs":{"href":"e_spatiallyRestrictedGenes.html"},"children":[{"name":"img","attribs":{"src":"assets/spatiallyrestricted.png","alt":"assets/spatiallyrestricted.png"},"children":[]}]}]}]},{"name":"Fragment","attribs":[],"children":[{"name":"a","attribs":{"style":{"font-weight":"600","color":"black"},"href":"e_DEWithoutReps.html","target":"_blank"},"children":["Differential expression without replicates with scRNAseq"]},{"name":"div","attribs":{"style":{"display":"inline-block"}},"children":[{"name":"a","attribs":{"href":"e_DEWithoutReps.html"},"children":[{"name":"img","attribs":{"src":"assets/de_1v1.png","alt":"assets/de_1v1.png"},"children":[]}]}]}]}],"minWidth":250},{"id":"howto","name":"howto","type":"character","show":false},{"id":"Description","name":"Description","type":"character","minWidth":200},{"id":"Technology","name":"Technology","type":"character","cell":[{"name":"div","attribs":{},"children":[{"name":"div","attribs":{},"children":["single cell spatial transcriptomics"]},{"name":"div","attribs":{"style":{"font-style":"italic"}},"children":["cosmx"]}]},{"name":"div","attribs":{},"children":[{"name":"div","attribs":{},"children":["single cell spatial transcriptomics"]},{"name":"div","attribs":{"style":{"font-style":"italic"}},"children":["cosmx"]}]},{"name":"div","attribs":{},"children":[{"name":"div","attribs":{},"children":["single cell spatial transcriptomics"]},{"name":"div","attribs":{"style":{"font-style":"italic"}},"children":["cosmx"]}]},{"name":"div","attribs":{},"children":[{"name":"div","attribs":{},"children":["single cell RNAseq"]},{"name":"div","attribs":{"style":{"font-style":"italic"}},"children":["10X chromium"]}]}]},{"id":"Toolkit","name":"Toolkit","type":"character","minWidth":50},{"id":"url","name":"","type":"character","cell":[{"name":"a","attribs":{"href":"e_CompositionChange.html","target":"_blank"},"children":["▶️"]},{"name":"a","attribs":{"href":"e_DEPseudobulk_insitu.html","target":"_blank"},"children":["▶️"]},{"name":"a","attribs":{"href":"e_spatiallyRestrictedGenes.html","target":"_blank"},"children":["▶️"]},{"name":"a","attribs":{"href":"e_DEWithoutReps.html","target":"_blank"},"children":["▶️"]}],"width":50},{"id":"SpecificTech","name":"SpecificTech","type":"character","searchable":true,"show":false}],"searchable":true,"dataKey":"b98330955524ad60717b60488f192879"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
<div id="refs">

</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
[1] dplyr_1.1.4       htmltools_0.5.8.1 reactable_0.4.4   workflowr_1.7.1  

loaded via a namespace (and not attached):
 [1] jsonlite_1.8.9      compiler_4.4.0      BiocManager_1.30.25
 [4] renv_1.0.5          promises_1.3.2      tidyselect_1.2.1   
 [7] Rcpp_1.0.14         stringr_1.5.1       git2r_0.33.0       
[10] callr_3.7.6         later_1.4.1         jquerylib_0.1.4    
[13] yaml_2.3.10         fastmap_1.2.0       R6_2.5.1           
[16] generics_0.1.3      knitr_1.49          htmlwidgets_1.6.4  
[19] tibble_3.2.1        rprojroot_2.0.4     bslib_0.9.0        
[22] pillar_1.10.1       rlang_1.1.5         cachem_1.1.0       
[25] stringi_1.8.4       reactR_0.6.1        httpuv_1.6.15      
[28] xfun_0.50           getPass_0.2-4       fs_1.6.5           
[31] sass_0.4.9          cli_3.6.3           withr_3.0.2        
[34] magrittr_2.0.3      crosstalk_1.2.1     ps_1.8.1           
[37] digest_0.6.37       processx_3.8.5      rstudioapi_0.17.1  
[40] lifecycle_1.0.4     vctrs_0.6.5         evaluate_1.0.3     
[43] glue_1.8.0          whisker_0.4.1       rmarkdown_2.29     
[46] httr_1.4.7          tools_4.4.0         pkgconfig_2.0.3    </code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
