---
title: "Dataset: IBD cosMX GarridoTrigo2023"
author: "Sarah Williams"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
bibliography: /home/s.williams/zotero_export/MyLibrary.bib
---


Data is from paper [_Macrophage and neutrophil heterogeneity at single-cell spatial resolution in human inflammatory bowel disease_](https://www.ncbi.nlm.nih.gov/pubmed/37495570)
from Garrido-Trigo et al 2023, [@garrido-trigoMacrophageNeutrophilHeterogeneity2023]

The study included 9 cosmx slides of colonic biopsies

* 3x HC - Healthy controls
* 3x UC - Ulcerative colitis
* 3x CD - Chrones's disease


Fastantically - not only have they made their raw and annotated data available, but have also shared their analysis code; https://github.com/HelenaLC/CosMx-SMI-IBD


They have also shared browseable interface here: https://servidor2-ciberehd.upc.es/external/garrido/app/


# Libraries

```{r}

library(tidyverse)
library(patchwork)
library(SpatialFeatureExperiment)
library(alabaster.sfe) # SaveObject
library(scran)  
library(scater)
library(scuttle)
library(bluster) # clustering parameters, needed for passing htreads to NNGraphParam
library(BiocParallel)


#library(alabaster.sfe)
# Requires v1.9.3 to use alabaster.sfe to save object.
# alabaster.sfe not yet in bioconductor production March 2025
#renv::install('pachterlab/SpatialFeatureExperiment@devel')
#renv::install('pachterlab/alabaster.sfe')

```

# Config

```{r}
dataset_dir      <- '~/projects/spatialsnippets/datasets/'
project_data_dir <- file.path(dataset_dir,'GSE234713_IBDcosmx_GarridoTrigo2023')
sample_dir            <- file.path(project_data_dir, "raw_data_for_sfe/") 
annotation_file       <- file.path(project_data_dir,"GSE234713_CosMx_annotation.csv.gz")
data_dir              <- file.path(project_data_dir, "processed_data/") 


sfe_01_loaded <- file.path(data_dir, "GSE234713_CosMx_IBD_sfe_01_loaded") 

# config
min_count_per_cell <- 50 
min_detected_genes_per_cell <- 20 # from paper.
max_pc_negs        <- 1.5
max_avg_neg        <- 0.5

sample_codes <- c(HC="Healthy controls",UC="Ulcerative colitis",CD="Crohn's disease")

```

# Data download

From the project description:

Raw data on GEO here; https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE234713

_Inflammatory bowel diseases (IBDs) including ulcerative colitis (UC) and Crohnâ€™s disease (CD) are chronic inflammatory diseases with increasing worldwide prevalence that show a perplexing heterogeneity in manifestations and response to treatment. We applied spatial transcriptomics at single-cell resolution (CosMx Spatial Molecular Imaging) to human inflamed and uninflamed intestine._


The following files were download from GEO

* GSE234713_CosMx_annotation.csv.gz
* GSE234713_CosMx_normalized_matrix.txt.gz
* GSE234713_RAW.tar: RAW data was downloaded via custom downlod in 3 batches, one per group
  * GSE234713_RAW_CD.tar
  * GSE234713_RAW_HC.tar
  * GSE234713_RAW_UC.tar
* GSE234713_ReadMe_SMI_Data_File.html

```{sh eval=FALSE}
cd raw_data
tar -xf GSE234713_RAW_CD.tar
tar -xf GSE234713_RAW_HC.tar
tar -xf GSE234713_RAW_UC.tar
```

Construct a folder with each slide's info.

NB: Also due to limited temp directory space, I need to gunzip the files (this really shouldn't be needed)
that get read by the efficient DT reading functions. Symptom of that is silently incomplete file reads.
https://www.linkedin.com/pulse/trivial-fix-after-3-hours-debugging-kirill-tsyganov/


```{sh eval=FALSE}
mkdir raw_data_for_sfe
cp raw_data/*tar.gz raw_data_for_sfe/
cd raw_data_for_sfe/
tar -xzf *.tar.gz


cd /home/s.williams/projects/spatialsnippets/datasets/GSE234713_IBDcosmx_GarridoTrigo2023/raw_data_for_sfe
for sample in GSM7473682_HC_a  GSM7473685_UC_a  GSM7473688_CD_a GSM7473683_HC_b  GSM7473686_UC_b  GSM7473689_CD_b GSM7473684_HC_c  GSM7473687_UC_c  GSM7473690_CD_c; do
  echo $sample
  #cd /home/s.williams/projects/spatialsnippets/datasets/GSE234713_IBDcosmx_GarridoTrigo2023/raw_data_for_sfe
  
  mkdir raw_data_for_sfe/${sample}
  cp raw_data/${sample}_* raw_data_for_sfe/${sample}
  gunzip raw_data_for_sfe/${sample}/*
  
  # Separately copy in the polygons files (supplied by authors)
  samplecode=${sample: -4}
  echo $samplecode
  cp /home/s.williams/projects/spatialsnippets/datasets/GSE234713_IBDcosmx_GarridoTrigo2023/polygons/${samplecode}.csv ${sample}/${sample}-polygons.csv
  
  
  
done 


```


Read in each sample as a 'SpatialExperiment' using a function from the *SpatialExperimentIO* package.

We don't have polygon files for this dataset, otherwise would load a a SpatialFeatureExperiment. A spatialFeatureExperiment object *is a* SpatialExperiment object with extra features (ie. it inherits from it).


# Data load

Data is loaded into a SpatialFeatureExperiment object with the following code (long running).

```{r, file='code/d_cosmxIBD_sfe_script.R', eval=FALSE}

```

Load in the saved object
```{r}
sfe <- readObject(sfe_01_loaded)
```


# Basic QC filter

## Min count per cell

For interest, not filtering on total counts.
```{r}
ggplot(colData(sfe), aes(x=total_count, col=sample_id)) +
  geom_density() + 
  scale_x_log10() +
  theme_bw() +
  ggtitle("Counts per cell")
```

## Min detected genes per cell

Distinct genes observed per cell (number of detected genes per cell)

```{r}
ggplot(colData(sfe), aes(x=distinct_genes, col=sample_id)) +
  geom_density() + 
  geom_vline(xintercept = min_detected_genes_per_cell, lty=3) +
  scale_x_log10() +
  theme_bw() +
  ggtitle("Counts per cell")
```


## Percent Negative probes

```{r}
ggplot(colData(sfe), aes(x=pc_neg, col=sample_id)) +
  geom_density() + 
  geom_vline(xintercept = max_pc_negs, lty=3) +
  scale_x_log10() +
  theme_bw() +
  ggtitle("Negative probe composition")
```

```{r}
ggplot(colData(sfe), aes(x=avg_neg, col=sample_id)) +
  geom_density() + 
  geom_vline(xintercept = max_avg_neg, lty=3) +
  theme_bw() +
  ggtitle("Negative probe average")
```


Use bottom right corner;
```{r}
ggplot(colData(sfe), aes(y=avg_neg, x=total_count)) +
  geom_point(pch=3, alpha=0.1) + 
  geom_hline(yintercept = max_avg_neg, lty=3) +
  geom_vline(xintercept = min_count_per_cell, lty=3) +
  scale_x_log10() + 
  theme_bw() +
  ggtitle("Negative probes vs counts")
```


## Apply filteres

From paper: _"Cells with an average negative control count greater than 0.5 and less than 20 detected features were filtered out."_

Applying a different threshold (arbitrarily), results will differ.





## How many cells per sample?

```{r}
table(sfe$tissue_sample)
```




# Basic plots



## UMAP {.tabset .tabset-pills}

### Sample

```{r}
plotUMAP(sfe, colour_by = 'tissue_sample', point_shape='.') +
  guides(colour = guide_legend(override.aes = list(shape=15)))
```


### Condition

```{r}
plotUMAP(sfe, colour_by = 'condition', point_shape='.') +
  guides(colour = guide_legend(override.aes = list(shape=15)))
```



### Condition X Sample

```{r fig.width=12, fig.height=3}
p1 <- plotUMAP(sfe[,sfe$group=='HC'], colour_by = 'tissue_sample', point_shape='.') + ggtitle ("Healthy Controls") + 
  guides(colour = guide_legend(override.aes = list(shape=15)))
p2 <- plotUMAP(sfe[,sfe$group=='UC'], colour_by = 'tissue_sample', point_shape='.') + ggtitle("Ulcerative colitis") +
  guides(colour = guide_legend(override.aes = list(shape=15)))
p3 <- plotUMAP(sfe[,sfe$group=='CD'], colour_by = 'tissue_sample', point_shape='.') + ggtitle("Crohn's disease")+
  guides(colour = guide_legend(override.aes = list(shape=15)))


p1 + p2 + p3
```

### Total counts

```{r}
plotUMAP(sfe, colour_by = 'total_count_log10', point_shape='.') 
```


### Clustered

```{r}
plotUMAP(sfe, colour_by = 'cluster_code', point_shape='.', text_by='cluster_code') +
  guides(colour = guide_legend(override.aes = list(shape=15)))
```


### celltype_subset

Classifications from Garrido-Trigo et al 2023.

```{r}
plotUMAP(sfe, colour_by = 'celltype_subset', text_by='celltype_subset', point_shape='.') +
  guides(colour = guide_legend(override.aes = list(shape=15)))

```

### celltype_SingleR2

Classifications from Garrido-Trigo et al 2023.

```{r fig.width=12}
plotUMAP(sfe, colour_by = 'celltype_SingleR2', point_shape='.') +
  guides(colour = guide_legend(override.aes = list(shape=15)))

```



## Clustering

Double checking the clusters line up with annotaiaon
```{r}
table(sfe$celltype_subset)
table(sfe$celltype_SingleR2)
table(sfe$cluster_code) 
```


Composition. Expect an obvious but imperfect grouping between clustering and cell typing. Will be using annotation for analysis.

```{r}
heatmap(table(sfe$celltype_SingleR2, sfe$celltype_subset))
heatmap(table(sfe$celltype_subset, sfe$cluster_code))
heatmap(table(sfe$celltype_SingleR2,sfe$cluster_code))
```



