---
title: "Dataset: IBD cosMX GarridoTrigo2023"
author: "Sarah Williams"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
bibliography: /home/s.williams/zotero_export/MyLibrary.bib
---


Data is from paper [_Macrophage and neutrophil heterogeneity at single-cell spatial resolution in human inflammatory bowel disease_](https://www.ncbi.nlm.nih.gov/pubmed/37495570)
from Garrido-Trigo et al 2023, [@garrido-trigoMacrophageNeutrophilHeterogeneity2023]

The study included 9 cosmx slides of colonic biopsies

* 3x HC - Healthy controls
* 3x UC - Ulcerative colitis
* 3x CD - Chrones's disease


Fastantically - not only have they made their raw and annotated data available, but have also shared their analysis code; https://github.com/HelenaLC/CosMx-SMI-IBD


They have also shared browseable interface here: https://servidor2-ciberehd.upc.es/external/garrido/app/


# Libraries

```{r}
library(Seurat)
library(SeuratObject)
library(tidyverse)
```

# Data download

From the project description:

Raw data on GEO here; https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE234713

_Inflammatory bowel diseases (IBDs) including ulcerative colitis (UC) and Crohn’s disease (CD) are chronic inflammatory diseases with increasing worldwide prevalence that show a perplexing heterogeneity in manifestations and response to treatment. We applied spatial transcriptomics at single-cell resolution (CosMx Spatial Molecular Imaging) to human inflamed and uninflamed intestine._


The following files were download from GEO

* GSE234713_CosMx_annotation.csv.gz
* GSE234713_CosMx_normalized_matrix.txt.gz
* GSE234713_RAW.tar: RAW data was downloaded via custom downlod in 3 batches, one per group
  * GSE234713_RAW_CD.tar
  * GSE234713_RAW_HC.tar
  * GSE234713_RAW_UC.tar
* GSE234713_ReadMe_SMI_Data_File.html

The polygon files defining the cell outlines were provided separately. Typically these would be included with the other files. 

* CD_a.csv : A .csv file which defines the outline of every cell. One file per slide.




```{sh eval=FALSE}
cd raw_data
tar -xf GSE234713_RAW_CD.tar
tar -xf GSE234713_RAW_HC.tar
tar -xf GSE234713_RAW_UC.tar
```



# Data load


```{r}
dataset_dir      <- '~/projects/spatialsnippets/datasets/'
project_data_dir <- file.path(dataset_dir,'GSE234713_IBDcosmx_GarridoTrigo2023')
sample_dir            <- file.path(project_data_dir, "raw_data_for_sfe/") 
annotation_file       <- file.path(project_data_dir,"GSE234713_CosMx_annotation.csv.gz")
data_dir              <- file.path(project_data_dir, "processed_data/") 


seurat_file_00_raw    <- file.path(data_dir, "GSE234713_CosMx_IBD_seurat_00_raw.RDS")
seurat_file_01_loaded <- file.path(data_dir, "GSE234713_CosMx_IBD_seurat_01_loaded.RDS")

# config
min_count_per_cell <- 100
max_pc_negs        <- 1.5
max_avg_neg        <- 0.5

sample_codes <- c(HC="Healthy controls",UC="Ulcerative colitis",CD="Crohn's disease")

 

```




## Raw data


## Load data

```{r}
# Load LoadNanostring.X for keeping metadata e.t.c
source('/home/s.williams/projects/loadnanostring/code/LoadNanostring.R')


load_sample_into_seurat <- function(the_sample){
  
  # Using a modified version of the LoadNanostring function, which keeps metadata, and loads a litte more effiently for large datasets (which is not this one.)
  # ..../datasets//GSE234713_IBDcosmx_GarridoTrigo2023/raw_data_for_sfe//GSM7473682_HC_a
  so <- LoadNanostring.X(file.path(sample_dir, the_sample), 
                         project = the_sample,
                         assay='RNA',
                         fov=the_sample, tempdir= '~/tmp')
  
  # sample info
  so$individual_code <- factor(substr(so$orig.ident,12,16))
  so$tissue_sample   <- factor(substr(so$orig.ident,12,16))
  so$group     <- factor(substr(the_sample, 12, 13), levels=names(sample_codes))
  so$condition <- factor(as.character(sample_codes[so$group]), levels=sample_codes)
  
  so$fov_name        <- paste0(so$individual_code,"_", str_pad(so$fov, 3, 'left',pad='0'))


  # Put neg probes into their own assay.
  neg_probes <- rownames(so)[grepl(x=rownames(so), pattern="NegPrb")]
  neg_matrix         <- GetAssayData(so,assay = 'RNA', layer = 'counts')[neg_probes,]
  #so[["negprobes"]] <- CreateAssayObject(counts = neg_matrix)
  so[["negprobes"]] <- CreateAssay5Object(counts = neg_matrix)
    
  ## and remove from the main one
  rna_probes  <- rownames(so)[(! rownames(so) %in% neg_probes)]
  so[['RNA']] <- subset( so[['RNA']],features = rna_probes)
  

  return(so)
  
}


```

```{r}
samples <- c('GSM7473682_HC_a','GSM7473683_HC_b','GSM7473684_HC_c',
             'GSM7473685_UC_a','GSM7473686_UC_b','GSM7473687_UC_c',
             'GSM7473688_CD_a','GSM7473689_CD_b','GSM7473690_CD_c')
sample_prefix <- paste0(substr(samples, 12,15))

```


```{r eval=FALSE}

so.list <- lapply(FUN=load_sample_into_seurat, X=samples)

#NB: merge is in SeuratObject packages, but must be called without :: 
options(future.globals.maxSize= 10*1024^3) # 10G.
so.raw <- merge(so.list[[1]], y=so.list[2:length(so.list)], add.cell.ids=sample_prefix)
#Error in getGlobalsAndPackages(expr, envir = envir, globals = globals) : 
#  The total size of the 8 globals exported for future expression (‘FUN()’) is 5.90 GiB.. This exceeds the maximum #allowed size of 500.00 MiB (option 'future.globals.maxSize'). The three largest globals are ‘FUN’ (5.90 GiB of #class ‘function’), ‘p’ (25.22 KiB of class ‘function’) and ‘slot<-’ (1.36 KiB of class ‘function’)
rm(so.list)

# save
saveRDS(so.raw, seurat_file_00_raw)

```

```{r}
# load previously saved.
so.raw <- readRDS(seurat_file_00_raw)
```

# Negative probe handling

```{r}
so.raw$pc_neg <-  ( so.raw$nCount_negprobes / so.raw$nCount_RNA ) * 100

so.raw[["negprobes"]] <- JoinLayers(so.raw[["negprobes"]]) # For caluclating these, need to have the negprobes merged
so.raw$avg_neg <-  colMeans(so.raw[["negprobes"]])   # only defined firsts sample.

```


<!-- ```{r} -->
<!-- # Put neg probes into their own assay. -->
<!-- neg_probes <- rownames(so.raw)[grepl(x=rownames(so.raw), pattern="NegPrb")] -->
<!-- neg_matrix         <- GetAssayData(so.raw,assay = 'RNA', layer = 'counts')[neg_probes,] -->
<!-- so.raw[["negprobes"]] <- CreateAssayObject(counts = neg_matrix) -->

<!-- ## and remove from the main one -->
<!-- rna_probes <- rownames(so)[(! rownames(so.raw) %in% neg_probes)] -->
<!-- so.raw         <- so[rna_probes,] -->
<!-- ``` -->




# Pull in annotation

Most cell annotations. 

```{r}
anno_table <- read_csv(annotation_file)

anno_table <- as.data.frame(anno_table)
rownames(anno_table) <- anno_table$id

head(so.raw@meta.data)
head(anno_table)

so.raw$full_cell_id      <- as.character(rownames(so.raw@meta.data))
so.raw$celltype_subset   <- factor(anno_table[so.raw$full_cell_id,]$subset)
so.raw$celltype_SingleR2 <- factor(anno_table[so.raw$full_cell_id,]$SingleR2)
so.raw$fov_name          <- factor(so.raw$fov_name)
so.raw$group          <- factor(so.raw$group, levels=c("CD","UC","HC"))
so.raw$condition      <- factor(so.raw$condition, levels=c("Crohn's disease",  'Ulcerative colitis', 'Healthy controls'))
table(is.na(so.raw$celltype_subset))
```


# Basic QC filter

## Min count per cell

```{r}
ggplot(so.raw@meta.data, aes(x=nCount_RNA, col=orig.ident)) +
  geom_density() + 
  geom_vline(xintercept = min_count_per_cell, lty=3) +
  scale_x_log10() +
  theme_bw() +
  ggtitle("Counts per cell")
```


## Percent Negative probes

```{r}
ggplot(so.raw@meta.data, aes(x=pc_neg, col=orig.ident)) +
  geom_density() + 
  geom_vline(xintercept = max_pc_negs, lty=3) +
  scale_x_log10() +
  theme_bw() +
  ggtitle("Negative probe composition")
```

```{r}
ggplot(so.raw@meta.data, aes(x=avg_neg, col=orig.ident)) +
  geom_density() + 
  geom_vline(xintercept = max_avg_neg, lty=3) +
  theme_bw() +
  ggtitle("Negative probe average")
```


Use bottom right corner;
```{r}
ggplot(so.raw@meta.data, aes(y=avg_neg, x=nCount_RNA)) +
  geom_point(pch=3, alpha=0.1) + 
  geom_hline(yintercept = max_avg_neg, lty=3) +
  geom_vline(xintercept = min_count_per_cell, lty=3) +
  scale_x_log10() + 
  theme_bw() +
  ggtitle("Negative probes vs counts")
```



## Apply filteres

From paper: _"Cells with an average negative control count greater than 0.5 and less than 20 detected features were filtered out."_

Also remove cell with no cell type annotations.
```{r}
###so.raw <- JoinLayers(so.raw)

so <- so.raw[ ,so.raw$nCount_RNA >= min_count_per_cell &
               so.raw$avg_neg <= max_avg_neg &
               !(is.na(so.raw$celltype_subset) )]
ncol(so.raw)
ncol(so)
```

```{r}
so <- readRDS(seurat_file_01_loaded)
```

## How many cells per sample?

```{r}
table(so@meta.data$orig.ident)
```



# Basic preprocessing

```{r eval=FALSE}
num_dims <- 15
# Run through preprocessing
so <- NormalizeData(so)

## Do per sample to mimic paper approach somewhat.
so <- FindVariableFeatures(so, nfeatures = 200) 


so <- ScaleData(so) # Just 2k variable features
so <- RunPCA(so, features = VariableFeatures(so))
so <- RunUMAP(so, dims=1:num_dims)
so <- FindNeighbors(so, dims = 1:num_dims)
so <- FindClusters(so)

```

# Basic plots



## UMAP {.tabset .tabset-pills}

### Sample

```{r}
DimPlot(so, group.by = 'orig.ident')
```


### Condition

```{r}
DimPlot(so, group.by = 'condition')
```



### Condition X Sample

```{r fig.width=12, fig.height=3}
DimPlot(so, group.by = 'orig.ident',split.by = "condition")
```

### Total counts

```{r}
FeaturePlot(so, 'nCount_RNA') + scale_colour_viridis_c(option="magma", direction = -1)
```


### Clustered

```{r}
DimPlot(so, group.by = 'seurat_clusters')
```


### celltype_subset

Classifications from Garrido-Trigo et al 2023.

```{r}
DimPlot(so, group.by = 'celltype_subset')
```

### celltype_SingleR2

Classifications from Garrido-Trigo et al 2023.

```{r}
DimPlot(so, group.by = 'celltype_SingleR2')
```

## Clustering

Double checking the clusters line up with annotaiaon
```{r}
table(so$celltype_subset)
table(so$celltype_SingleR2)
```


Composition
```{r}
heatmap(table(so$celltype_subset,so$celltype_SingleR2))
```


# Save data

```{r eval=FALSE}
saveRDS(so, seurat_file_01_loaded)
```

