---
title: "Dataset: Mouse Alzheimers Plaques cosMX Mallach et al 2024"
author: "Sarah Williams"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
bibliography: /home/s.williams/zotero_export/MyLibrary.bib
---

Paper [Microglia-astrocyte crosstalk in the amyloid plaque niche of an Alzheimerâ€™s disease mouse model, as revealed by spatial transcriptomics](https://www.sciencedirect.com/science/article/pii/S2211124724005448?via%3Dihub)
[@mallachMicrogliaastrocyteCrosstalkAmyloid2024] explores the spatial transcritome of amaloid plaques in a mouse model.

Their work includes an analysis of cosMX samples from of 4 mouse brain samples. 

They have shared their full raw dataset via GEO; under accession [GSE263793](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE263793).

----

Note that the prepressing below makes no attempt to assign cell types or identify amyloid plaques. 


# Data download


The Cosmx dataset were downloaded from' https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE263791

Consisting of two slides (AD2_AD3_CosMx/GSM8199188_ID61-ID62_S10 and AD4_AD5_CosMx/GSM8199189_ID67-ID68_S18) - the following files
were downloaded;
```{}
GSM8199188_ID61-ID62_S10-polygons.csv.gz
GSM8199188_ID61-ID62_S10_CellComposite.tar.gz
GSM8199188_ID61-ID62_S10_CellLabels.tar.gz
GSM8199188_ID61-ID62_S10_CellOverlay.tar.gz
GSM8199188_ID61-ID62_S10_CompartmentLabels.tar.gz
GSM8199188_ID61-ID62_S10_exprMat_file.csv.gz
GSM8199188_ID61-ID62_S10_fov_positions_file.csv.gz
GSM8199188_ID61-ID62_S10_metadata_file.csv.gz
GSM8199188_ID61-ID62_S10_tx_file.csv.gz

GSM8199189_ID67-ID68_S18-polygons.csv.gz
GSM8199189_ID67-ID68_S18_CellComposite.tar.gz
GSM8199189_ID67-ID68_S18_CellLabels.tar.gz 
GSM8199189_ID67-ID68_S18_CellOverlay.tar.gz
GSM8199189_ID67-ID68_S18_CompartmentLabels.tar.gz
GSM8199189_ID67-ID68_S18_exprMat_file.csv.gz 
GSM8199189_ID67-ID68_S18_fov_positions_file.csv.gz
GSM8199189_ID67-ID68_S18_metadata_file.csv.gz 
GSM8199189_ID67-ID68_S18_tx_file.csv.gz
```


# Setup

```{r}
library(Seurat)
library(tidyverse)
library(magrittr) # %<>% is used somewhere (assignment pipe)
library(SingleR) # Cell refernce
library(scRNAseq) # dataset
library(scuttle) # sce normalisation
```


```{r}
dataset_dir      <- '~/projects/spatialsnippets/datasets/'
project_data_dir <- file.path(dataset_dir,'GSE263793_Mallach2024_AlzPlaque')
sample_dir            <- file.path(project_data_dir, "raw_data/") 
#annotation_file       <-
data_dir              <- file.path(project_data_dir, "processed_data/") 


seurat_file_00_loaded <- file.path(data_dir, "GSE263793_AlzPlaque_seurat_00_loaded.RDS")
seurat_file_01_preprocessed <- file.path(data_dir, "GSE263793_AlzPlaque_seurat_01_preprocessed.RDS")

# config
min_count_per_cell <- 100
max_pc_negs        <- 0.5

source("code/subset_obj_seurat_v2.R")



cols_x30 <- c(
  "#cec0e8","#343800","#ffa3ca","#381f00","#940029",
  "#2cdbd9", "#7900da","#009c28","#002fe4","#2d0013",
  "#b200cb","#00b367","#d06eff","#d78a00","#00256e",
  "#cacb67","#ff7de5","#a1d38e","#fd0073","#006042",
  "#ff4b2c","#0083b5","#b75200","#85a0ff","#7a5f00",
  "#560047","#f0be69","#00406d","#ff917c","#009138"

  )

```



## Load data


```{r}

load_sample_into_seurat <- function(the_slide){

  fileroot <- slide2fileroot[the_slide]
  
  #https://github.com/satijalab/seurat/issues/5819
  slide_with_dots <- gsub("_",".", the_slide)
  
  
  #sample_metadata_file <- file.path(sample_dir, the_slide,  paste0(fileroot,'_metadata_file.csv.gz'))
  #sample_mtx_file      <- file.path(sample_dir, the_slide,  paste0(fileroot,'_exprMat_file.csv.gz'))
  #sample_molecules_file <- file.path(sample_dir, the_slide,  paste0(fileroot,'_tx_file.csv.gz'))
  
  #GSM8199188_ID61-ID62_S10_CellComposite.tar.gz  
  #GSM8199188_ID61-ID62_S10_CompartmentLabels.tar.gz   
  #GSM8199188_ID61-ID62_S10_metadata_file.csv.gz
  #GSM8199188_ID61-ID62_S10_CellLabels.tar.gz     
  #GSM8199188_ID61-ID62_S10_exprMat_file.csv.gz        
  #GSM8199188_ID61-ID62_S10-polygons.csv.gz
  #GSM8199188_ID61-ID62_S10_CellOverlay.tar.gz    
  #GSM8199188_ID61-ID62_S10_fov_positions_file.csv.gz  
  #GSM8199188_ID61-ID62_S10_tx_file.csv.gz
  
  so <- LoadNanostring(file.path(sample_dir, the_slide), fov=slide_with_dots, assay = "RNA")
  
  
  so$slide_fov      <- slide_with_dots
  so$slide_fov_orig <- the_slide
  so$run_name       <-  as.character(slide2fileroot[the_slide])

  
  
  # # Will probably not consider (ns) fovs as pseudoreplicates here, buecase its
  # # tiled across two samples per slide.
  # # And the composition of brain structure across fovs is very different, so they're 
  # # not technical repicates. 
  # # STILL want to record cell's FOV membership though!
  # fov_pos_file <- file.path(sample_dir, the_slide, paste0(as.character(slide2fileroot[the_slide]),'_fov_positions_file.csv.gz'))
  # # nanostring fov not Seurat fov!
  # fov_pos <- read_csv(fov_pos_file)
  # # Need to calculate outer edges of each fov, not just one corner (or centre?)
  # # Get centroid of each cell
  # coords <- GetTissueCoordinates(so, image=slide_with_dots)
  # rownames(coords) <- coords$cell


  return(so)
}


slide2fileroot <- list( AD2_AD3_CosMx= 'GSM8199188_ID61-ID62_S10',
                        AD4_AD5_CosMx= 'GSM8199189_ID67-ID68_S18')

```



```{r eval=FALSE }

slides <- c('AD2_AD3_CosMx', 'AD4_AD5_CosMx')
sample_prefix <- gsub("_","",gsub("_CosMx","",slides))

so.list <- lapply(FUN=load_sample_into_seurat, X=slides)

#NB: merge is in SeuratObject packages, but must be called without :: 
so.raw <- merge(so.list[[1]], y=so.list[2:length(so.list)], add.cell.ids=sample_prefix)
#rm(so.list)

saveRDS(so.raw, seurat_file_00_loaded)
```

```{r}
so.raw <- readRDS(seurat_file_00_loaded)
```

# Negative probe handling

```{r}

# Join layers for some broad stats
so.raw[["RNA"]] <- JoinLayers(so.raw[["RNA"]] )


# Add gene names and tag negative probes in the RNA assay/feature metadata.
gene_meta <- so.raw[['RNA']]@meta.data
gene_meta$feature <- rownames(so.raw)
gene_meta$CodeClass <- ifelse(substr(gene_meta$feature, 1,6) == "NegPrb", 
                              "NegativeProbe", "RNAProbe" )
# prioro to cell filtering
gene_meta$total_count_unfiltered <- rowSums(GetAssayData(so.raw, assay="RNA", layer='counts'))
gene_meta$mean_count_unfiltered <- gene_meta$total_count_unfiltered / ncol(so.raw)

so.raw[['RNA']]<- AddMetaData(so.raw[['RNA']], gene_meta)


# Count how many negatives/RNA probes detected per cell
neg_matrix <- GetAssayData(so.raw, assay="RNA", layer='counts')[so.raw[['RNA']]@meta.data$CodeClass == "NegativeProbe",]

so.raw$nCount_neg    <- colSums(neg_matrix)
so.raw$nCount_nonneg <- so.raw$nCount_RNA - so.raw$nCount_neg
so.raw$neg_pc        <- so.raw$nCount_neg / so.raw$nCount_RNA * 100



# split layers back again 
# Layers have new names, which seem better than counts.1, counts.2
# counts.AD2_AD3_CosMx, counts.AD4_AD5_CosMx
so.raw[["RNA"]] <- split(so.raw[["RNA"]], f = so.raw$slide_fov)

```






# Basic QC filter

## Min count per cell

```{r}
ggplot(so.raw@meta.data, aes(x=nCount_RNA, col=slide_fov)) +
  geom_density() + 
  geom_vline(xintercept = min_count_per_cell, lty=3) +
  scale_x_log10() +
  theme_bw() +
  ggtitle("Counts per cell")
```


```{r}
ggplot(so.raw@meta.data, aes(x=nCount_RNA, y=nFeature_RNA, col=slide_fov)) +
  geom_point(pch='.', alpha=0.5) + 
  geom_smooth() + 
  scale_x_log10() + scale_y_log10() + 
  theme_bw() 
```



## Percent Negative probes

```{r}
ggplot(so.raw@meta.data, aes(x=neg_pc, col=slide_fov)) +
  geom_density() +
  geom_vline(xintercept = max_pc_negs, lty=3) +
  scale_x_log10() +
  theme_bw() +
  ggtitle("Negative probe composition")
```

```{r}
ggplot(so.raw@meta.data, aes(x=nCount_RNA, y=neg_pc, col=slide_fov)) +
  geom_point(pch=3, alpha=0.3) +
  scale_x_log10() +
  geom_vline(xintercept = min_count_per_cell, lty=3) +
  geom_hline(yintercept = max_pc_negs, lty=3) +
  theme_bw() +
  ggtitle("Negative probe composition")
table(so.raw$neg_pc >= 0.5)
```


## Negative vs RNA probes

```{r}
gene_info <- so.raw[['RNA']]@meta.data


ggplot(gene_info, aes(x=mean_count_unfiltered, col=CodeClass)) +
  geom_density() +
  geom_rug(data=filter(gene_info, CodeClass == "RNAProbe")) + 
  geom_rug(data=filter(gene_info, CodeClass == "NegativeProbe")) +
  scale_x_continuous(transform = 'log2') +
  theme_bw() +
  ggtitle("Negative Vs RNA Probe expression")
```



## Apply filteres

From paper: _"To remove cells that may have still been contaminated with, we subclustered the microglial cells, identified a cluster of cells characterized by higher counts of negative probe counts, and removed this cluster (141 cells) from subsequent analysis. A total of 37,840 high-quality cells across all tissue sections were retained for subsequent analysis."_



"_We discarded all transcripts that had a higher correlation than the top two negative probes (for a total of 56 probes) from further analysis, as likely candidates of nonspecific binding._"

```{r}
so <- so.raw[ ,so.raw$nCount_RNA >= min_count_per_cell &
               so.raw$neg_pc<= max_pc_negs  ]
ncol(so.raw)
ncol(so)
```



# Basic preprocessing

```{r eval=FALSE}
num_dims <- 15

# Run through preprocessing
so <- NormalizeData(so)

# With the two slides as layers, common features should be found..
so <- FindVariableFeatures(so, nfeatures = 200) 
so <- ScaleData(so) 
so <- RunPCA(so, features = VariableFeatures(so))
so <- RunUMAP(so, dims=1:num_dims)
so <- FindNeighbors(so, dims = 1:num_dims)
so <- FindClusters(so)


so$cell <- colnames(so)
```

One we prepared earlier;
```{r}
so <- readRDS(seurat_file_01_preprocessed)
```

```{r}
VariableFeaturePlot(so)
ElbowPlot(so)
```


# Figure out samples

There are two mice per slide - unsure of what they are though.

```{r}
ImageDimPlot(so, fov = "AD2.AD3.CosMx", cols = "polychrome", axes = TRUE)
ImageDimPlot(so, fov = "AD4.AD5.CosMx", cols = "polychrome", axes = TRUE)
```




Define some sample names based on tissue slices. These do not correlate to original paper. 

```{r}
coords <- merge(x=so@meta.data, y=GetTissueCoordinates(so), by='cell')

coords1 <- merge(x=so@meta.data, y=GetTissueCoordinates(so, image = "AD2.AD3.CosMx"), by='cell', all=FALSE)

coords1$sample <- ifelse(coords1$y < 0, 'sample1', "sample2")

coords2 <- merge(x=so@meta.data, y=GetTissueCoordinates(so, image = "AD4.AD5.CosMx"), by='cell', all=FALSE)
coords2$sample <- ifelse(coords2$y < 0, 'sample3', "sample4")

coord_lookup <- c(setNames(coords1$sample, nm=coords1$cell), setNames(coords2$sample, nm=coords2$cell))
so$sample <-as.factor(as.character(coord_lookup[so$cell]))



```

Samples recorded
```{r}
ImageDimPlot(so, fov = "AD2.AD3.CosMx",group.by = 'sample', axes = TRUE)
ImageDimPlot(so, fov = "AD4.AD5.CosMx",group.by = 'sample', axes = TRUE)
```


Edges of each sample for plotting
```{r}

cell2sample <- so$sample
cell2slide  <- so$slide_fov
cell_coords <- rbind(GetTissueCoordinates(so, image ='AD2.AD3.CosMx'),
                     GetTissueCoordinates(so, image ='AD4.AD5.CosMx') ) 
cell_coords$sample <- cell2sample[cell_coords$cell]
cell_coords$slide_fov <- cell2slide[cell_coords$cell]
sample_coords <- cell_coords %>% group_by(slide_fov, sample) %>%
  summarise(min_x = min(x), 
            max_x = max(x),
            min_y = min(y), 
            max_y = max(y) )
cell_coords
```



```{r eval=FALSE}
saveRDS(so, file = seurat_file_01_preprocessed)
```


# Basic annotion

Don't have the full analysis annotation to hand, and not interested in fully replicating the paper. Just add some very basic annotations to demonstrate downstream analyses.

## Cell types


### Cell typing

Run singleR with a mouse brain reference to determine some cell types.

The [@Zeisel2015] dataset from paper ['Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq'](https://www.science.org/doi/10.1126/science.aaa1934)
is a nice simple annotation that gives us a few basic cell types to work with. It is made easily accessible via the [scRNAseq](https://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html) data package.

```{r}
available_ds <- as_tibble(scRNAseq::listDatasets() ) %>% dplyr::filter(Taxonomy == '10090', Part == 'brain')
DT::datatable(available_ds)
```
<!-- ```{r} -->
<!-- #all.ds <- listDatasets() -->
<!-- #sce.tasic <- TasicBrainData() # tried,  too detailed? -->
<!-- #sce.ref <- JessaBrainData() # tried, too detailed (also what brain part?) -->
<!-- ``` -->


<!-- tried others, too much detail for what we want -->

<!--Follow : https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2 --? 

<!-- is RTCD good because it explicitly does two predictions per cell (spot), and you pick the primary? -->
<!-- how is spatial data used??? --> 

Download the reference dataset, see the cell types present.
```{r}
sce.ref <- scRNAseq::ZeiselBrainData()
sce.ref <- logNormCounts(sce.ref) # lognorm reference data
table(sce.ref$level1class)
```

Make cell type calls at the cell and cluster level.
```{r}
# cluster names
so$cluster <- factor( paste0("c", so$seurat_clusters),
                      levels=paste0("c",levels(so$seurat_clusters)))

# Join layers for some broad stats
so.joined   <- JoinLayers(so[["RNA"]] )
norm_matrix <- GetAssayData(so.joined, assay= "RNA", layer='data')

predictions <- SingleR( test   = norm_matrix, 
                        ref    = sce.ref,
                        labels = sce.ref$level1class    )


predictions.cluster <- SingleR( test   = norm_matrix, 
                        ref    = sce.ref,
                        labels = sce.ref$level1class,
                        clusters = so$cluster)

```


And store them

 Store it

```{r}


# Cluster level predictions
so$cluster_pred           <- factor(predictions.cluster[so$cluster, "pruned.labels"])
so$cluster_pred_deltanext <- signif(predictions.cluster[so$cluster, "delta.next"],digits = 3)

# c1 => c1: pyramidal SS
cluster_lookup     <- 
  setNames(predictions.cluster$pruned.labels,
           nm=rownames(predictions.cluster))
clustername_lookup <- 
  setNames(paste0(names(cluster_lookup),": ", cluster_lookup ),
           nm = names(cluster_lookup))
so$cluster_name <- factor(as.character(clustername_lookup[so$cluster]), levels=clustername_lookup)


# cell level predictons
so$pred           <- factor(predictions[so$cell, "pruned.labels"])
so$pred_deltanext <- factor(predictions[so$cell, "delta.next"])

```

<!-- # if less than this many cells, classify as other. -->
<!-- min_considered_type <- 20 -->
<!-- type_counts  <- table(so$pred) -->
<!-- type_common  <- names(type_counts)[type_counts >= min_considered_type] -->
<!-- so$pred_main <- as.character(so$pred) -->
<!-- so$pred_main[ ! (so$pred_main %in% type_common) ] <- "Other" -->
<!-- so$pred_main <- factor(so$pred_main) -->


Relationship between predictions
```{R}
library(viridis)
library(pheatmap)
plotScoreHeatmap(predictions.cluster)

```

How do the cell level and cluster level assignments compare? - Clustered groupings are mostly matched to individual cell calls.

```{r}
plottable <- so@meta.data[,c('pred', "cluster_pred")] %>%
  group_by(pred, cluster_pred) %>%
  summarise(n=n())

ggplot(plottable, aes(x=cluster_pred, y=n, fill=pred)) +
  geom_bar(position="fill", stat="identity") + 
  theme_bw() +
  coord_flip() + 
  theme(legend.position = "bottom") +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values=cols_x30)


plottable <- so@meta.data[,c('pred', "cluster_name")] %>%
  group_by(pred, cluster_name) %>%
  summarise(n=n())

ggplot(plottable, aes(x=cluster_name, y=n, fill=pred)) +
  geom_bar(position="fill", stat="identity") + 
  theme_bw() +
  coord_flip() + 
  theme(legend.position = "bottom") +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values=cols_x30)


```



And how each level of cell typing looks spatially;

```{r}
so.s <- subset_opt(so, cells=so$cell[so$sample == 'sample2'])
```

```{r}
cluster_cols <- c(
  c0  = 'yellow',
  c1  = 'royalblue',
  c2  = 'turquoise4',
  c3  = 'turquoise2',
  c4  = 'red',
  c5  = 'olivedrab1',
  c6  = 'yellow3',
  c7  = 'darkviolet',
  c8  = 'tomato3',
  c9  = 'navy',
  c10 = 'olivedrab',
  c11 = 'springgreen',
  c12 = 'springgreen4',
  c13 = 'blue',
  c14 = 'plum3',
  c15 = 'tomato',
  c16 = 'lightblue2'
)
cluster_cols_names <- setNames(cluster_cols, nm=levels(so$cluster_name))


```

```{r fig.width=12}
ImageDimPlot(so.s,  fov = "AD2.AD3.CosMx",group.by = 'pred', boundaries = "segmentation", border.color = 'black', axes = T, crop=TRUE)
ImageDimPlot(so.s,  fov = "AD2.AD3.CosMx",group.by = 'cluster_pred', boundaries = "segmentation", border.color = 'black', axes = T, crop=TRUE)
ImageDimPlot(so.s,  fov = "AD2.AD3.CosMx",group.by = 'cluster_name', boundaries = "segmentation", border.color = 'black', axes = T, crop=TRUE)
```


With better colours
```{r}
ImageDimPlot(so.s,  fov = "AD2.AD3.CosMx",group.by = 'cluster_name', boundaries = "segmentation", border.color = 'black', axes = T, crop=TRUE, cols = cluster_cols_names) 
```











<!-- ```{r} -->
<!-- # remove CR cells because there aren't enough of them for annotation -->
<!-- allen.cortex.ref <- subset(allen.cortex.ref, subset = subclass != "CR") -->
<!-- counts <- GetAssayData(allen.cortex.ref, assay = "RNA", slot = "counts") -->
<!-- cluster <- as.factor(allen.cortex.ref$subclass) -->
<!-- names(cluster) <- colnames(allen.cortex.ref) -->
<!-- nUMI <- allen.cortex.ref$nCount_RNA -->
<!-- names(nUMI) <- colnames(allen.cortex.ref) -->
<!-- nUMI <- colSums(counts) -->
<!-- levels(cluster) <- gsub("/", "-", levels(cluster)) -->
<!-- reference <- Reference(counts, cluster, nUMI) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- # run RCTD with many cores -->
<!-- RCTD.1 <- create.RCTD(query.1, reference, max_cores = 8) -->
<!-- RCTD.1 <- run.RCTD(RCTD.1, doublet_mode = "doublet") -->

<!-- ``` -->








## Neighbourhood niches

```{r eval=FALSE}
so.1 <- subset_opt(so, cells=so$cell[so$slide_fov == 'AD2.AD3.CosMx'])
so.1 <- BuildNicheAssay(so.1, fov = "AD2.AD3.CosMx", group.by = "seurat_clusters", niches.k = 5, neighbors.k = 30)

so.2 <- subset_opt(so, cells=so$cell[so$slide_fov  == 'AD4.AD5.CosMx'])
so.2 <- BuildNicheAssay(so.2, fov = "AD4.AD5.CosMx", group.by = "seurat_clusters", niches.k = 5, neighbors.k = 30)
```



# Basic plots

## UMAP {.tabset .tabset-pills}

### Slide


```{r}
DimPlot(so, group.by = 'slide_fov')
```
### Sample

```{r}
DimPlot(so, group.by = 'sample')
```

### Cluster

```{r}
DimPlot(so)
```

### Cluster labelled

```{r}
DimPlot(so, group.by = 'cluster_pred')
```

### Prediction

```{r}
DimPlot(so, group.by = 'pred_main')
```





### Total read count

```{r}
so$log_nCountRNA <- log10(so$nCount_RNA)
FeaturePlot(so, features = 'log_nCountRNA')
```


### Negative probe percentage

```{r}
FeaturePlot(so, features = 'neg_pc')
```


# Spatial example plot


```{r}
ImageDimPlot(so, fov = "AD2.AD3.CosMx",group.by = 'sample', axes = TRUE)
```


Just one sample?
```{r}


so.s <- subset_opt(so, cells=so$cell[so$sample == 'sample2'])

```

```{r}
ImageDimPlot(so.s,  fov = "AD2.AD3.CosMx",group.by = 'seurat_clusters', boundaries = "segmentation", border.color = 'black', axes = T, crop=TRUE)
```




```{r}
ImageFeaturePlot(so.s,  features = 'Apoe',fov = "AD2.AD3.CosMx", boundaries = 'segmentation',border.color = 'black')
ImageFeaturePlot(so.s,  features = 'Lpl',fov = "AD2.AD3.CosMx", boundaries = 'segmentation',border.color = 'black')

```


And with some interesting genes  (Needs SF package.)
```{r}
genes <- c('Apoe', 'Lpl', 'Trem2') # DAM markers - disease-associated microglia
ImageDimPlot(so.s,  fov = "AD2.AD3.CosMx",
             molecules = genes, 
             group.by = 'sample', boundaries = "segmentation",cols = c("grey30"),
             border.color = 'black', axes = T, crop=TRUE)
```

```{r}
ImageDimPlot(so.s,  fov = "AD2.AD3.CosMx",
             molecules = genes[1], 
             group.by = 'cluster_name', boundaries = "segmentation",
             alpha=0.4,
             cols = cluster_cols_names,
             mols.size = 1,
             border.color = 'black', axes = T, crop=TRUE)

```



# References


