---
title: "Dataset: Wnt4 KO in Developing mouse kidney"
author: "Sarah Williams"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
bibliography: /home/s.williams/zotero_export/MyLibrary.bib
---


Data from paper [Forming nephrons promote nephron progenitor maintenance and branching morphogenesis via paracrine BMP4 signalling under the control of Wnt4](https://www.biorxiv.org/content/10.1101/2023.11.19.567482v1.full) [@moreauFormingNephronsPromote2023a]

This study included 10X chromium single cell RNAseq data from 4 conditions, with 3-4 E14.5 mice pooled per group. 

* Sample1 (Wnt4FloxKO):  Wnt4Flox/Flox Six2-Conditional Wnt4 Knockout
* Sample2 (Wnt4FloxHet):  Wnt4Flox/+ Six2-Conditional Wnt4 Het
* Sample3 (Wnt4Het): Wnt4 GCE/+ Control Wnt4 Het
* Sample4 (Wnt4KO): Wnt4 GCE/GCE Knockout Wnt4

In that paper they explain that complete or conditional homozygous knockout of Wnt4 gene results in abnormal kidney development, and they use scRNAseq data to explore effects at cellular level. [@moreauFormingNephronsPromote2023a]


# Libraries

```{r}
library(Seurat)
#library(SeuratObject)
library(tidyverse)
```

# Data load

Loading data from a previously annotated object used in the analysis.

```{r}
dataset_dir      <- '~/projects/spatialsnippets/datasets'
project_data_dir <- file.path(dataset_dir,'Wnt4KO_Moreau2023')


seurat_file_loaded <- file.path(project_data_dir, "Wnt4KOE14.5_10_withdeconclust.rds")
seurat_file_saved  <- file.path(project_data_dir,   "Wnt4KOE14.5_11_ss.rds")


```



# Load and tidy data

```{r}
so <- readRDS(seurat_file_loaded)

# There's alot of historical/working feilds in the metadata. 
# Slim down to just useful ones.
keep_cols <- c(
  "barcode","orig.ident","nCount_RNA","nFeature_RNA" ,"sample", # basic annotation/stats.
  "Genotype","GTeffect", "GTshort" , # same info, 3 ways
  "MT_pc", "cc" ,"G1" ,"S" ,"G2M", "doublet_score" , # was filtered on scrublet.
  # "cluster_content","cluster_lab"    # More detailed groupings used in working  
  "Celltype", "CelltypeCode",          # final types in paper
  "CelltypeBroad", "CelltypeBroadCode" # broader groupings
)

so@meta.data <- so@meta.data[,keep_cols]


# Reduce to a simple case - KO vs Het (WT)
# Or keep these in to demo using all 4 in a model for 1 vs 1 (useful information)
#so <- so[,so$GTshort %in% c("Wnt4Het", "Wnt4KO")]
#so$Genotype <- droplevels(so$Genotype)
#so$GTeffect <- droplevels(so$GTeffect)
#so$GTshort  <- droplevels(so$GTshort)
#so$sample  <- droplevels(so$sample)
```


## Data tour

```{r}
DimPlot(so, group.by="CelltypeCode",label = TRUE) + NoLegend()
DimPlot(so, group.by="CelltypeBroad",label = TRUE) + NoLegend()
DimPlot(so, group.by="GTshort")
DimPlot(so, group.by="cc")
```

```{r fig.width=10}
FeaturePlot(so, c('nCount_RNA','MT_pc'))
```



4 Samples, n=1 per group. Can use 'GTshort' to identify the sample.
```{r}
select(so@meta.data, sample, Genotype, GTeffect, GTshort) %>% 
  as_tibble() %>% 
  group_by( sample, Genotype, GTeffect, GTshort) %>%
  summarise(num_cells=n(), .groups = 'drop') %>%
  DT::datatable()
```









# References
