---
title: "Differential expression between groups using pseudobulk"
author: "Sarah Williams"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
bibliography: /home/s.williams/zotero_export/MyLibrary.bib
---


```{r angrylibrarian, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)
```

<!-- keywords

insituspatial
seurat
de
test

-->

# Overview

Once we have identified cell types present in the samples, its typical to test how
gene expression changes between experimental conditions, within each different cell type. 
Some cell types may be dramatically affected by the experimental conditions, while others are not. Likewise some genes may change only in a specific cell type, whereas others show a more general difference.


This document describes how to apply a pseudobulk approach to test for differences between groups, accounting for biological replicates. This is very similar to how a non-spatial single cell experiment may be analysed.

In a pseudobulk approach counts are obtaing by pooling together groups of cells; in this case cells from the of the same type from the same fov. These pooled counts can then be analysed more like a bulk RNAseq experiment. 

Note that there are other approaches to calculate differential expression in this kind of data - including those that make use of individual cells [@Soneson2017].




![](assets/insitu_spatial_pseudobulk_de.png)

This requires:

* Biological replicates for each group
* Assigned cell types
* [Optionally] Multiple fovs measured per sample


For example:

* What genes are differentially expressed in epithelial cells in Crohn's disease vs healthy individuals?
* How do genes change with treatment in each different cell type in my sample?


Steps:

1. Calculate pseudobulk
2. Filter to testable pseudobulk groups (enough cells to pool)
3. Filter to testable genes (enough expression to see changes)
4. Test for changes in gene expression
5. Plot DE results and individual genes. 


# Worked example 

How does gene expression change within each cell type between Ulcerative colitis or Crohn's disease, and Healthy controls?

Using data from [@garrido-trigoMacrophageNeutrophilHeterogeneity2023]

## Load libraries and data

```{r}
library(Seurat)
library(speckle)
library(tidyverse)
library(limma)
library(DT)
library(edgeR)
```

```{r}
data_dir              <- file.path("~/projects/spatialsnippets/datasets/GSE234713_IBDcosmx_GarridoTrigo2023/processed_data") 
seurat_file_01_loaded <- file.path(data_dir, "GSE234713_CosMx_IBD_seurat_01_loaded.RDS")
```


```{r}
so <- readRDS(seurat_file_01_loaded)
```

## Experimental design

<!-- Move this to own page? --->

There are three individuals per condition (one tissue sample from each individual). With multiple fovs on each physical tissue sample.
```{r}
sample_table <- select(as_tibble(so@meta.data), condition, individual_code, fov_name) %>%
  unique() %>% 
  group_by(condition, individual_code) %>% 
  summarise(n_fovs= n(), item = str_c(fov_name, collapse = ", "))

DT::datatable(sample_table)
```






## Count how many cells of each type in your data

Using a pseudobulk approach.

* Need at least x reads in a cell to include it
* Need at least x cells of a celltype within an fov to include that
* Can only test where we have at least 2 samples on each side of a contrast.


```{r}
min_reads_per_cell <- 200

ggplot(so@meta.data, aes(x=nCount_RNA)) +
  geom_density() +
  geom_vline(xintercept = min_reads_per_cell, lty=3) +
  scale_x_log10() +
  theme_bw()+
  ggtitle("How many reads per cell?")


so<- so[,so$nCount_RNA >= min_reads_per_cell]
```



We will pool each celltype within each fov (cluster_group). But there needs to be a certain number of cells for that to work - less than a certain number of cells and a pseudobulk pool will be excluded.


Note there are much fewer t-cells overall, but given that we have a high number of samples, there should still be enough to include.  Its typical that some of the less common cell types are difficult or impossible to reliably test.

```{r}
min_cells_per_fovcluster <- 20

so$fov_cluster <- paste0(so$fov_name,"_", so$celltype_subset)

celltype_summary_table <- so@meta.data %>% 
  group_by(condition, group, individual_code, fov_name, celltype_subset, fov_cluster) %>%
  summarise(cells=n(), .groups = 'drop')
DT::datatable(celltype_summary_table)
```


```{r}
ggplot(celltype_summary_table, aes(x=cells, col=celltype_subset)) +
  geom_density() +
  geom_vline(xintercept=min_cells_per_fovcluster, lty=3) +
  geom_rug() +
  scale_x_log10() +
  theme_bw() +
  ggtitle("How many cells per fov-cluster?")
```


Record the names of those fov_clusters that contain enough cells to be used. 
Will use this later to filter.
```{r}
passed_fov_clusters <- celltype_summary_table$fov_cluster[celltype_summary_table$cells >= min_cells_per_fovcluster]
```


## Calculate pseudobulk

Now use the _PseudobulkExpression()_ function to sum up each gene's gene expression across each *fov_cluster*. 

```{r}
pseudobulk_counts <- PseudobulkExpression(so, assays = "RNA", layer="counts",  method = 'aggregate', group.by = 'fov_cluster')
pseudobulk_counts_matrix <- pseudobulk_counts[["RNA"]]

# Change - back to _. Ideally we'd have neither, but - will cause problems later
colnames(pseudobulk_counts_matrix)<-gsub("-","_",colnames(pseudobulk_counts_matrix))
```


Now instead of counts for each individual cell, we have our pseodubulk matrix 
with the pooled sum of counts for each celltype within each fov region. So the numbers are much higher, fewer zeros. But each pool is a different size, which we will adress later with normalisation. 
```{r}
pseudobulk_counts_matrix[1:10,1:4]
```


Next, filter to only those passed fov_clusters we saw earlier - this is what we'll use for calculating differential expression.  

We also need an annotation table to tell us what is in each fov_cluster (sample, celltype, condition...). Build this by filtering the celltype summary table made earlier. If we filter it to the same samples, in the same order as the pseudubulk matrix, we can then use those columns of data to build our differential expression model in the next step.

```{r}
# Filter the pseudobulk table
pseudobulk_counts_matrix <- pseudobulk_counts_matrix[,passed_fov_clusters]

# And filter the celltype summary into an annotation table for only those fov_clusters
# pull in relevant annotation in a matched order
pseudobulk_anno_table <- celltype_summary_table
match_order <- match(passed_fov_clusters, pseudobulk_anno_table$fov_cluster)
pseudobulk_anno_table <- pseudobulk_anno_table[match_order,]

# Double check that the order of samples in the annotation table matches the pseudobulk table. 
# If this is wrong then the results will be nonsense!
stopifnot(all(colnames(pseudobulk_counts_matrix) == pseudobulk_anno_table$fov_cluster  ))
```


## Calculate Differential Expression

We now have a counts matrix, and an annotation table that describes its samples. This looks very much like a bulk RNAseq experiment. 

```{r}
pseudobulk_counts_matrix[1:10,1:4]
head(pseudobulk_anno_table)
```

We will use a fairly standard limma differential expression analysis. 

We will process each celltype one at a time as follows:

1) Subset the counts matrix and annotation table to the celltype
2) Check that we have enough replicates for that celltype to run the test(s). We need to have fovs from at least 2 biological samples on each side of our test (e.g. 2treatment, 2 control). Even though we have enough tissue samples, if its a less common cell type we might have filtered them away!
2) Build a model. In our case, a simple one that looks for variouat by disease group, blocked on individual.
3) Run the test.



How to build the model and run the test will need to be customised foryour experimental design. Online resources that describe approaches for bulk RNAseq analyses can be applied to pseubulk analyses - see the 'More Information' section of this document for suggestions. 


In this case the line `model.matrix( ~0 + group)` builds a model where gene expression
is expected to vary by group (UC/CD/HC). The ~0 part indicates a 'intercept', which is convenient becase it means we can include all 3 groups in our contrasts, rather than one of them being treated as the baseline. ( NB: Mathematically, you could use ~group alone without intercept, and define contrasts accordingly to get the same result. )

But, we also need to tae into acoount our individuals. We do this by *blocking* on individual, and calculageing a intra duplicate correlation with _duplicateCorrelation()_. 


If instead we didn't have multiple fovs per replicate, we might use a model like `~0 + group + individual`; no blocking or use of duplicateCorrelation(). That is more typical of a typical of a bulk rnaseq or single cell RNAseq pseuobulk analysis.


```{r}
min_samples_to_calc <- 2 # require 2 samples on on either side of contrast

de_result_list <- list()

# celltype_subset is a matrix
for (the_celltype in levels(so$celltype_subset)) {
  
  anno_table.this   <- pseudobulk_anno_table[pseudobulk_anno_table$celltype_subset == the_celltype,]
  count_matrix.this <- pseudobulk_counts_matrix[,anno_table.this$fov_cluster]
 
  print(the_celltype)
  
  # skip clusters with nothing 
  if( nrow(anno_table.this) < 1 ) {next}
  
  # Setup objects for limma
  dge <- DGEList(count_matrix.this)
  dge <- calcNormFactors(dge)
  
  # Build model
  group           <- anno_table.this$group
  individual_code <- anno_table.this$individual_code

  # To do any calculations, we need at least 2 pseudobulk groups per contrast.
  # there are plenty in this experiemnt, but with less replicates and rare cell types
  # it may be neccesary to check and skip certain contrasts

  # Model design 
  design    <- model.matrix( ~0 + group)

  vm  <- voom(dge, design = design, plot = FALSE)
  
  # Adding dupliate correlation to use individual fovs, rather than pooled per biosample
  corrfit <- duplicateCorrelation(vm, design, block=individual_code)
  fit     <- lmFit(vm, design, correlation = corrfit$consensus, block=individual_code) 

  # Then fit contrasts and run ebayes
  contrasts <- makeContrasts(UCvHC  = groupUC - groupHC, 
                             CDvHC  = groupCD - groupHC,
                           levels=coef(fit))

  fit <- contrasts.fit(fit, contrasts)
  fit <- eBayes(fit)

  
  for ( the_coef in colnames(contrasts) ) {
    de_result.this <- topTable(fit, n = Inf, adjust.method = "BH", coef = the_coef) %>%
      rownames_to_column("target") %>%
      mutate(contrast=the_coef,
             celltype=the_celltype) %>%
      select(celltype,contrast,target,everything()) %>%
      arrange(P.Value)
    
    
      de_result_list[[paste(the_celltype, the_coef, sep="_")]] <- de_result.this
    
  }

  

 
}

# Join together results for all celltypes, and pull out those with a singificant adjusted p-value
de_results_all <- bind_rows(de_result_list)
de_results_sig <- filter(de_results_all, adj.P.Val < 0.01)
```



Table of significant results.

```{r}
DT::datatable(de_results_sig)
```


## DE plots

The below plots show the logFC calculate for each gene versus its average expression across all samples. This is a useful diagnostic plot to evaluate your differential expression results.

* Only big changes are significant at lower expressions, simply becuase higher expression means more statistical confidence.
* A lopsided plot might indicate some oddness around the normalisation of your data (e.g. extremely differnet cell counts). 


[[ADD pvalue plot?]]


```{r}
library(ggrepel) # gg_repel, For non-overlapping gene labels


make_ma_style_plot <- function(res_table, pval_threshold = 0.01, n_genes_to_label = 10) {
  p <- ggplot(res_table, aes(x=AveExpr, y=logFC, col=adj.P.Val < pval_threshold) ) +
    geom_hline(yintercept = c(0), col='grey80') +
    geom_point(pch=3) +
    geom_text_repel(data    = head(arrange(filter(res_table , adj.P.Val < pval_threshold ), P.Value), n=5),
                    mapping = aes(label=target), col="red" ) +
    theme_bw() +
    geom_hline(yintercept = c(-1,1), lty=3) +
    scale_colour_manual(values = c('FALSE'="black", 'TRUE'="red")) +
    theme(legend.position = 'none')
  return(p)
}

```

```{r}
#res_table.UCvHC.epi <- filter(de_results_all, contrast == "UCvHC", celltype=="epi")

make_ma_style_plot(res_table = filter(de_results_all, contrast == "UCvHC", celltype=="epi"))
make_ma_style_plot(res_table = filter(de_results_all, contrast == "UCvHC", celltype=="tcells"))
make_ma_style_plot(res_table = filter(de_results_all, contrast == "UCvHC", celltype=="stroma"))



```

## Check some examples

Its always worth visualising how the expression of your differentially expressed genes really looks, with respect to your experimental design. How best to do this depends on your experiment. 


The results suggests that TNFRSF18 was significantly DE between individuals with Ulcerative Colitis and Healthy Controls in plasma cells. There's some very convenient seurat plots below;

```{r}
VlnPlot(subset(so, celltype_subset == "plasmas"), features = "TNFRSF18", group.by = 'group', alpha = 0.1)
FeaturePlot(so, "TNFRSF18", split.by = "group")
```

But it gets difficult to summarise data at the single cell level. We can also use the the normalised pseudobulk expression to see how gene expression varies within each fov,individual,celltype and condition - The plot below shows an overview of TNFRSF18 across the entire experiment.  

NB: We have to plot normalised expression instead of the raw counts as there are vastly different numbers of cells in each fov+celltype grouping. 

```{r}
# Get tmm normalised coutns for all pseudobulk
# WHen we did the DE we calculated this a celltype at a time, so values might differ slightly!
dge <- DGEList(pseudobulk_counts_matrix)
dge <- calcNormFactors(dge)
norm_pseudobulk <- cpm(dge , log=TRUE) # uses tmm normalisation

# Plot expression for TNFRSF18
plottable <- cbind(pseudobulk_anno_table, expression = norm_pseudobulk["TNFRSF18",])
ggplot(plottable, aes(x=individual_code, y=expression, col=condition )) +
  geom_boxplot(outlier.shape = NA) +
  geom_point() + 
  theme_bw() + 
  theme(axis.text.x=element_text(angle = -90, hjust = 0)) + 
  facet_wrap(~celltype_subset) 

```


And you can compare that with IGHG1, which was flagged at differentially expressed across multiple cell types.

```{r}
plottable <- cbind(pseudobulk_anno_table, expression = norm_pseudobulk["IGHG1",])
ggplot(plottable, aes(x=individual_code, y=expression, col=condition )) +
  geom_boxplot(outlier.shape = NA) +
  geom_point() + 
  theme_bw() + 
  theme(axis.text.x=element_text(angle = -90, hjust = 0)) + 
  facet_wrap(~celltype_subset) 
```

# Code Snippet



```{r eval=FALSE}
library(Seurat)
library(edgeR)
library(limma)


min_reads_per_cell       <- 200
min_cells_per_fovcluster <- 20
min_samples_to_calc      <- 2 # require 2 samples on on either side of contrast


# Remove cells with too few counts
so <- so[,so$nCount_RNA >= min_reads_per_cell]


# Define fov+cluster groups, with all relevant sample annotation
# remove those with too few cells.
so$fov_cluster <- paste0(so$fov_name,"_", so$celltype_subset)
celltype_summary_table <- so@meta.data %>% 
  group_by(condition, group, individual_code, fov_name, celltype_subset, fov_cluster) %>%
  summarise(cells=n(), .groups = 'drop')


# Calculate pseudobulk for each fov+cluster group
pseudobulk_counts <- PseudobulkExpression(so, assays = "RNA", layer="counts",  method = 'aggregate', group.by = 'fov_cluster')
pseudobulk_counts_matrix <- pseudobulk_counts[["RNA"]]
# Change - back to _. Ideally have neither and skip this step
colnames(pseudobulk_counts_matrix)<-gsub("-","_",colnames(pseudobulk_counts_matrix))


# Determin fov_clusters with eough cells
# Filter both pseudobulk matrix and pseudobulk annotation 
passed_fov_clusters <- celltype_summary_table$fov_cluster[celltype_summary_table$cells >= min_cells_per_fovcluster]
pseudobulk_counts_matrix <- pseudobulk_counts_matrix[,passed_fov_clusters]
pseudobulk_anno_table    <- celltype_summary_table[passed_fov_clusters,]


# Calculate DE across every celltype
de_result_list <- list()

for (the_celltype in levels(so$celltype_subset)) {
  
  anno_table.this   <- pseudobulk_anno_table[pseudobulk_anno_table$celltype_subset == the_celltype,]
  count_matrix.this <- pseudobulk_counts_matrix[,anno_table.this$fov_cluster]
 
  print(the_celltype)
  
  # skip clusters with nothing 
  if( nrow(anno_table.this) < 1 ) {next}
  
  # Setup objects for limma
  dge <- DGEList(count_matrix.this)
  dge <- calcNormFactors(dge)
  
  # Build model
  group           <- anno_table.this$group
  individual_code <- anno_table.this$individual_code

  # To do any calculations, we need at least 2 pseudobulk groups per contrast.
  # there are plenty in this experiemnt, but with less replicates and rare cell types
  # it may be neccesary to check and skip certain contrasts

  # Model design 
  design    <- model.matrix( ~0 + group)

  vm  <- voom(dge, design = design, plot = FALSE)
  
  # Adding dupliate correlation to use individual fovs, rather than pooled per biosample
  corrfit <- duplicateCorrelation(vm, design, block=individual_code)
  fit     <- lmFit(vm, design, correlation = corrfit$consensus, block=individual_code) 

  # Then fit contrasts and run ebayes
  contrasts <- makeContrasts(UCvHC  = groupUC - groupHC
                           levels=coef(fit))

  fit <- contrasts.fit(fit, contrasts)
  fit <- eBayes(fit)

  
  for ( the_coef in colnames(contrasts) ) {
    de_result.this <- topTable(fit, n = Inf, adjust.method = "BH", coef = the_coef) %>%
      rownames_to_column("target") %>%
      mutate(contrast=the_coef,
             celltype=the_celltype) %>%
      select(celltype,contrast,target,everything()) %>%
      arrange(P.Value)
    
    
      de_result_list[[paste(the_celltype, the_coef, sep="_")]] <- de_result.this
    
  }

}

de_results_all <- bind_rows(de_result_list)
de_results_sig <- filter(de_results_all, adj.P.Val < 0.01)


```


# Results

```{r}
DT::datatable(head(de_results_all))
```

This table is the typical output of _limma_ tests; With a couple of extra columns added by our code.

* **celltype**: The celltype being tested (Added by example code)
* **contrast**: The contrast being tested (Added by example code)
* **target** : The gene name (Added by example code, is the rowname in limma output)
* **rownames** : The tested cell types
* **logFC** : Log 2 fold change between tested groups. For a test of Test-Con; 
    + At logFC +1, A is doubled B. 
    + At logFC -1, A is half of B.  
    + A logFC 0 indicates no change.
* **AveExpr** : Average expression of a gene across all replicates. 
* **t** : Moderated T-statistic. See Limma documentation.
* **P.Value** : P.value
* **adj.P.Val** : A multiple-hypothesis corrected p-value
* **B** : B statistic (rarely used). See Limma documentation.


# More Information

* ['Ochestrating single cell analysis with bioconductor' book chapter 'DE analyses between conditions'](https://bioconductor.org/books/release/OSCA/) : An explanation of the 'pseudobulk' approch to single cell differential expression calculation.
* [Bias, robustness and scalability in single-cell differential expression analysis](https://www.nature.com/articles/nmeth.4612) [@Soneson2017] : A review of single cell differential expression calculation methods. 
* [Seurat Differntial expression Vignette](https://satijalab.org/seurat/articles/de_vignette) : How to run differential expression analyses using Seurat.
* [limma documentation](https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf) [@Ritchie2015]: The complete manual to limma. 
* [Differential Expression with Limma-Voom UC davis bioinformatics training](https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html) : A more accessible explanation of bulk RNAseq analyses using limma. 
* [Interactions and contrasts](https://genomicsclass.github.io/book/pages/interactions_and_contrasts.html) : An excellent visual explanation of how to build linear models for more complex multi-factor experimental designs (e.g. treatment _and_ genotype). Part of a larger [Data Analysis for Genomics class](https://github.com/genomicsclass/book) resource.



# References

