---
title: "Example Title"
author: 
   - "Bessie Bioinformatician"
   - "Robert DropTable-Students (User Review)"
   - "Steph Stats (Statistical Review)"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
bibliography: /home/s.williams/zotero_export/MyLibrary.bib
---



```{r angrylibrarian, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)
```


<!-- keywords
# Not yet used, but add keywords as appriopriate. 

insituspatial
seurat
differntialexpression
test

-->


# Overview


Sometimes there are no biological replicates, yet you still want to make a comparison. 
While not ideal, its possible. Individual cells may be treated as 'replicates' to explore the 
difference between two samples.


Note that 'pooling' multiple unlabelled biological samples before library prep will still 
count as 'one' replicate no many how many samples are in the pool - because
we have no way to tell what sample a cell comes from. (NB: Using cell hashing approaches to tag samples before pooling avoids this issue.)



<!--![](assets/cluster_prop_diff.png)-->


This requires:

* Cell clusters
* 2 or more samples (or grouping to compare)


For example:

* What genes are differentially expressed between the knockout (n=1) and control(n=1), for every cell type.
* In my n=1 pilot study - what is differentially expressed between two of my stromal clusters?


Steps:

1. Filter to testable genes (enough expression to see changes)
2. Test for changes in gene expression
3. Plot DE results and individual genes. 



# Worked example 

Data from paper [Forming nephrons promote nephron progenitor maintenance and branching morphogenesis via paracrine BMP4 signalling under the control of Wnt4](https://www.biorxiv.org/content/10.1101/2023.11.19.567482v1.full) [@moreauFormingNephronsPromote2023a]

This study included 10X chromium single cell RNAseq data from 4 conditions, with 3-4 E14.5 mice pooled per group. 

* Sample1 (Wnt4FloxKO):  Wnt4Flox/Flox Six2-Conditional Wnt4 Knockout
* Sample2 (Wnt4FloxHet):  Wnt4Flox/+ Six2-Conditional Wnt4 Het
* Sample3 (Wnt4Het): Wnt4 GCE/+ Control Wnt4 Het
* Sample4 (Wnt4KO): Wnt4 GCE/GCE Knockout Wnt4

In that paper they explain that complete or conditional homozygous knockout of Wnt4 gene results in abnormal kidney development, and they use scRNAseq data to explore effects at cellular level. [@moreauFormingNephronsPromote2023a]


In this vignette, we will test for differential expression in each cell type for two 1 vs 1 comparisons:

* Wnt4KO vs Wnt4Het : Complete knockout vs heterozgote
* Wnt4FloxKO vs Wnt4FloxHet : Conditional (Flox) knockout vs heterozygote


## Load Libraries and Data

```{r}
library(Seurat)
library(edgeR)
library(limma)
library(DT)
library(tidyverse)


dataset_dir      <- '~/projects/spatialsnippets/datasets'
seurat_file      <- file.path(dataset_dir, 'Wnt4KO_Moreau2023',  "Wnt4KOE14.5_11_ss.rds")

so <- readRDS(seurat_file)
```


## Experimental Design

```{r}
select(so@meta.data, sample, Genotype, GTeffect, GTshort) %>% 
  as_tibble() %>% 
  group_by( sample, Genotype, GTeffect, GTshort) %>%
  summarise(num_cells=n(), .groups = 'drop') %>%
  DT::datatable()
```


## Filter



### Counts per cell

A minimum-counts-per-cell threshold was already applied to this dataset (during preprocessing) so nothing to do here.


```{r}
# This is plenty
min(so$nCount_RNA)
```


### Cells per group

How many cells are there for each celltype for each sample? If there are too few on either side of the contrast, we won't be able to test. 


How many is too few? 50 might be a good threshold. In this case however, we might 
consider being very permissive and allowing just 20 cells to look at some celltypes
that are clearly reduced or increased in number between conditions (e.g. c9,c10,c11)
They're interesting in this experiment, but have to keep in mind during interpretation
that results may be less reliable.

Using the 'GTshort' annotation for our sample names.

We will apply that filter when running the differential expression.


```{r}
min_cells_per_group <- 20 # used later
table(so$CelltypeCode, so$GTshort)
```


## Run just one differential expression


Talk about options. ...


* layer vs assay
* SCT vs RNA 
* count vs data vs data scale


## Run differential expression for every cluster


```{r}
# Set threhoehsolds
min_cells_per_group <- 20


# Calculate DE across every celltype
# Empty list to collect results
de_result_list <- list()
de_result_sig_list <- list()

# the contrasts
# If we store them in a list we can give the nice names
contrast_list <-  list( # 'Test'      vs  'control'  
                       FloxWnt4KOvsFloxHet  = c('Wnt4FloxKO',      'Wnt4FloxHet' ) ,
                       TotalWnt4KOvsHet     = c('Wnt4KO',          'Wnt4Het'     ) )

## Or you could autogenerated them
#make_contrast_name <- function(contrast_parts){ paste0(contrast_parts[1],"vs",contrast_parts[1])}
#contrast_names <- sapply(FUN=make_contrast_name, X=contrast_list) 
#names(contrast_list) <- contrast_names


# note keeping all 4 samples before two contrasts. (actually does that matter)

#the_celltype = 'c15: Podocyte'
#the_celltype = "c16: Blood"
#contrast_name <- 'TotalWnt4KOvsHet'
  
for (the_celltype in levels(so$CelltypeCode)[5:10]) {
  
  # Subset to one cell type. 
  print(the_celltype)
  
  so.this <- subset(so, CelltypeCode == the_celltype)
  
  # count how many cells within each sample (GTshort)
  # And list which samples have more than the minimum
  cells_per_sample <- table(so.this$GTshort)
  print(cells_per_sample)
  samples_with_enough_cells <- names(cells_per_sample)[cells_per_sample > min_cells_per_group]
  
  # SUbset to only those samples (might be all cells, might be no cells)
  so.this <- subset(so.this, GTshort %in% samples_with_enough_cells)
  

  # For each listed contrast, do both sides of the contrast have enough cells?
  # This is of course much simpler if you only have one contrast!
  for (contrast_name in names(contrast_list)) {
    
    # from mycontrastname pull out list of the two samples involved;
    # c('test', 'control')
    contrast <- contrast_list[[contrast_name]]
    
    # Only run this contrast if both sides pass!
    if (all(contrast %in% samples_with_enough_cells)) {
      print(contrast_name) 

      # We need to tell Seurat to group by _sample_ not cluster.
      Idents(so.this) <- so.this$GTshort
      
      
      de_result <- FindMarkers(so.this, 
                               ident.1  = contrast[1], 
                               ident.2 = contrast[2],
                               slot     = 'data',
                               test.use = "wilcox", # the default
                               min.pct  = 0.01,   # Note 
                               max.cells.per.ident = 100 # 1000 # if you have really big clusters, set this to subsample!
                               )
  
      
      # It can be helpful to know the average expression of a gene
      # This will give us the average (per cell) within this celltype.
      avg_expression <- rowMeans(GetAssayData(so.this, assay = 'SCT', layer="data"))
      
      
      
      de_result.formatted <- de_result %>%
      rownames_to_column("target") %>%
      mutate(contrast=contrast_name,
             celltype=the_celltype,
             avg_expression=avg_expression[target]) %>%
      select(celltype,contrast,target,avg_expression, everything()) %>%
      arrange(p_val)
  
      # Filter to just significant results, optionally by log2FC.
      de_result.sig <- filter(de_result, 
                         p_val_adj < 0.01,
                         abs(avg_log2FC) > log2(1.5) )

      # Record these results in a list to combine
      full_name <- paste(contrast_name, the_celltype)
      de_result_list[[full_name]] <- de_result.formatted
      de_result_sig_list[[full_name]] <- de_result.sig
      
    }
  }
}

# Join together results for all celltypes, and pull out those with a singificant adjusted p-value
de_results_all <- bind_rows(de_result_list)
de_results_sig <- bind_rows(de_result_sig_list)


```

Check out the full set of significant DE genes

```{r}
DT::datatable(de_results_sig)
```



Save the results.

```{r eval=FALSE}
# Save the full set of results as a tab-deliminated text file.
# This is useful for parsing later e.g. for functional enrichment
write_tsv(x = de_results_all, "~/myproject/de_results_all.tsv")

# If you don't have too many contrasts and celltypes, 
# you can save it as an excel file, with each contrast in a seprate tab.
library(writexl)
write_xlsx(de_result_list,      "~/myproject/de_results_all.xlsx")
write_xlsx(de_result_sig_list,  "~/myproject/de_results_sig.xlsx")
```






# Code Snippet

Copy and paste this bit into your own script to adapt to your own data. 
Make as generic as possible, just a framework.

```{r eval=FALSE}

# This code does not run

```



# Results

```{r}
# Display results table of stats
```


Have a dot point list of what _every_ field in the default output is. Refer to docs wherever possible.

* **Column name 1** : This is xyz
* **Column name 2** : This is xyz



# More information 

List of useful resources. Papers, vignettes, pertinent forum posts 

* [Wnt4 KO in developing mouse kidney - 10X Chromium scRNAseq](d_Wnt4KO.Rmd): [Forming nephrons promote nephron progenitor maintenance and branching morphogenesis via paracrine BMP4 signalling under the control of Wnt4](https://www.biorxiv.org/content/10.1101/2023.11.19.567482v1.full) [@moreauFormingNephronsPromote2023a] : The paper with this data.


# Refereneces

<!-- nothing here, autopopulated from bibliography: in json -->

