---
title: "Spatially resticted genes"
author: "Sarah Williams"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
bibliography: /home/s.williams/zotero_export/MyLibrary.bib
---

```{r angrylibrarian, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)
```


# Overview


It is possible to test for genes that are expressed in a non-random spatial pattern. These might be restricted to regions of a tissue (e.g. epithelia), near structures or simply having very high expression in selected cells only (e.g. immunoglobulins in plasma cells). 

One popular approach to find these genes is the MoransI test of spatial autocorrelation.

![](assets/spatiallyrestricted.png)


This example will show how to use the morans test within Seurat to find spatially restricted genes.


This requires:

* X,Y coordinates of individual transcripts, or their cells.
* Annotation of the tissue samples in the cell metadata (if multiple samples)

There is no need for celltype annotation. 


For example:

* What cells are spatially restricted in this tissue due to some structure?
* Are some cancer-related genes showing a restricted spatial expression (and does this align with tumour-dense regions?)



Steps:

1. Subset experiment to each tissue 
2. Calculate Moran's I for each sample
3. Join and inspect results




# Worked example 

Paper [Microglia-astrocyte crosstalk in the amyloid plaque niche of an Alzheimer’s disease mouse model, as revealed by spatial transcriptomics](https://www.sciencedirect.com/science/article/pii/S2211124724005448?via%3Dihub)[@mallachMicrogliaastrocyteCrosstalkAmyloid2024] explores the spatial transcription of amyloid plaques in a mouse model.

Their work includes an analysis of cosMx samples from of 4 mouse brain samples. 

Those sections show very distinct spatial patterning of gene expression due to the different brain structures. This example will test which genes are expressed in a spatially restricted pattern, independently of any celltype annotations. e.g.

* Present within or nearby a given structure
* Existing in 'clumps' e.g. high expression on a small subset of cells



### Load libraries and data

Load relevant libraries
```{r}
# NB: The Rfast2 and ape packages may need to be installed
# to use the moransI test (in addition to Seurat)
# The GSL system library might also need to be installed, if it isn't already.
# install.packages('Rfast2')
# install.packages('ape')


library(Seurat)
library(tidyverse)
library(DT)


# Certain stps may result in error:
#Error in getGlobalsAndPackages(expr, envir = envir, globals = globals) : 
#  The total size of the 14 globals exported for future expression (‘FUN()’) is 1.21 GiB.. This exceeds the maximum allowed size of 500.00 MiB (option 'future.globals.maxSize'). The three largest globals are ‘FUN’ (824.48 MiB of class ‘function’), ‘x’ (411.83 MiB of class ‘list’) and ‘over’ (940.20 KiB of class ‘function’)
# this can be fixed by increaseing the future.globals.maxsize variable (to an acceptable amount of ram)
options(future.globals.maxSize= 5*1024^3) # Set to 5Gb

```


And load the preprocessed Seurat object.
```{r}
dataset_dir      <- '~/projects/spatialsnippets/datasets/GSE263793_Mallach2024_AlzPlaque/processed_data/'
seurat_file_01_preprocessed <- file.path(dataset_dir,  "GSE263793_AlzPlaque_seurat_01_preprocessed.RDS")

so <- readRDS(seurat_file_01_preprocessed)
```

### Spatially variable features

Morans test can be slow to run, so save time by only running it on variable features. Variable features are those with a non-even distribution across cells, and are
routinely deteced with _FindVariableFeatures()_ during preprocessing. Non variable features are unlikely to be spatially restricted.


For the purpose of this demo, recalculate just the top 10 variable features. The actual number for a real experiment could be judged from the variable features plot below, e.g. 100-200-2000 (or whatever you use for PCA - it really depends on your panel!). 

```{r}
num_variable_features = 30 # Test only! Should be much larger. Consider using all your VariableFeatures used for UMAPs, et.c.
so <-  FindVariableFeatures(so, nfeatures=num_variable_features)
VariableFeaturePlot(so)
```


We will look for _spatially_ variable features on each of our tissue samples independently; in this case 4 samples across 2 slides. But first, just test one.


We do this because all probes are essentially going to be restricted to the tissue itself, not the surrounding empty slide. This might be particularly noticeable on a panel of many small cores.



First, subset to just one tissue sample.

```{r}
so.sample <- subset( so, subset= sample == 'sample1')
```


Then find the spatially variable genes with _FindSpatiallyVariableFeatures()_ function. 

That code _should_ look like this:

```{r eval=FALSE}
so.sample <- FindSpatiallyVariableFeatures(
  so.sample, 
  assay = "RNA", 
  features = VariableFeatures(so.sample),    
  selection.method = "moransi",
  layer = "counts")
```

<!-- ### OLD BUG -->

<!-- However, right now, there is a bug with the current FindSpatiallyVariableFeatures function, described [here](https://github.com/satijalab/seurat/issues/8226). Likely that issue will be closed when this is fixed! -->

<!-- If you see the following error - try a workaround, using a customised version of this that avoids the issue in this dataset. The custom version, may be found [here](https://github.com/swbioinf/spatialsnippets/blob/main/code/spatially_variable_features_code.R) -->

<!-- ``` -->
<!-- Error in `LayerData<-`: -->
<!-- ! 'layer' must be a single non-empty string -->
<!-- ``` -->

<!-- The adjusted function is called as: -->
<!-- ```{r} -->
<!-- # Workaround  -->
<!-- # Available: https://github.com/swbioinf/spatialsnippets/blob/main/code/spatially_variable_features_code.R -->
<!-- source("code/spatially_variable_features_code.R") -->
<!-- so.sample <- FindSpatiallyVariableFeatures.Seurat_EDITED( -->
<!--   so.sample,  -->
<!--   assay = "RNA",  -->
<!--   features = VariableFeatures(so.sample),     -->
<!--   selection.method = "moransi", -->
<!--   layer = "counts") -->
<!-- ``` -->


However, right now, there is a bug with the current _FindSpatiallyVariableFeatures()_ function with this dataset. If you find the R code segfaults (ie. R exits entirely) with the message "*address (nil), cause 'memory not mapped'*", consider the workaroudn below. Also, check issue ticket [here](https://github.com/satijalab/seurat/issues/10099), it may be fixed.

```{r}
so.sample.assay <- so.sample[['RNA']]
# Or equivalently
# so.sample.assay <- GetAssay(so.sample, "RNA")

# Make sure that the rownames of the coordinate table is the cell names
# Might not occur for certain types of data.
tc <- GetTissueCoordinates(so.sample)
rownames(tc)<- tc$cell

# When given an 'assay' FindSpatiallyVariableFeatures returns an assay. Put it back in the seurat object.
so.sample[['RNA']] <- FindSpatiallyVariableFeatures(
  so.sample.assay,
  layer = "scale.data",
  spatial.location = tc,
  features = VariableFeatures(so.sample),
  selection.method = "moransi",
  nfeatures=10 # mark top 10 spatially variable
  )
```



### Resume


Now we have a a seurat object with the moransI scores embedded in the feature metatdata of the 'RNA' assay.


```{r}
gene_metadata <- so.sample[["RNA"]]@meta.data
#NB: This is geme metadata, different to the usual *cell* metadata found at so.sample@meta.data
# so.sample[['RNA']]  retreives the 'RNA' assay.
DT::datatable(head(gene_metadata), width='100%')
```

The whole gene-metadata includes other columns, and in fact the columns we are interested in only have values for the 'variable' genes that we tested. So, make a summary table with just the relevant data.
```{r}
gene_metadata_morans <- 
  filter(gene_metadata, !is.na(moransi.spatially.variable.rank)) %>%
  select(feature, 
         MoransI_observed, MoransI_p.value, moransi.spatially.variable,moransi.spatially.variable.rank) %>% 
  arrange(moransi.spatially.variable.rank)
  
DT::datatable(gene_metadata_morans, width = '100%')
```


### Plot results

We can pull out the most significant genes from that table.

```{r}
top_genes = gene_metadata_morans$feature[1:3]
top_genes
```

Here we are plotting the top 3 genes on that slide. Each has a different but clear reasons for being spatially restricted. Ptgds and Penk seem to be restricted to specific regions of the tissue. Wherease Drd4 seems to have high expression in a subset of cells - its proximity to itself also triggers the significance n the moransI test. 

NB: Genes without any sort of spatial pattern (e.g negative controls) might still have some sort of moransI test significance - since they're still restricted to the tissue itself, it isn't random.


```{r fig.width=10}
ImageDimPlot(so.sample,  fov = "AD2.AD3.CosMx",
             molecules = top_genes, 
             group.by = 'sample', cols = c("grey30"), # Make all cells grey.
             boundaries = "segmentation",
             border.color = 'black', axes = T, crop=TRUE)
```

<!-- ```{r} -->
<!-- # get averages of these 3 genes in  -->
<!-- norm_matrix_means <- FetchData(so.sample, c('cluster_name', 'Ptgds', 'Penk', 'Drd4')) %>% -->
<!--   group_by(cluster_name) %>% -->
<!--   summarise(Ptgds = mean(Ptgds), -->
<!--             Penk  = mean(Penk), -->
<!--             Drd4  = mean(Drd4)) %>% -->
<!--   column_to_rownames(var = 'cluster_name') %>% # turn it into a matrix -->
<!--   as.matrix() -->

<!-- ``` -->
<!-- ```{r} -->
<!-- heatmap(norm_matrix_means, -->
<!--         scale = "none",     # plot normalised expression, no scaling. -->
<!--         Rowv = NA, Colv=NA, # no heirachical clustering -->
<!--         cexCol = 1.5,       # adjust gene name sizes  -->
<!--         margin = c(5,10)    # make window wide enough for labels. -->
<!--         ) -->
<!-- ``` -->

## Run across all samples

Realistically, we would want to test all samples. Here we run the test on each tissue sample separately. 

```{r}
samples <- levels(so@meta.data$sample)
results_list <- list()

for (the_sample in samples) {
  
  so.sample <- subset( so, subset= sample == the_sample)
  
  # Again, this should be: 
  #so.sample <- FindSpatiallyVariableFeatures(
  #  so.sample, 
  #  assay = "RNA", 
  #  features = VariableFeatures(so.sample),    
  #  selection.method = "moransi",
  #  layer = "counts"
  #)
  
  # Workaround running FindSpatiallyVariableFeatures at the assay level
  so.sample.assay <- so.sample[['RNA']]
  tc <- GetTissueCoordinates(so.sample)
  rownames(tc)<- tc$cell
  
  # When given an 'assay' FindSpatiallyVariableFeatures returns an assay. Put it back in the seurat object.
  so.sample[['RNA']] <- FindSpatiallyVariableFeatures(
    so.sample.assay,
    layer = "scale.data",
    spatial.location = tc,
    features = VariableFeatures(so.sample),
    selection.method = "moransi",
    nfeatures=10 # mark top 10 spatially variable
    )
  # --- end workaround
  
  gene_metadata <- so.sample[["RNA"]]@meta.data
  results <- 
  select(gene_metadata, 
         feature, 
         MoransI_observed, 
         MoransI_p.value, 
         moransi.spatially.variable,
         moransi.spatially.variable.rank) %>% 
    filter(!is.na(moransi.spatially.variable.rank)) %>% # only tested
    arrange(moransi.spatially.variable.rank) %>%
    mutate(sample = the_sample) %>%
    select(sample, everything())

  
  results_list[[the_sample]] <- results
}

results_all <- bind_rows(results_list)
```

Display results for variable genes
```{r}
DT::datatable(results_all, width='100%') 
```

Plot some of the moransI scores. We should keep in mind that the shape of the tissue itself limits where the targest can fall, but some genes are clearly more spatially restricted than others. 

```{r}
# unique set of genes that were top 10 morans in any sample
plot_genes <- filter(results_all, moransi.spatially.variable.rank <= 10) %>% pull(feature) %>% unique()
ggplot(filter(results_all, feature %in% plot_genes), aes(x=feature, y=MoransI_observed)) +
  geom_boxplot() +
  geom_point( mapping=aes(col=sample)) +
  theme_bw() +
  ggtitle("MoransI Per sample")

```


## Check an example

Ttr had a much higher Moran's I in sample4 than sample1. Plotting its distribution in both demonstrates the difference - there's a very high expression region in sample4 not seen in sample1.

```{r fig.width=12}
ImageDimPlot(
  subset( so, subset = sample == 'sample1'),  
             fov          = "AD2.AD3.CosMx",
             molecules    = 'Ttr', 
             group.by     = 'sample', cols = c("grey30"), # Make all cells grey.
             boundaries   = "segmentation",
             border.color = 'black', axes = T, crop=TRUE)

ImageDimPlot(
  subset( so, subset = sample == 'sample4'),  
             fov          = "AD4.AD5.CosMx", # note the slide it is on.
             molecules    = 'Ttr', 
             group.by     = 'sample', cols = c("grey30"), # Make all cells grey.
             boundaries   = "segmentation",
             border.color = 'black', axes = T, crop=TRUE)
```





# Code Snippet

Assumes that tissue samples are in a metadata column data called 'sample'.
If there are multiple slides, it may be neccessary to call joinlayers.

```{r eval=FALSE}

library(Seurat)
library(tidyverse)
library(DT)

# Load edited function, see https://github.com/satijalab/seurat/issues/8226
# Available here: https://github.com/swbioinf/spatialsnippets/blob/main/code/spatially_variable_features_code.R
source("spatially_variable_features_code.R")


## If not alread run, find variable features
#num_variable_features = 1000 # Choose based on likely results and acceptable runtime
#so <-  FindVariableFeatures(so, nfeatures=num_variable_features)


# Record moransI results for each sample, one by one.
samples <- levels(so@meta.data$sample)
results_list <- list()
for (the_sample in samples) {
  
  so.sample <- subset( so, subset= sample == the_sample)
  
  
  # TRY THIS FIRST
  # This is the Seurat method, and may have been fixed.
  #so.sample <- FindSpatiallyVariableFeatures(
  #  so.sample, 
  #  assay = "RNA", 
  #  features = VariableFeatures(so.sample),    
  #  selection.method = "moransi",
  #  layer = "counts"
  #)
  
  # workaround
  # If the Seurat method above crashes (segfault), a workaround is
  # funning FindSpatiallyVariableFeatures at the assay level
  so.sample.assay <- so.sample[['RNA']]
  tc <- GetTissueCoordinates(so.sample)
  rownames(tc)<- tc$cell
  # When given an 'assay' FindSpatiallyVariableFeatures returns an assay. Put it back in the seurat object.
  so.sample[['RNA']] <- FindSpatiallyVariableFeatures(
    so.sample.assay,
    layer = "scale.data",
    spatial.location = tc,
    features = VariableFeatures(so.sample),
    selection.method = "moransi",
    nfeatures=10 # mark top 10 spatially variable
    )
  # --- end workaround
  
  
  # Format output table
  gene_metadata <- so.sample[["RNA"]]@meta.data
  results <- 
  select(gene_metadata, 
         feature, 
         MoransI_observed, 
         MoransI_p.value, 
         moransi.spatially.variable,
         moransi.spatially.variable.rank) %>% 
    filter(!is.na(moransi.spatially.variable.rank)) %>% # only tested
    arrange(moransi.spatially.variable.rank) %>%
    mutate(sample = the_sample) %>%
    select(sample, everything())

  
  results_list[[the_sample]] <- results
}

# Collect output result 
results_all <- bind_rows(results_list)

```



# Results

```{r}
DT::datatable(results_all, width='100%')
```

* **sample**: _(not a default column, added by code)_: What tissue sample the test was run on.
* **feature** : The gene being tested
* **MoransI_observed** : The moransI statistic calculated. Higher values indicate more spatial correlation, 0 is completely random, and negative values indicate anti-correlation (ie repulsion). 
* **MoransI_p.value** : P-value for the moransI test.
* **moransi.spatially.variable** : Is this gene spatially restricted? True or false value. The number of TRUEs is controlled by the nfeatures parameter of _FindSpatiallyVariableFeatures_.
* **moransi.spatially.variable.rank** : Ranking of the genes by spatial correlation, where 1 is the most distincly spatially restricted.




# More information 


* [Microglia-astrocyte crosstalk in the amyloid plaque niche of an Alzheimer’s disease mouse model, as revealed by spatial transcriptomics](https://www.sciencedirect.com/science/article/pii/S2211124724005448?via%3Dihub): Data used in this example. [@mallachMicrogliaastrocyteCrosstalkAmyloid2024]
* [MoransI wikipedia](https://en.wikipedia.org/wiki/Moran%27s_I): What is MoransI test, with pictures. 
* [Seurat Spatially variable features](https://satijalab.org/seurat/articles/spatial_vignette): This is actually the sequencing based technology vignette (for visium data), but it covers the FindSpatiallyVariableFeatures function
* [FindSpatiallyVariableFeatures() Bug report](https://github.com/satijalab/seurat/issues/8226) : Link to the bug report on the Seurat repo in github. Can check the status of this issue here.
* [Voyager toolkit - spatial statistics for cosmx vignette](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#local-spatial-statistics-of-marker-genes) : Here is a more focussed and detailed vignette on calculating spatial statistics for gene expression using the Voyager package. It does not use Seurat objects, it is rather using the bioconductor-compatible SpatialExperiment class.

# References





